
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>dms_merge &#8212; dms_tools 1.2.2 documentation</title>
    <link rel="stylesheet" href="_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.2.2',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="dms_correlate" href="dms_correlate.html" />
    <link rel="prev" title="dms_logoplot" href="dms_logoplot.html" /> 
  </head>
  <body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="dms_correlate.html" title="dms_correlate"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="dms_logoplot.html" title="dms_logoplot"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">dms_tools 1.2.2 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="programs.html" accesskey="U">Programs</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="dms-merge">
<span id="id1"></span><h1><a class="toc-backref" href="#id2"><code class="docutils literal"><span class="pre">dms_merge</span></code></a><a class="headerlink" href="#dms-merge" title="Permalink to this headline">¶</a></h1>
<div class="contents topic" id="contents">
<p class="topic-title first">Contents</p>
<ul class="simple">
<li><a class="reference internal" href="#dms-merge" id="id2"><code class="docutils literal"><span class="pre">dms_merge</span></code></a><ul>
<li><a class="reference internal" href="#overview" id="id3">Overview</a></li>
<li><a class="reference internal" href="#command-line-usage" id="id4">Command-line usage</a><ul>
<li><a class="reference internal" href="#Positional Arguments" id="id5">Positional Arguments</a></li>
<li><a class="reference internal" href="#Named Arguments" id="id6">Named Arguments</a></li>
</ul>
</li>
<li><a class="reference internal" href="#examples" id="id7">Examples</a><ul>
<li><a class="reference internal" href="#rescaling-preferences-by-a-stringency-parameter" id="id8">Rescaling preferences by a stringency parameter</a></li>
<li><a class="reference internal" href="#averaging-differential-selection" id="id9">Averaging differential selection</a></li>
<li><a class="reference internal" href="#averaging-preferences" id="id10">Averaging preferences</a></li>
<li><a class="reference internal" href="#adding-and-subtracting-preferences" id="id11">Adding and subtracting preferences</a></li>
<li><a class="reference internal" href="#adding-counts" id="id12">Adding counts</a></li>
<li><a class="reference internal" href="#adding-counts-with-normalization" id="id13">Adding counts with normalization</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="section" id="overview">
<h2><a class="toc-backref" href="#id3">Overview</a><a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">dms_merge</span></code> is a program included with the <a class="reference external" href="http://jbloomlab.github.io/dms_tools">dms_tools</a> package. It merges preferences, differential selection values, or differential preferences in different files by averaging them or adding / subtracting them. It also sums the counts from count files together. See <a class="reference internal" href="#examples">Examples</a> for illustrations of how you might do this.</p>
<p>After you install <a class="reference external" href="http://jbloomlab.github.io/dms_tools">dms_tools</a>, this program will be available to run at the command line.</p>
</div>
<div class="section" id="command-line-usage">
<h2><a class="toc-backref" href="#id4">Command-line usage</a><a class="headerlink" href="#command-line-usage" title="Permalink to this headline">¶</a></h2>
<p>Merge preferences, differential selection, or differential preferences by averaging or adding / subtracting the values in multiple files. Alternatively, sum counts by adding the counts for the characters at each site across multiple files. All files must specify the same character type: can be nucleotide, codon, or amino acid (see “–excludestop” if using amino acids). This script is part of dms_tools (version 1.1.20) written by the Bloom Lab (see <a class="reference external" href="https://github.com/jbloomlab/dms_tools/graphs/contributors">https://github.com/jbloomlab/dms_tools/graphs/contributors</a> for all contributors). Detailed documentation is at <a class="reference external" href="http://jbloomlab.github.io/dms_tools/">http://jbloomlab.github.io/dms_tools/</a></p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">dms_merge</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">excludestop</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">minus</span> <span class="n">MINUS</span> <span class="p">[</span><span class="n">MINUS</span> <span class="o">...</span><span class="p">]]</span>
                 <span class="p">[</span><span class="o">--</span><span class="n">sitediffselfile</span> <span class="n">SITEDIFFSELFILE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">v</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">--</span><span class="n">stringencyparameter</span> <span class="n">STRINGENCYPARAMETER</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">normalize</span><span class="p">]</span>
                 <span class="p">[</span><span class="o">--</span><span class="n">chartype</span> <span class="p">{</span><span class="n">DNA</span><span class="p">,</span><span class="n">codon</span><span class="p">,</span><span class="n">aa</span><span class="p">}]</span>
                 <span class="n">outfile</span> <span class="p">{</span><span class="n">average</span><span class="p">,</span><span class="n">median</span><span class="p">,</span><span class="nb">sum</span><span class="p">,</span><span class="n">rescale</span><span class="p">}</span> <span class="n">infiles</span> <span class="p">[</span><span class="n">infiles</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
<div class="section" id="Positional Arguments">
<h3><a class="toc-backref" href="#id5">Positional Arguments</a><a class="headerlink" href="#Positional Arguments" title="Permalink to this headline">¶</a></h3>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd>outfile</kbd></td>
<td><p class="first">Created output file with merged values; removed if it already exists. If the “infiles” do not all have the same wildtype residue at a site, then the wildtype is indicated as “?” in “outfile”.</p>
<p>For merge_method of “average”, then this file will be of the same type as infiles (preferences, mutation-level differential selection values, or differential preferences). For merging differential selection, any values of NaN are ignored.</p>
<p>For merge_method of “median”, then infiles and outfile must be of the format of the <a href="#id1"><span class="problematic" id="id2">*</span></a>mutdiffsel.txt files created by dms_diffselection.</p>
<p>For merge_method of “sum”, this file may be either preferences, differential preferences, or counts:</p>
<p>If any of the files in infiles and –minus give preferences and the sum at each site is one, then outfile gives preferences.</p>
<p>If any of the files in infiles and –minus give preferences and the sum at each site is zero, then outfile gives differential preferences.</p>
<p>If all of the files in infiles and –minus are differential preferences, the outfile gives differential preferences.</p>
<p>If all of the files in infiles are counts, the outfile gives counts.</p>
<p>For merge_method of “rescale”, this created file will give preferences (after re-scaling).</p>
<p>If outfile gives preferences, it will be in the preferences_file format. Note that this file will not have any columns giving 95% credible intervals, as these cannot be calculated when merging files.</p>
<p>If outfile gives differential preferences, it will be in the diffpreferences_file format. Note that this file will not have any columns giving posterior probabilities, as these cannot be calculated when merging files.</p>
<p>If outfile gives mutation-level differential selection values, it will be in the format of the <a href="#id3"><span class="problematic" id="id4">*</span></a>mutdiffsel.txt file created by dms_diffselection. To create a corresponding <a href="#id5"><span class="problematic" id="id6">*</span></a>sitediffsel.txt file, see the –sitediffselfile option.</p>
<p class="last">If outfile gives counts, it will be in the dms_counts format.</p>
</td></tr>
<tr><td class="option-group">
<kbd>merge_method</kbd></td>
<td><p class="first">Possible choices: average, median, sum, rescale</p>
<p class="last">How to merge: If “average” then all “infiles” must specify preferences, differential selection on mutations (mutdiffsel.txt file), or differential preferences; these are then averaged (by taking the mean) in “outfile”. If “median”, then all “infiles” must specify differential selection on mutations (mutdiffsel.txt file); these are then averaged by taking the median. If “sum” then “infiles” can either all be count files, all be differential preferences, or can be a combination of preferences and differential preferences that (along with any additional files specified by “–minus”) sum to a total preference or a total differential preference of zero at each site. If “rescale”, then only one infile can be specified and it must be preferences; these preferences are then scaled by the value provided by “stringencyparameter”</p>
</td></tr>
<tr><td class="option-group">
<kbd>infiles</kbd></td>
<td><p class="first">Files to average or sum. Must all have the same sites and character type, but do not need to have the same wildtype residue at each site. Note that for differential selection, must be the <a href="#id1"><span class="problematic" id="id2">*</span></a>mutdiffsel.txt file.</p>
<p class="last">Note that for differential selection, the files can only be the <a href="#id3"><span class="problematic" id="id4">*</span></a>mutdiffsel.txt files created by dms_diffselection, not the <a href="#id5"><span class="problematic" id="id6">*</span></a>sitediffsel.txt files. However, you can create a <a href="#id7"><span class="problematic" id="id8">*</span></a>sitediffsel.txt file from the merged <a href="#id9"><span class="problematic" id="id10">*</span></a>mutdiffsel.txt files using the –sitediffselfile option.</p>
</td></tr>
</tbody>
</table>
</div>
<div class="section" id="Named Arguments">
<h3><a class="toc-backref" href="#id6">Named Arguments</a><a class="headerlink" href="#Named Arguments" title="Permalink to this headline">¶</a></h3>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd>--excludestop</kbd></td>
<td><p class="first">If we are using amino acids, do we remove stop codons (denoted by “*”)? We only remove stop codons if this argument is specified. If this option is used, then any files with stop codons have these codons removed (re-normalizing preferences to sum to one, and differential preferences to sum to zero) before the merge.</p>
<p class="last">Default: False</p>
</td></tr>
<tr><td class="option-group">
<kbd>--minus</kbd></td>
<td><p class="first">Files to subtract when summing. Can only be used if “merge_method” is “sum” and if files are either preferences or differential preferences.</p>
<p class="last">Files should be in the formats of a preferences_file or a diffpreferences_file. Currently, this option cannot be used with counts files.</p>
</td></tr>
<tr><td class="option-group" colspan="2">
<kbd>--sitediffselfile</kbd></td>
</tr>
<tr><td>&#160;</td><td><p class="first">If merging differential selection, ‘infiles’ and ‘outfile’ are the ‘<a href="#id1"><span class="problematic" id="id2">*</span></a>mutdiffsel.txt’ file. If you also want to create a ‘<a href="#id3"><span class="problematic" id="id4">*</span></a>sitdiffsel.txt’ file for the merged ‘<a href="#id5"><span class="problematic" id="id6">*</span></a>mutdiffsel.txt’ file, specify the name of that file here.</p>
<p class="last">After merging the mutation-level differential selection values in <a href="#id7"><span class="problematic" id="id8">*</span></a>mutdiffsel.txt infiles, you can create a new <a href="#id9"><span class="problematic" id="id10">*</span></a>sitediffsel.txt file (in the format of those created by dms_diffselection with this option.</p>
</td></tr>
<tr><td class="option-group">
<kbd>-v, --version</kbd></td>
<td>show program’s version number and exit</td></tr>
<tr><td class="option-group" colspan="2">
<kbd>--stringencyparameter</kbd></td>
</tr>
<tr><td>&#160;</td><td>Stringency parameter used to rescale preferences. Can only be used if “merge_method” is “rescale”.</td></tr>
<tr><td class="option-group">
<kbd>--normalize</kbd></td>
<td><p class="first">Whether to normalize the counts at each site to the minimum number of counts observed at that site across all provided count files, before summing the counts. Can only be used if “merge_method” is “sum”.</p>
<p class="last">Default: False</p>
</td></tr>
<tr><td class="option-group">
<kbd>--chartype</kbd></td>
<td><p class="first">Possible choices: DNA, codon, aa</p>
<p>Characters for which counts are summed: “DNA” = counts for DNA; “codon” = counts for codons; “aa” = counts for amino acids (possibly including stop codons, see “–excludestop”). This option only needs to be specified when summing count files.</p>
<p class="last">Default: “codon”</p>
</td></tr>
</tbody>
</table>
</div>
</div>
<div class="section" id="examples">
<h2><a class="toc-backref" href="#id7">Examples</a><a class="headerlink" href="#examples" title="Permalink to this headline">¶</a></h2>
<div class="section" id="rescaling-preferences-by-a-stringency-parameter">
<h3><a class="toc-backref" href="#id8">Rescaling preferences by a stringency parameter</a><a class="headerlink" href="#rescaling-preferences-by-a-stringency-parameter" title="Permalink to this headline">¶</a></h3>
<p>Use:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">dms_merge</span> <span class="n">rescaledprefs</span><span class="o">.</span><span class="n">txt</span> <span class="n">rescale</span> <span class="n">prefs</span><span class="o">.</span><span class="n">txt</span> <span class="o">--</span><span class="n">stringencyparameter</span> <span class="mf">2.5</span>
</pre></div>
</div>
<p>to rescale the preferences in <code class="docutils literal"><span class="pre">prefs.txt</span></code> by a stringency parameter of <span class="math">\(beta = 2.5\)</span>. The created file <code class="docutils literal"><span class="pre">rescaledprefs.txt</span></code> will be in the <a class="reference internal" href="fileformats.html#preferences-file"><span class="std std-ref">Preferences file</span></a> format.</p>
</div>
<div class="section" id="averaging-differential-selection">
<h3><a class="toc-backref" href="#id9">Averaging differential selection</a><a class="headerlink" href="#averaging-differential-selection" title="Permalink to this headline">¶</a></h3>
<p>Use:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">dms_merge</span> <span class="n">avgmutdiffsel</span><span class="o">.</span><span class="n">txt</span> <span class="n">average</span> <span class="n">mutdiffsel1</span><span class="o">.</span><span class="n">txt</span> <span class="n">mutdiffsel2</span><span class="o">.</span><span class="n">txt</span>
</pre></div>
</div>
<p>to average the two mutation differential selection files to create <code class="docutils literal"><span class="pre">avgmutdiffsel.txt</span></code> in the format of a <code class="docutils literal"><span class="pre">*mutdiffsel.txt</span></code> file from <a class="reference internal" href="dms_diffselection.html#dms-diffselection"><span class="std std-ref">dms_diffselection</span></a>. If you also want to create a <code class="docutils literal"><span class="pre">*sitediffsel.txt</span></code> file corresponding to <code class="docutils literal"><span class="pre">avgmutdiffsel.txt</span></code>, use:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">dms_merge</span> <span class="n">avgmutdiffsel</span><span class="o">.</span><span class="n">txt</span> <span class="n">average</span> <span class="n">mutdiffsel1</span><span class="o">.</span><span class="n">txt</span> <span class="n">mutdiffsel2</span><span class="o">.</span><span class="n">txt</span> <span class="o">--</span><span class="n">sitediffselfile</span> <span class="n">avgsitediffsel</span><span class="o">.</span><span class="n">txt</span>
</pre></div>
</div>
<p>to create <code class="docutils literal"><span class="pre">avgsitediffsel.txt</span></code> with the site differential selection values in the format created by <a class="reference internal" href="dms_diffselection.html#dms-diffselection"><span class="std std-ref">dms_diffselection</span></a>.</p>
</div>
<div class="section" id="averaging-preferences">
<h3><a class="toc-backref" href="#id10">Averaging preferences</a><a class="headerlink" href="#averaging-preferences" title="Permalink to this headline">¶</a></h3>
<p>Use:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">dms_merge</span> <span class="n">avgprefs</span><span class="o">.</span><span class="n">txt</span> <span class="n">average</span> <span class="n">prefs1</span><span class="o">.</span><span class="n">txt</span> <span class="n">prefs2</span><span class="o">.</span><span class="n">txt</span>
</pre></div>
</div>
<p>The created file <code class="docutils literal"><span class="pre">avgprefs.txt</span></code> will have the <a class="reference internal" href="fileformats.html#preferences-file"><span class="std std-ref">Preferences file</span></a> format.</p>
</div>
<div class="section" id="adding-and-subtracting-preferences">
<h3><a class="toc-backref" href="#id11">Adding and subtracting preferences</a><a class="headerlink" href="#adding-and-subtracting-preferences" title="Permalink to this headline">¶</a></h3>
<p>The command:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">dms_merge</span> <span class="n">summedprefs</span><span class="o">.</span><span class="n">txt</span> <span class="nb">sum</span> <span class="n">prefs1</span><span class="o">.</span><span class="n">txt</span> <span class="n">prefs2</span><span class="o">.</span><span class="n">txt</span> <span class="o">--</span><span class="n">minus</span> <span class="n">prefs3</span><span class="o">.</span><span class="n">txt</span>
</pre></div>
</div>
<p>would create an output file <code class="docutils literal"><span class="pre">summedprefs.txt</span></code> in the <a class="reference internal" href="fileformats.html#preferences-file"><span class="std std-ref">Preferences file</span></a> format.</p>
<p>The command:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">dms_merge</span> <span class="n">summedprefs</span><span class="o">.</span><span class="n">txt</span> <span class="nb">sum</span> <span class="n">prefs1</span><span class="o">.</span><span class="n">txt</span> <span class="n">prefs2</span><span class="o">.</span><span class="n">txt</span>
</pre></div>
</div>
<p>is <strong>invalid</strong> since it does not create preferences that sum to one at each site (instead they sum to two).</p>
<p>The command:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">dms_merge</span> <span class="n">diffs</span><span class="o">.</span><span class="n">txt</span> <span class="nb">sum</span> <span class="n">prefs1</span><span class="o">.</span><span class="n">txt</span> <span class="n">prefs2</span><span class="o">.</span><span class="n">txt</span> <span class="o">--</span><span class="n">minus</span> <span class="n">prefs3</span><span class="o">.</span><span class="n">txt</span> <span class="n">prefs4</span><span class="o">.</span><span class="n">txt</span>
</pre></div>
</div>
<p>would create a file <code class="docutils literal"><span class="pre">diffs.txt</span></code> in the <a class="reference internal" href="fileformats.html#diffpreferences-file"><span class="std std-ref">Differential preferences file</span></a> format since the total for each site sums to zero.</p>
</div>
<div class="section" id="adding-counts">
<h3><a class="toc-backref" href="#id12">Adding counts</a><a class="headerlink" href="#adding-counts" title="Permalink to this headline">¶</a></h3>
<p>The command:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">dms_merge</span> <span class="n">summedcounts</span><span class="o">.</span><span class="n">txt</span> <span class="nb">sum</span> <span class="n">counts1</span><span class="o">.</span><span class="n">txt</span> <span class="n">counts2</span><span class="o">.</span><span class="n">txt</span> <span class="o">--</span><span class="n">chartype</span> <span class="n">codon</span>
</pre></div>
</div>
<p>sums counts from the two provided counts files and writes them to the file <code class="docutils literal"><span class="pre">summedcounts.txt</span></code> in the <a class="reference internal" href="fileformats.html#dms-counts"><span class="std std-ref">Deep mutational scanning counts file</span></a> format.</p>
</div>
<div class="section" id="adding-counts-with-normalization">
<h3><a class="toc-backref" href="#id13">Adding counts with normalization</a><a class="headerlink" href="#adding-counts-with-normalization" title="Permalink to this headline">¶</a></h3>
<p>The command:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">dms_merge</span> <span class="n">summedcounts</span><span class="o">.</span><span class="n">txt</span> <span class="nb">sum</span> <span class="n">counts1</span><span class="o">.</span><span class="n">txt</span> <span class="n">counts2</span><span class="o">.</span><span class="n">txt</span> <span class="o">--</span><span class="n">chartype</span> <span class="n">codon</span> <span class="o">--</span><span class="n">normalize</span>
</pre></div>
</div>
<p>normalizes counts at each site to the minimum number of total counts observed at that site in <code class="docutils literal"><span class="pre">counts1.txt</span></code> and <code class="docutils literal"><span class="pre">counts2.txt</span></code>. The normalized counts are then summed and written to the file <code class="docutils literal"><span class="pre">summedcounts.txt</span></code> in the <a class="reference internal" href="fileformats.html#dms-counts"><span class="std std-ref">Deep mutational scanning counts file</span></a> format.</p>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#"><code class="docutils literal"><span class="pre">dms_merge</span></code></a><ul>
<li><a class="reference internal" href="#overview">Overview</a></li>
<li><a class="reference internal" href="#command-line-usage">Command-line usage</a><ul>
<li><a class="reference internal" href="#Positional Arguments">Positional Arguments</a></li>
<li><a class="reference internal" href="#Named Arguments">Named Arguments</a></li>
</ul>
</li>
<li><a class="reference internal" href="#examples">Examples</a><ul>
<li><a class="reference internal" href="#rescaling-preferences-by-a-stringency-parameter">Rescaling preferences by a stringency parameter</a></li>
<li><a class="reference internal" href="#averaging-differential-selection">Averaging differential selection</a></li>
<li><a class="reference internal" href="#averaging-preferences">Averaging preferences</a></li>
<li><a class="reference internal" href="#adding-and-subtracting-preferences">Adding and subtracting preferences</a></li>
<li><a class="reference internal" href="#adding-counts">Adding counts</a></li>
<li><a class="reference internal" href="#adding-counts-with-normalization">Adding counts with normalization</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="dms_logoplot.html"
                        title="previous chapter"><code class="docutils literal"><span class="pre">dms_logoplot</span></code></a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="dms_correlate.html"
                        title="next chapter"><code class="docutils literal"><span class="pre">dms_correlate</span></code></a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/dms_merge.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="dms_correlate.html" title="dms_correlate"
             >next</a> |</li>
        <li class="right" >
          <a href="dms_logoplot.html" title="dms_logoplot"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">dms_tools 1.2.2 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="programs.html" >Programs</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2014, Jesse Bloom.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.6.3.
    </div>
  </body>
</html>