
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>dms_subassemble &#8212; dms_tools 1.2.2 documentation</title>
    <link rel="stylesheet" href="_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.2.2',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="dms_matchsubassembledbarcodes" href="dms_matchsubassembledbarcodes.html" />
    <link rel="prev" title="dms_barcodedsubamplicons" href="dms_barcodedsubamplicons.html" /> 
  </head>
  <body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="dms_matchsubassembledbarcodes.html" title="dms_matchsubassembledbarcodes"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="dms_barcodedsubamplicons.html" title="dms_barcodedsubamplicons"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">dms_tools 1.2.2 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="programs.html" accesskey="U">Programs</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="dms-subassemble">
<span id="id1"></span><h1><a class="toc-backref" href="#id2"><code class="docutils literal"><span class="pre">dms_subassemble</span></code></a><a class="headerlink" href="#dms-subassemble" title="Permalink to this headline">¶</a></h1>
<div class="contents topic" id="contents">
<p class="topic-title first">Contents</p>
<ul class="simple">
<li><a class="reference internal" href="#dms-subassemble" id="id2"><code class="docutils literal"><span class="pre">dms_subassemble</span></code></a><ul>
<li><a class="reference internal" href="#overview" id="id3">Overview</a></li>
<li><a class="reference internal" href="#subassembly-sequencing" id="id4">Subassembly sequencing</a></li>
<li><a class="reference internal" href="#subassembly-algorithm" id="id5">Subassembly algorithm</a></li>
<li><a class="reference internal" href="#command-line-usage" id="id6">Command-line usage</a><ul>
<li><a class="reference internal" href="#Positional Arguments" id="id7">Positional Arguments</a></li>
<li><a class="reference internal" href="#Named Arguments" id="id8">Named Arguments</a></li>
</ul>
</li>
<li><a class="reference internal" href="#output-files" id="id9">Output files</a><ul>
<li><a class="reference internal" href="#log-file" id="id10">Log file</a></li>
<li><a class="reference internal" href="#subassembled-variants-file" id="id11">Subassembled variants file</a></li>
<li><a class="reference internal" href="#all-reads-by-barcode-file" id="id12">All reads by barcode file</a></li>
<li><a class="reference internal" href="#summary-statistics-file" id="id13">Summary statistics file</a></li>
<li><a class="reference internal" href="#alignable-reads-per-barcode-file" id="id14">Alignable reads per barcode file</a></li>
<li><a class="reference internal" href="#mutations-among-subassembled-variants-file" id="id15">Mutations among subassembled variants file</a></li>
<li><a class="reference internal" href="#read-start-sites-file" id="id16">Read start sites file</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="section" id="overview">
<h2><a class="toc-backref" href="#id3">Overview</a><a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<p>This program can process the FASTQ reads generated by <a class="reference internal" href="#subassembly-sequencing">Subassembly sequencing</a> to subassemble gene variants identified by a unique random barcode sequence at their 3’ end. It provides linkage information across sites.</p>
<p>After you install <a class="reference external" href="http://jbloomlab.github.io/dms_tools">dms_tools</a>, this program will be available to run at the command line.</p>
</div>
<div class="section" id="subassembly-sequencing">
<h2><a class="toc-backref" href="#id4">Subassembly sequencing</a><a class="headerlink" href="#subassembly-sequencing" title="Permalink to this headline">¶</a></h2>
<p>Subassembly is a technique to use short read sequencing to sequence longer sequences (e.g. full-length genes) by attached a unique barcode (a string of <code class="docutils literal"><span class="pre">N</span></code> nucleotides) to each variant of the sequence, and then associating short reads that span the entire sequence with this barcode to build up the full sequence. The technique is originally described in <a class="reference external" href="http://www.nature.com/nmeth/journal/v7/n2/full/nmeth.1416.html">Hiatt et al (2010)</a>.</p>
<p>The <code class="docutils literal"><span class="pre">dms_subassemble</span></code> program currently works for the case in which the unique barcode is at the 3’ end of the gene, and the fragments to be subassembled are generated by PCR at defined locations. The R1 read captures the barcode, and the R2 read begins at one of the defined PCR locations and captures part of the variant’s sequence.</p>
<p>Here we give an example of the experimental workflow to subassemble <em>recA</em> from <em>E. coli</em>.</p>
<p>In order to subassemble the gene, it is necessary to generate appropriately sized subamplicons that contain the barcode sequence at their 3’ ends (if the subamplicons are too large, then they won’t form clusters efficiently). To do this, the plasmids were cut with XbaI &amp; NdeI or XbaI &amp; BstEII to remove 640 bp or 587 bp upstream of the barcode, respectively. The digested plasmids were then recircularized and used as template for the subamplicons spanning nucleotide positions -1 through 545. Uncut plasmid was used as template for the remaining subamplicons, which span nucleotide positions 481 through the terminator and barcode sequences downstream of the recA gene. Each subamplicon is amplified with a common reverse primer annealing to an Illumina adapter sequence that has been incorporated into the plasmid immediately downstream of the barcode sequence, and with one of 7 different forward primers that anneal to different regions of the recA gene. The forward primers were designed such that the distance from the 3’-end of one primer is <span class="math">\(\le 200\)</span> bp from the 3’-end of the next forward primer. The template and primer combinations used to generate the subamplicons are shown below:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>                         Rnd1F-1: 5’-GTGACTGGAGTTCAGACGTGTGCTCTTCCGATCTccggcatgacaggagtaaa-3’                                                                                                        Rnd1F172: 5’-GTGACTGGAGTTCAGACGTGTGCTCTTCCGATCTcttggggcaggtggtctg                                                                                                                    Rnd1F354: 5’-GTGACTGGAGTTCAGACGTGTGCTCTTCCGATCTtcgacaacctgctgtgctc
  XbaI &amp; NdeI digested recA with coding sequence in caps: 5’-ccggtattacccggcatgacaggagtaaaaATGGCTATCGACGAAAACAAACAGAAAGCGTTGGCGGCAGCACTGGGCCAGATTGAGAAACAATTTGGTAAAGGCTCCATCATGCGCCTGGGTGAAGACCGTTCCATGGATGTGGAAACCATCTCTACCGGTTCGCTTTCACTGGATATCGCGCTTGGGGCAGGTGGTCTGCCGATGGGCCGTATCGTCGAAATCTACGGACCGGAATCTTCCGGTAAAACCACGCTGACGCTGCAGGTGATCGCCGCAGCGCAGCGTGAAGGTAAAACCTGTGCGTTTATCGATGCTGAACACGCGCTGGACCCAATCTACGCACGTAAACTGGGCGTCGATATCGACAACCTGCTGTGCTCCCAGCCGGACACCGGCGAGCAGGCACTGGAAATCTGTGACGCCCTGGCGCGTTCTGGCGCAGTAGACGTTATCGTCGTTGACTCCGTGGCGGCACTGACGCCGAAAGCGGAAATCGAAGGCGAAATCGGCGACTCTCATActagaNNNNNNNNNNNNNNNNNNagatcggaagagcgtcgtgtagggaaagagtgt-3’
  XbaI &amp; NdeI digested recA with coding sequence in caps: 3’-ggccataatgggccgtactgtcctcattttTACCGATAGCTGCTTTTGTTTGTCTTTCGCAACCGCCGTCGTGACCCGGTCTAACTCTTTGTTAAACCATTTCCGAGGTAGTACGCGGACCCACTTCTGGCAAGGTACCTACACCTTTGGTAGAGATGGCCAAGCGAAAGTGACCTATAGCGCGAACCCCGTCCACCAGACGGCTACCCGGCATAGCAGCTTTAGATGCCTGGCCTTAGAAGGCCATTTTGGTGCGACTGCGACGTCCACTAGCGGCGTCGCGTCGCACTTCCATTTTGGACACGCAAATAGCTACGACTTGTGCGCGACCTGGGTTAGATGCGTGCATTTGACCCGCAGCTATAGCTGTTGGACGACACGAGGGTCGGCCTGTGGCCGCTCGTCCGTGACCTTTAGACACTGCGGGACCGCGCAAGACCGCGTCATCTGCAATAGCAGCAACTGAGGCACCGCCGTGACTGCGGCTTTCGCCTTTAGCTTCCGCTTTAGCCGCTGAGAGTATgatctNNNNNNNNNNNNNNNNNNtctagccttctcgcagcacatccctttctcaca-5’
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     Rnd1R: 3’-TCTAGCCTTCTCGCAGCACATCCCTTTCTCACA-5’

                         Rnd1F-1: 5’-GTGACTGGAGTTCAGACGTGTGCTCTTCCGATCTccggcatgacaggagtaaa-3’                                                                                                        Rnd1F172: 5’-GTGACTGGAGTTCAGACGTGTGCTCTTCCGATCTcttggggcaggtggtctg                                                                                                                    Rnd1F354: 5’-GTGACTGGAGTTCAGACGTGTGCTCTTCCGATCTtcgacaacctgctgtgctc
XbaI &amp; BstEII digested recA with coding sequence in caps: 5’-ccggtattacccggcatgacaggagtaaaaATGGCTATCGACGAAAACAAACAGAAAGCGTTGGCGGCAGCACTGGGCCAGATTGAGAAACAATTTGGTAAAGGCTCCATCATGCGCCTGGGTGAAGACCGTTCCATGGATGTGGAAACCATCTCTACCGGTTCGCTTTCACTGGATATCGCGCTTGGGGCAGGTGGTCTGCCGATGGGCCGTATCGTCGAAATCTACGGACCGGAATCTTCCGGTAAAACCACGCTGACGCTGCAGGTGATCGCCGCAGCGCAGCGTGAAGGTAAAACCTGTGCGTTTATCGATGCTGAACACGCGCTGGACCCAATCTACGCACGTAAACTGGGCGTCGATATCGACAACCTGCTGTGCTCCCAGCCGGACACCGGCGAGCAGGCACTGGAAATCTGTGACGCCCTGGCGCGTTCTGGCGCAGTAGACGTTATCGTCGTTGACTCCGTGGCGGCACTGACGCCGAAAGCGGAAATCGAAGGCGAAATCGGCGACTCTCATATGGGCCTTGCGGCACGTATGATGAGCCAGGCGATGCGTAAGCTGGCGGGTAACctagaNNNNNNNNNNNNNNNNNNagatcggaagagcgtcgtgtagggaaagagtgt-3’
XbaI &amp; BstEII digested recA with coding sequence in caps: 3’-ggccataatgggccgtactgtcctcattttTACCGATAGCTGCTTTTGTTTGTCTTTCGCAACCGCCGTCGTGACCCGGTCTAACTCTTTGTTAAACCATTTCCGAGGTAGTACGCGGACCCACTTCTGGCAAGGTACCTACACCTTTGGTAGAGATGGCCAAGCGAAAGTGACCTATAGCGCGAACCCCGTCCACCAGACGGCTACCCGGCATAGCAGCTTTAGATGCCTGGCCTTAGAAGGCCATTTTGGTGCGACTGCGACGTCCACTAGCGGCGTCGCGTCGCACTTCCATTTTGGACACGCAAATAGCTACGACTTGTGCGCGACCTGGGTTAGATGCGTGCATTTGACCCGCAGCTATAGCTGTTGGACGACACGAGGGTCGGCCTGTGGCCGCTCGTCCGTGACCTTTAGACACTGCGGGACCGCGCAAGACCGCGTCATCTGCAATAGCAGCAACTGAGGCACCGCCGTGACTGCGGCTTTCGCCTTTAGCTTCCGCTTTAGCCGCTGAGAGTATACCCGGAACGCCGTGCATACTACTCGGTCCGCTACGCATTCGACCGCCCATTGgatctNNNNNNNNNNNNNNNNNNtctagccttctcgcagcacatccctttctcaca-5’
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          Rnd1R: 3’-TCTAGCCTTCTCGCAGCACATCCCTTTCTCACA-5’

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         Rnd1F481: 5’-GTGACTGGAGTTCAGACGTGTGCTCTTCCGATCTggaaatcgaaggcgaaatc-3’                                                                                   Rnd1F633: 5’-GTGACTGGAGTTCAGACGTGTGCTCTTCCGATCTgcaacccggaaaccactac-3’                                                                                                          Rnd1F809: 5’-GTGACTGGAGTTCAGACGTGTGCTCTTCCGATCTcctctacggcgaaggtatca-3’                                                              Rnd1F941: 5’-GTGACTGGAGTTCAGACGTGTGCTCTTCCGATCTtgcctggctgaaagataacc-3’
               E. coli recA with coding sequence in caps: 5’-ccggtattacccggcatgacaggagtaaaaATGGCTATCGACGAAAACAAACAGAAAGCGTTGGCGGCAGCACTGGGCCAGATTGAGAAACAATTTGGTAAAGGCTCCATCATGCGCCTGGGTGAAGACCGTTCCATGGATGTGGAAACCATCTCTACCGGTTCGCTTTCACTGGATATCGCGCTTGGGGCAGGTGGTCTGCCGATGGGCCGTATCGTCGAAATCTACGGACCGGAATCTTCCGGTAAAACCACGCTGACGCTGCAGGTGATCGCCGCAGCGCAGCGTGAAGGTAAAACCTGTGCGTTTATCGATGCTGAACACGCGCTGGACCCAATCTACGCACGTAAACTGGGCGTCGATATCGACAACCTGCTGTGCTCCCAGCCGGACACCGGCGAGCAGGCACTGGAAATCTGTGACGCCCTGGCGCGTTCTGGCGCAGTAGACGTTATCGTCGTTGACTCCGTGGCGGCACTGACGCCGAAAGCGGAAATCGAAGGCGAAATCGGCGACTCTCATATGGGCCTTGCGGCACGTATGATGAGCCAGGCGATGCGTAAGCTGGCGGGTAACCTGAAGCAGTCCAACACGCTGCTGATCTTCATCAACCAGATCCGTATGAAAATTGGTGTGATGTTCGGCAACCCGGAAACCACTACCGGTGGTAACGCGCTGAAATTCTACGCCTCTGTTCGTCTCGACATCCGTCGTATCGGCGCGGTGAAAGAGGGCGAAAACGTGGTGGGTAGCGAAACCCGCGTGAAAGTGGTGAAGAACAAAATCGCTGCGCCGTTTAAACAGGCTGAATTCCAGATCCTCTACGGCGAAGGTATCAACTTCTACGGCGAACTGGTTGACCTGGGCGTAAAAGAGAAGCTGATCGAGAAAGCAGGCGCGTGGTACAGCTACAAAGGTGAGAAGATCGGTCAGGGTAAAGCGAATGCGACTGCCTGGCTGAAAGATAACCCGGAAACCGCGAAAGAGATCGAGAAGAAAGTACGTGAGTTGCTGCTGAGCAACCCGAACTCAACGCCGGATTTCTCTGTAGATGATAGCGAAGGCGTAGCAGAAACTAACGAAGATTTTTAAtcgtcttgtttgatacacaagggtcgcatctgcggcccttttgcttttttaagttgtaaggatatgccattctagaNNNNNNNNNNNNNNNNNNagatcggaagagcgtcgtgtagggaaagagtgt-3’
               E. coli recA with coding sequence in caps: 3’-ggccataatgggccgtactgtcctcattttTACCGATAGCTGCTTTTGTTTGTCTTTCGCAACCGCCGTCGTGACCCGGTCTAACTCTTTGTTAAACCATTTCCGAGGTAGTACGCGGACCCACTTCTGGCAAGGTACCTACACCTTTGGTAGAGATGGCCAAGCGAAAGTGACCTATAGCGCGAACCCCGTCCACCAGACGGCTACCCGGCATAGCAGCTTTAGATGCCTGGCCTTAGAAGGCCATTTTGGTGCGACTGCGACGTCCACTAGCGGCGTCGCGTCGCACTTCCATTTTGGACACGCAAATAGCTACGACTTGTGCGCGACCTGGGTTAGATGCGTGCATTTGACCCGCAGCTATAGCTGTTGGACGACACGAGGGTCGGCCTGTGGCCGCTCGTCCGTGACCTTTAGACACTGCGGGACCGCGCAAGACCGCGTCATCTGCAATAGCAGCAACTGAGGCACCGCCGTGACTGCGGCTTTCGCCTTTAGCTTCCGCTTTAGCCGCTGAGAGTATACCCGGAACGCCGTGCATACTACTCGGTCCGCTACGCATTCGACCGCCCATTGGACTTCGTCAGGTTGTGCGACGACTAGAAGTAGTTGGTCTAGGCATACTTTTAACCACACTACAAGCCGTTGGGCCTTTGGTGATGGCCACCATTGCGCGACTTTAAGATGCGGAGACAAGCAGAGCTGTAGGCAGCATAGCCGCGCCACTTTCTCCCGCTTTTGCACCACCCATCGCTTTGGGCGCACTTTCACCACTTCTTGTTTTAGCGACGCGGCAAATTTGTCCGACTTAAGGTCTAGGAGATGCCGCTTCCATAGTTGAAGATGCCGCTTGACCAACTGGACCCGCATTTTCTCTTCGACTAGCTCTTTCGTCCGCGCACCATGTCGATGTTTCCACTCTTCTAGCCAGTCCCATTTCGCTTACGCTGACGGACCGACTTTCTATTGGGCCTTTGGCGCTTTCTCTAGCTCTTCTTTCATGCACTCAACGACGACTCGTTGGGCTTGAGTTGCGGCCTAAAGAGACATCTACTATCGCTTCCGCATCGTCTTTGATTGCTTCTAAAAATTagcagaacaaactatgtgttcccagcgtagacgccgggaaaacgaaaaaattcaacattcctatacggtaagatctNNNNNNNNNNNNNNNNNNtctagccttctcgcagcacatccctttctcaca-5’
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     Rnd1R: 3’-TCTAGCCTTCTCGCAGCACATCCCTTTCTCACA-5’
</pre></div>
</div>
<p>After these subamplicons are generated in the Round 1 PCR, a second round of PCR is used to add the remaining Illumina adapter sequences and TruSeq index sequences for multiplexing. For example, here are two primers used in the Round 2 PCR reactions:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>&gt;Rnd2IndexF: forward primer for the Round 2 PCR that adds the Illumina adaptor (uppercase) and a TruSeq index, shown here as 6 “n” nucleotides
5’-CAAGCAGAAGACGGCATACGAGATnnnnnngtgactggagttcagacgtgtgctcttcc-3’

&gt;Rnd2R: reverse primer for the Round 2 PCR that adds the Illumina adaptor (uppercase)
5’-AATGATACGGCGACCACCGAGATCTacactctttccctacacgacgctcttccgatct-3’
</pre></div>
</div>
<p>The products of the Round 2 PCR are sequenced with asymmetric paired-end reads with Read1 sequencing the barcode (<span class="math">\(\ge 18\)</span> nucleotides), and with Read2 of sufficient length to overlap with the next subamplicon (<span class="math">\(\ge 200\)</span> nucleotides). The sequencing data are analyzed with the standard Illumina pipeline to generate FASTQ files containing the R1 and R2 reads.</p>
</div>
<div class="section" id="subassembly-algorithm">
<h2><a class="toc-backref" href="#id5">Subassembly algorithm</a><a class="headerlink" href="#subassembly-algorithm" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">dms_subassemble</span></code> implements the following algorithm:</p>
<ol class="arabic">
<li><p class="first">The paired <em>R1</em> and <em>R2</em> reads are read from <code class="docutils literal"><span class="pre">r1files</span></code> and <code class="docutils literal"><span class="pre">r2files</span></code>, and any nucleotide that has a Q score &lt; <code class="docutils literal"><span class="pre">--minq</span></code> is converted to an <code class="docutils literal"><span class="pre">N</span></code> (e.g. considered ambiguous).</p>
</li>
<li><p class="first">The <em>R1</em> read is only processed to extract the barcode (the first <code class="docutils literal"><span class="pre">--barcodelength</span></code> nucleotides) and any remaining nucleotides in <em>R1</em> are discarded. The read pair is then discarded if:</p>
<blockquote>
<div><ul class="simple">
<li>There are any <code class="docutils literal"><span class="pre">N</span></code> nucleotides in the barcode.</li>
<li>The fraction of <code class="docutils literal"><span class="pre">N</span></code> nucleotides in the trimmed <em>R2</em> read is &gt; <code class="docutils literal"><span class="pre">--maxlowqfrac</span></code>.</li>
</ul>
</div></blockquote>
</li>
<li><p class="first">An attempt is made to gaplessly align the trimmed <em>R2</em> read at each site specified by <code class="docutils literal"><span class="pre">alignspecs</span></code>. Briefly, <code class="docutils literal"><span class="pre">alignspecs</span></code> specifies pairs of numbers <em>REFSEQSTART</em> and <em>R2START</em>. For <em>R2</em> read, we try to align the read starting at nucleotide <em>R2START</em> (1, 2, … numbering) at nucleotide <em>REFSEQSTART</em> in <code class="docutils literal"><span class="pre">refseq</span></code>. If <code class="docutils literal"><span class="pre">--trimR2</span></code> has its default value of <em>auto</em>, we only try to align up to where the next subamplicon would be (effectively trimming the unneeded part of <em>R2</em> from its 3’ end). If the read gaplessly aligns with no more than <code class="docutils literal"><span class="pre">--maxmuts</span></code> mutations of character type <code class="docutils literal"><span class="pre">--chartype</span></code>, the alignment is considered successful. Read that fail to align are written <code class="docutils literal"><span class="pre">&lt;outprefix&gt;_unaligned.txt</span></code> depending on the value of <code class="docutils literal"><span class="pre">--no_write_unaligned</span></code> and then discarded. Otherwise, reads are retained.</p>
</li>
<li><p class="first">After collecting all alignable trimmed <em>R2</em> reads for a barcode, we then see if we have enough coverage to subassemble the gene. We call identities in the gene only if the character (which may be a codon character rather than a nucleotide character, see <code class="docutils literal"><span class="pre">--chartype</span></code>) has at least <code class="docutils literal"><span class="pre">--minreadspersite</span></code> non-ambiguous (not <code class="docutils literal"><span class="pre">N</span></code>) reads covering it and <span class="math">\(\ge\)</span> <code class="docutils literal"><span class="pre">--minreadconcurrence</span></code> of this reads concur on its identity.</p>
</li>
</ol>
</div>
<div class="section" id="command-line-usage">
<h2><a class="toc-backref" href="#id6">Command-line usage</a><a class="headerlink" href="#command-line-usage" title="Permalink to this headline">¶</a></h2>
<p>Subassemble barcoded variants with linkage from short-read sequences. R1 has the barcode, R2 has the sequence fragment. This script is part of dms_tools (version 1.1.20) written by the Bloom Lab (see <a class="reference external" href="https://github.com/jbloomlab/dms_tools/graphs/contributors">https://github.com/jbloomlab/dms_tools/graphs/contributors</a> for all contributors). Detailed documentation is at <a class="reference external" href="http://jbloomlab.github.io/dms_tools/">http://jbloomlab.github.io/dms_tools/</a></p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">dms_subassemble</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">barcodelength</span> <span class="n">BARCODELENGTH</span><span class="p">]</span>
                       <span class="p">[</span><span class="o">--</span><span class="n">trimR2</span> <span class="p">{</span><span class="n">auto</span><span class="p">,</span><span class="n">none</span><span class="p">}]</span> <span class="p">[</span><span class="o">--</span><span class="n">minq</span> <span class="n">MINQ</span><span class="p">]</span>
                       <span class="p">[</span><span class="o">--</span><span class="n">maxlowqfrac</span> <span class="n">MAXLOWQFRAC</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">maxmuts</span> <span class="n">MAXMUTS</span><span class="p">]</span>
                       <span class="p">[</span><span class="o">--</span><span class="n">minreadspersite</span> <span class="n">MINREADSPERSITE</span><span class="p">]</span>
                       <span class="p">[</span><span class="o">--</span><span class="n">minreadconcurrence</span> <span class="n">MINREADCONCURRENCE</span><span class="p">]</span>
                       <span class="p">[</span><span class="o">--</span><span class="n">chartype</span> <span class="p">{</span><span class="n">codon</span><span class="p">}]</span> <span class="p">[</span><span class="o">--</span><span class="n">no_write_barcode_reads</span><span class="p">]</span>
                       <span class="p">[</span><span class="o">--</span><span class="n">purgefrac</span> <span class="n">PURGEFRAC</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">v</span><span class="p">]</span>
                       <span class="n">outprefix</span> <span class="n">refseq</span> <span class="n">r1files</span> <span class="n">r2files</span> <span class="n">alignspecs</span>
                       <span class="p">[</span><span class="n">alignspecs</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
<div class="section" id="Positional Arguments">
<h3><a class="toc-backref" href="#id7">Positional Arguments</a><a class="headerlink" href="#Positional Arguments" title="Permalink to this headline">¶</a></h3>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd>outprefix</kbd></td>
<td><p class="first">Prefix for output files.</p>
<p class="last">See Output files for a list of the created files.</p>
</td></tr>
<tr><td class="option-group">
<kbd>refseq</kbd></td>
<td><p class="first">Existing FASTA file containing wildtype gene we are subassembling.</p>
<p class="last">This file should specify a valid in-frame coding sequence.</p>
</td></tr>
<tr><td class="option-group">
<kbd>r1files</kbd></td>
<td>Comma-separated list of R1 FASTQ files (no spaces). Files can optionally be gzipped (extension .gz).</td></tr>
<tr><td class="option-group">
<kbd>r2files</kbd></td>
<td>Like ‘r1files’ but for R2. Must be same number of comma-separated entires as for ‘r1files’.</td></tr>
<tr><td class="option-group">
<kbd>alignspecs</kbd></td>
<td><p class="first">This argument is repeated to specify each possible alignment location for R2. Each specification is two comma-delimited integers (no spaces): ‘REFSEQSTART,R2START’. REFSEQSTART is nucleotide (1, 2, … numbering) in ‘refseq’ where nucleotide R2START in R2 aligns.</p>
<p>It is important to set alignspecs so that you don’t count the part of the subamplicon that is in the primer binding site, since the nucleotide identities in this region come from the primers rather than the templates being sequences. Typically, R2START would be one greater than the length of the gene-binding region of the primer to avoid this.</p>
<p class="last">The alignments will fail if you don’t set alignspecs exactly correctly, as the program only tries gapless alignment.</p>
</td></tr>
</tbody>
</table>
</div>
<div class="section" id="Named Arguments">
<h3><a class="toc-backref" href="#id8">Named Arguments</a><a class="headerlink" href="#Named Arguments" title="Permalink to this headline">¶</a></h3>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group" colspan="2">
<kbd>--barcodelength</kbd></td>
</tr>
<tr><td>&#160;</td><td><p class="first">Length of barcode (NNN…) which starts at beginning of R1.</p>
<p class="last">Default: 18</p>
</td></tr>
<tr><td class="option-group">
<kbd>--trimR2</kbd></td>
<td><p class="first">Possible choices: auto, none</p>
<p>Trim R2 read? If ‘auto’, trim until start of next subamplicon specified by ‘alignspecs’; if ‘none’ no trimming.</p>
<p class="last">Default: “auto”</p>
</td></tr>
<tr><td class="option-group">
<kbd>--minq</kbd></td>
<td><p class="first">Nucleotides with Q scores &lt; this number are converted to N.</p>
<p class="last">Default: 20</p>
</td></tr>
<tr><td class="option-group">
<kbd>--maxlowqfrac</kbd></td>
<td><p class="first">Only retain if fraction of N nucleotides in R2 read &lt;= this.</p>
<p class="last">Default: 0.15</p>
</td></tr>
<tr><td class="option-group">
<kbd>--maxmuts</kbd></td>
<td><p class="first">Only align read if &lt;= this many mismatches with “refseq” counted in terms of “chartype”.</p>
<p class="last">Default: 4</p>
</td></tr>
<tr><td class="option-group" colspan="2">
<kbd>--minreadspersite</kbd></td>
</tr>
<tr><td>&#160;</td><td><p class="first">Call site only barcodes when &gt;= this many reads give it a non-ambiguous identity.</p>
<p class="last">Default: 2</p>
</td></tr>
<tr><td class="option-group" colspan="2">
<kbd>--minreadconcurrence</kbd></td>
</tr>
<tr><td>&#160;</td><td><p class="first">Only call sites when &gt;= this fraction of reads concur.</p>
<p class="last">Default: 0.75</p>
</td></tr>
<tr><td class="option-group">
<kbd>--chartype</kbd></td>
<td><p class="first">Possible choices: codon</p>
<p>Character for which we are counting mutations. Currently “codon” is only allowed value (in the future “nucleotide” might be added).</p>
<p class="last">Default: “codon”</p>
</td></tr>
<tr><td class="option-group" colspan="2">
<kbd>--no_write_barcode_reads</kbd></td>
</tr>
<tr><td>&#160;</td><td><p class="first">Don’t write all barcodes and assigned reads to a file (saves space/time to not do this).</p>
<p>Default: False</p>
<p class="last">If you set this option, then the program does not create the All reads by barcode file. If you are not debugging, you may not want this file as it is very large.</p>
</td></tr>
<tr><td class="option-group">
<kbd>--purgefrac</kbd></td>
<td><p class="first">Randomly purge reads with this probability (subsample the data).</p>
<p class="last">Default: 0</p>
</td></tr>
<tr><td class="option-group">
<kbd>-v, --version</kbd></td>
<td>show program’s version number and exit</td></tr>
</tbody>
</table>
</div>
</div>
<div class="section" id="output-files">
<h2><a class="toc-backref" href="#id9">Output files</a><a class="headerlink" href="#output-files" title="Permalink to this headline">¶</a></h2>
<p>The following output files are created. Each has the prefix specified by <code class="docutils literal"><span class="pre">outprefix</span></code> with the following suffixes.</p>
<div class="section" id="log-file">
<h3><a class="toc-backref" href="#id10">Log file</a><a class="headerlink" href="#log-file" title="Permalink to this headline">¶</a></h3>
<p>This file has the suffix <code class="docutils literal"><span class="pre">.log</span></code> and tracks the progress of the program.</p>
</div>
<div class="section" id="subassembled-variants-file">
<h3><a class="toc-backref" href="#id11">Subassembled variants file</a><a class="headerlink" href="#subassembled-variants-file" title="Permalink to this headline">¶</a></h3>
<p>This file has the suffix <code class="docutils literal"><span class="pre">_subassembled_variants.txt</span></code>. It is the primary output file, and lists all barcodes that can be subassembled according to the parameters passed to <code class="docutils literal"><span class="pre">dms_subassemble</span></code>.</p>
<p>It is in the <a class="reference internal" href="fileformats.html#subassembly-file"><span class="std std-ref">Subassembly file</span></a> format. Note that for codon sequences (<code class="docutils literal"><span class="pre">--chartype</span></code> of <code class="docutils literal"><span class="pre">codon</span></code>), the mutations are numbered according to the <strong>codon</strong> position in 1, 2, … numbering, not the nucleotide position.</p>
<p>Here are a few example lines:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">TATTACATCTGCCCCCAA</span> <span class="n">ATGGCTATCGACGAAAACAAACAGAAAGCGTTGGCGGCAGCACTGGGCCAGATTGAGAAACAATTTGGTAAAGGCTCCATCATGCGCCTGGGTGAAGACCGTTCCATGGATGTGGAAACCATCTCTACCGGTTCGCTTTCACTGGATATCGCGCTTGGGGCAGGTGGTCTGCCGATGGGCCGTATCGTCGAAATCTACGGACCGGAATCTTCCGGTAAAACCACGCTGACGCTGCAGGTGATCGCCGCAGCGCAGCGTGAAGGTAAAACCTGTGCGTTTATCGATGCTGAACACGCGCTGGACCCAATCTACGCACGTAAACTGGCAGTCGATATCGACAACCTGCTGTGCTCCCAGCCGGACACCGGCGAGCAGGCACTGGAAATCTGTGACGCCCTGGCGCGTTCTGGCGCAGTAGACGTTATCGTCGTTGACTCCGTGGCGGCACTGACGCCGAAAGCGGAAATCGAAGGCGAAATCGGCGACTCTCATATGGGCCTTGCGGCACGTATGATGAGCCAGGCGATGCGTAAGCTGGCGGGTAACCTGAAGCAGTCCAACACGCTGCTGATCTTCATCAACCAGATCCGTATGAAAATTGGTGTGATGTTCGGCAACCCGGAAACCACTACCGGTGGTAACGCGCTGAAATTCTACGCCTCTGTTCGTCTCGACATCCGTCGTATCGGCGCGGTGAAAGAGGGCGAAAACGTGGTGGGTAGCGAAACCCGCGTGAAAGTGGTGAAGAACAAAATCGCTGCGCCGTTTAAACAGGCTGAATTCCAGATCCTCTACGGCGAAGGTATCAACTTCTACGGCGAACTGGTTGACCTGGGCGTAAAAGAGAAGCTGATCGAGAAAGCAGGCGCGTGGTACAGCTACAAAGGTGAGAAGATCGGTCAGGGTAAAGCGAATGCGACTGCCTGGCTGAAAGATAACCCGGAAACCGCGAAAGAGATCGAGAAGAAAGTACGTGAGTTGCTGCTGAGCAACCCGAACTCAACGCCGGATTTCTCTGTAGATGATAGCGAAGGCGTAGCAGAAACTAACGAAGATTTTTAA</span> <span class="n">GGC109GCA</span>
<span class="n">AACTCTGTGTTCCCATCA</span> <span class="n">ATGGCTATCGACGAAAACAAACAGAAAGCGTTGGCGGCAGCACTGGGCCAGATTGAGAAACAATTTGGTAAAGGCTCCATCATGCGCCTGGGTGAAGACCGTTCCATGGATGTGGAAACCATCTCTACCGGTTCGCTTTCACTGGATATCGCGCTTGGGGCAGGTGGTCTGCCGATGGGCCGTATCGTCGAAATCTACGGACCGGAATCTTCCGGTAAAACCACGCTGACGCTGCAGGTGATCGCCGCAGCGCAGCGTGAAGGTAAAACCTGTGCGTTTATCGATGCTGAACACGCGCTGGACCCAATCTACGCACGTAAACTGGGCGTCGATATCGACAACCTGCTGTGCTCCCAGCCGGACACCGGCGAGCAGGCACTGGAAATCTGTGACGCCCTGGCGCGTTCTGGCGCAGTAGACGTTATCGTCGTTGACTCCGTGGCGGCACTGACGCCGAAAGCGGAAATCGAAGGCGAAATCGGCGACTCTCATATGGGCCTTGCGGCACGTATGATGAGCCAGGCGATGCGTAAGCTGGCGGGTAACCTGAAGCAGTCCAACACGCTGCTGATCTTCATCAACCAGATCCGTATGAAAATTGGTGTGATGTTCGGCAACCCGGAAACCACTACCGGTGGTAACGCGCTGAAATTCTACGCCTCTGTTCGTCTCGACATCCGTCGTATCGGCGCGGTGAAAGAGGGCGAAAACGTGGTGGGTAGCGAAACCCGCGTGAAAGTGGTGAAGAACAAAATCGCTGCGCCGTTTAAACAGGCTGAATTCCAGATCCTCTACGGCGAAGGTATCAACTTCTACGGCGAACTGGTTGACCTGGGCGTAAAAGAGAAGCTGATCGAGAAAGCAGGCGCGTGGTACAGCTACAAAGGTGAGAAGATCGGTCAGGGTAAAGCGAATGCGACTGCCTGGCTGAAAGATAACCCGGAAACCGCGAAAGAGATCGAGAAGAAAGTACGTGAGTTGCTGCTGAGCAACCCGAACTCAACGCCGGATTTCTCTGTAGATGATAGCGAAGGCGTAGCAGAAACTAACGAAGATTTTTAA</span> <span class="n">no_mutations</span>
<span class="n">GTTAACCGATCAACGCAA</span> <span class="n">ATGGCTATCGACGAAAACAAACAGAAAGCGTTGGCGGCAGCACTGGGCCAGATTGAGAAACAATTTGGTAAAGGCTCCATCATGCGCCTGGGTGAAGACCGTTCCATGGATGTGGAAACCATCTCTACCGGTTCGCTTGCACTGGATATCGCGCTTGGGGCAGGTGGTCTGCCGATGGGCCGTATCGTCGAAATCTACGGACCGGAATCTTCCGGTAAAACCACGCTGACGCTGCAGGTGATCGCCGCAGCGCAGCGTGAAGGTAAAACCTGTGCGTTTATCGATGCTGAACACGCGCTGGACCCAATCTACGCACGTAAACTGGGCGTCGATATCGACAACCTGCTGTGCTCCCAGCCGGACACCGGCGAGCAGGCACTGGAAATCTGTGACGCCCTGGCGCGTTCTGGCGCAGTAGACGTTATCGTCGTTGACTCCGTGGCGGCACTGACGCCGAAAGCGGAAATCGAAGGCGAAATCGGCGACTCTCATATGGGCCTTGCGGCACGTATGATGAGCCAGGCGATGCGTAAGCTGGCGGGTAACCTGAAGCAGTCCAACACGCTGCTGATCTTCATCAACCAGATCCGTATGAAAATTGGTGTGATGTTCGGCAACCCGGAAACCACTACCGGTGGTAACGCGCTGAAATTCTACGCCTCTGTTCGTCTCGACATCCGTCGTATCGGGGCGGTGAAAGAGGGCGAAAACGTGGTGGGTAGCGAAACCCGCGTGAAAGTGGTGAAGAACAAAATCGCTGCGCCGTTTAAACAGGCTGAATTCCAGATCCTCTACGGCGAAGGTATCAACTTCTACGGCGAACTGGTTGACCTGGGCGTAAAAGAGAAGCTGATCGAGAAAGCAGGCGCGTGGTACAGCTACAAAGGTGAGAAGATCGGTCAGGGTAAAGCGAATGCGACTGCCTGGCTGAAAGATAACCCGGAAACCGCGAAAGAGATCGAGAAGAAAGTACGTGAGTTGCTGCTGAGCAACCCGAACTCAACGCCGGATTTCTCTGTAGATGATAGCGAAGGCGTAGCAGAAACTAACGAAGATTTTTAA</span> <span class="n">TCA47GCA</span><span class="p">,</span><span class="n">GGC230GGG</span>
</pre></div>
</div>
<p>Each line lists a barcode, then the sequence subassembled for that barcode, and finally any mutations relative to <code class="docutils literal"><span class="pre">refseq</span></code>.</p>
</div>
<div class="section" id="all-reads-by-barcode-file">
<h3><a class="toc-backref" href="#id12">All reads by barcode file</a><a class="headerlink" href="#all-reads-by-barcode-file" title="Permalink to this headline">¶</a></h3>
<p>This file has the suffix <code class="docutils literal"><span class="pre">_all_reads_by_barcode.txt</span></code>. It is a <strong>very large</strong> text file that lists each read that matches each barcode (both those successfully subassembled and those that aren’t). It also explains why or why not a barcode was subassembled.</p>
<p>This file is <strong>not</strong> created if you set the <code class="docutils literal"><span class="pre">--no_write_barcode_reads</span></code>.</p>
</div>
<div class="section" id="summary-statistics-file">
<h3><a class="toc-backref" href="#id13">Summary statistics file</a><a class="headerlink" href="#summary-statistics-file" title="Permalink to this headline">¶</a></h3>
<p>This file has the suffix <code class="docutils literal"><span class="pre">_summarystats.txt</span></code>. It lists summary statistics about the subassembly. Here is an example:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">barcodes</span> <span class="p">(</span><span class="n">total</span><span class="p">)</span> <span class="o">=</span> <span class="mi">642641</span>
<span class="n">barcodes</span> <span class="n">successfully</span> <span class="n">subassembled</span> <span class="o">=</span> <span class="mi">98281</span>
<span class="n">barcodes</span> <span class="k">with</span> <span class="n">at</span> <span class="n">least</span> <span class="n">one</span> <span class="n">alignable</span> <span class="n">read</span> <span class="o">=</span> <span class="mi">615204</span>
<span class="n">read</span> <span class="n">pairs</span> <span class="p">(</span><span class="n">total</span><span class="p">)</span> <span class="o">=</span> <span class="mi">29611710</span>
<span class="n">read</span> <span class="n">pairs</span> <span class="n">aligned</span> <span class="n">at</span> <span class="n">site</span> <span class="mi">1</span> <span class="o">=</span> <span class="mi">2442839</span>
<span class="n">read</span> <span class="n">pairs</span> <span class="n">aligned</span> <span class="n">at</span> <span class="n">site</span> <span class="mi">172</span> <span class="o">=</span> <span class="mi">1944582</span>
<span class="n">read</span> <span class="n">pairs</span> <span class="n">aligned</span> <span class="n">at</span> <span class="n">site</span> <span class="mi">354</span> <span class="o">=</span> <span class="mi">2794256</span>
<span class="n">read</span> <span class="n">pairs</span> <span class="n">aligned</span> <span class="n">at</span> <span class="n">site</span> <span class="mi">481</span> <span class="o">=</span> <span class="mi">2352744</span>
<span class="n">read</span> <span class="n">pairs</span> <span class="n">aligned</span> <span class="n">at</span> <span class="n">site</span> <span class="mi">633</span> <span class="o">=</span> <span class="mi">2982037</span>
<span class="n">read</span> <span class="n">pairs</span> <span class="n">aligned</span> <span class="n">at</span> <span class="n">site</span> <span class="mi">809</span> <span class="o">=</span> <span class="mi">3024340</span>
<span class="n">read</span> <span class="n">pairs</span> <span class="n">aligned</span> <span class="n">at</span> <span class="n">site</span> <span class="mi">941</span> <span class="o">=</span> <span class="mi">2758838</span>
<span class="n">read</span> <span class="n">pairs</span> <span class="n">purged</span> <span class="n">due</span> <span class="n">to</span> <span class="n">low</span> <span class="n">quality</span> <span class="o">=</span> <span class="mi">8965508</span>
<span class="n">read</span> <span class="n">pairs</span> <span class="n">that</span> <span class="n">are</span> <span class="n">alignable</span> <span class="o">=</span> <span class="mi">18299636</span>
<span class="n">read</span> <span class="n">pairs</span> <span class="n">that</span> <span class="n">are</span> <span class="n">alignable</span> <span class="ow">and</span> <span class="nb">map</span> <span class="n">to</span> <span class="n">a</span> <span class="n">subassembled</span> <span class="n">barcode</span> <span class="o">=</span> <span class="mi">8068574</span>
<span class="n">read</span> <span class="n">pairs</span> <span class="n">that</span> <span class="n">are</span> <span class="n">unalignable</span> <span class="o">=</span> <span class="mi">2346566</span>
<span class="n">read</span> <span class="n">pairs</span> <span class="n">that</span> <span class="n">fail</span> <span class="n">Illumina</span> <span class="nb">filter</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">sites</span> <span class="k">with</span> <span class="n">insufficient</span> <span class="n">concurrence</span> <span class="n">due</span> <span class="n">to</span> <span class="n">mismatch</span> <span class="n">between</span> <span class="n">mutant</span> <span class="ow">and</span> <span class="n">wildtype</span> <span class="n">characters</span> <span class="o">=</span> <span class="mi">52936</span>
<span class="n">sites</span> <span class="k">with</span> <span class="n">insufficient</span> <span class="n">concurrence</span> <span class="n">due</span> <span class="n">to</span> <span class="n">mismatch</span> <span class="n">between</span> <span class="n">two</span> <span class="n">mutant</span> <span class="n">characters</span> <span class="o">=</span> <span class="mi">68862</span>
</pre></div>
</div>
</div>
<div class="section" id="alignable-reads-per-barcode-file">
<h3><a class="toc-backref" href="#id14">Alignable reads per barcode file</a><a class="headerlink" href="#alignable-reads-per-barcode-file" title="Permalink to this headline">¶</a></h3>
<p>This file has the suffix <code class="docutils literal"><span class="pre">_alignablereadsperbarcode.txt</span></code>. It gives the distribution of the number of alignable reads per barcode. Here is an example of the first few lines:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">nreads</span>  <span class="n">nbarcodes</span>
<span class="mi">0</span>   <span class="mi">27437</span>
<span class="mi">1</span>   <span class="mi">153500</span>
<span class="mi">2</span>   <span class="mi">10580</span>
<span class="mi">3</span>   <span class="mi">8714</span>
<span class="mi">4</span>   <span class="mi">8670</span>
<span class="mi">5</span>   <span class="mi">8825</span>
<span class="mi">6</span>   <span class="mi">8991</span>
<span class="mi">7</span>   <span class="mi">8958</span>
</pre></div>
</div>
</div>
<div class="section" id="mutations-among-subassembled-variants-file">
<h3><a class="toc-backref" href="#id15">Mutations among subassembled variants file</a><a class="headerlink" href="#mutations-among-subassembled-variants-file" title="Permalink to this headline">¶</a></h3>
<p>This file has the suffix <code class="docutils literal"><span class="pre">_nmuts_among_subassembled.txt</span></code>. It gives the distribution of the number of mutations per variant among subassembled variants. Here is an example:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">nmuts</span>   <span class="n">nvariants</span>
<span class="mi">0</span>   <span class="mi">35614</span>
<span class="mi">1</span>   <span class="mi">32442</span>
<span class="mi">2</span>   <span class="mi">17902</span>
<span class="mi">3</span>   <span class="mi">7705</span>
<span class="mi">4</span>   <span class="mi">3046</span>
<span class="mi">5</span>   <span class="mi">1113</span>
<span class="mi">6</span>   <span class="mi">316</span>
<span class="mi">7</span>   <span class="mi">107</span>
<span class="mi">8</span>   <span class="mi">33</span>
<span class="mi">9</span>   <span class="mi">2</span>
<span class="mi">10</span>  <span class="mi">1</span>
</pre></div>
</div>
</div>
<div class="section" id="read-start-sites-file">
<h3><a class="toc-backref" href="#id16">Read start sites file</a><a class="headerlink" href="#read-start-sites-file" title="Permalink to this headline">¶</a></h3>
<p>This file has the suffix <code class="docutils literal"><span class="pre">_refseqstarts.txt</span></code>. It gives the number of reads that start at each of the positions in <code class="docutils literal"><span class="pre">refseq</span></code> specified in <code class="docutils literal"><span class="pre">alignspecs</span></code>. Here is an example:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">refseqstart</span> <span class="n">nreads</span>
<span class="mi">1</span>   <span class="mi">2442839</span>
<span class="mi">172</span> <span class="mi">1944582</span>
<span class="mi">354</span> <span class="mi">2794256</span>
<span class="mi">481</span> <span class="mi">2352744</span>
<span class="mi">633</span> <span class="mi">2982037</span>
<span class="mi">809</span> <span class="mi">3024340</span>
<span class="mi">941</span> <span class="mi">2758838</span>
</pre></div>
</div>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#"><code class="docutils literal"><span class="pre">dms_subassemble</span></code></a><ul>
<li><a class="reference internal" href="#overview">Overview</a></li>
<li><a class="reference internal" href="#subassembly-sequencing">Subassembly sequencing</a></li>
<li><a class="reference internal" href="#subassembly-algorithm">Subassembly algorithm</a></li>
<li><a class="reference internal" href="#command-line-usage">Command-line usage</a><ul>
<li><a class="reference internal" href="#Positional Arguments">Positional Arguments</a></li>
<li><a class="reference internal" href="#Named Arguments">Named Arguments</a></li>
</ul>
</li>
<li><a class="reference internal" href="#output-files">Output files</a><ul>
<li><a class="reference internal" href="#log-file">Log file</a></li>
<li><a class="reference internal" href="#subassembled-variants-file">Subassembled variants file</a></li>
<li><a class="reference internal" href="#all-reads-by-barcode-file">All reads by barcode file</a></li>
<li><a class="reference internal" href="#summary-statistics-file">Summary statistics file</a></li>
<li><a class="reference internal" href="#alignable-reads-per-barcode-file">Alignable reads per barcode file</a></li>
<li><a class="reference internal" href="#mutations-among-subassembled-variants-file">Mutations among subassembled variants file</a></li>
<li><a class="reference internal" href="#read-start-sites-file">Read start sites file</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="dms_barcodedsubamplicons.html"
                        title="previous chapter"><code class="docutils literal"><span class="pre">dms_barcodedsubamplicons</span></code></a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="dms_matchsubassembledbarcodes.html"
                        title="next chapter"><code class="docutils literal"><span class="pre">dms_matchsubassembledbarcodes</span></code></a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/dms_subassemble.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="dms_matchsubassembledbarcodes.html" title="dms_matchsubassembledbarcodes"
             >next</a> |</li>
        <li class="right" >
          <a href="dms_barcodedsubamplicons.html" title="dms_barcodedsubamplicons"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">dms_tools 1.2.2 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="programs.html" >Programs</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2014, Jesse Bloom.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.6.3.
    </div>
  </body>
</html>