<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>dms_inferdiffprefs &#8212; dms_tools 1.1.20 documentation</title>
    
    <link rel="stylesheet" href="_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.1.20',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="dms_tools 1.1.20 documentation" href="index.html" />
    <link rel="up" title="Programs" href="programs.html" />
    <link rel="next" title="dms_logoplot" href="dms_logoplot.html" />
    <link rel="prev" title="dms_diffselection" href="dms_diffselection.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="dms_logoplot.html" title="dms_logoplot"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="dms_diffselection.html" title="dms_diffselection"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">dms_tools 1.1.20 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="programs.html" accesskey="U">Programs</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="dms-inferdiffprefs">
<span id="id1"></span><h1><a class="toc-backref" href="#id2"><code class="docutils literal"><span class="pre">dms_inferdiffprefs</span></code></a><a class="headerlink" href="#dms-inferdiffprefs" title="Permalink to this headline">¶</a></h1>
<div class="contents topic" id="contents">
<p class="topic-title first">Contents</p>
<ul class="simple">
<li><a class="reference internal" href="#dms-inferdiffprefs" id="id2"><code class="docutils literal"><span class="pre">dms_inferdiffprefs</span></code></a><ul>
<li><a class="reference internal" href="#overview" id="id3">Overview</a></li>
<li><a class="reference internal" href="#should-you-use-this-program" id="id4">Should you use this program?</a></li>
<li><a class="reference internal" href="#command-line-usage" id="id5">Command-line usage</a></li>
<li><a class="reference internal" href="#output" id="id6">Output</a></li>
<li><a class="reference internal" href="#do-not-worry-about-informational-messages" id="id7">Do not worry about Informational Messages</a></li>
<li><a class="reference internal" href="#runtime-and-convergence" id="id8">Runtime and convergence</a></li>
<li><a class="reference internal" href="#examples" id="id9">Examples</a></li>
</ul>
</li>
</ul>
</div>
<div class="section" id="overview">
<h2><a class="toc-backref" href="#id3">Overview</a><a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">dms_inferdiffprefs</span></code> is a program included with the <a class="reference external" href="http://jbloomlab.github.io/dms_tools">dms_tools</a> package. It infers the differential preference <span class="math">\(\Delta\pi_{r,x}\)</span> for each character <span class="math">\(x\)</span> (an amino acid, codon, or nucleotide) at each site <span class="math">\(r\)</span> in an alternative selection versus a control selection. Values of <span class="math">\(\Delta\pi_{r,x} &gt; 0\)</span> indicate that <span class="math">\(x\)</span> is favored in the alternative selection versus the control selection at site <span class="math">\(r\)</span>.</p>
<p>The inference is done by <a class="reference external" href="http://en.wikipedia.org/wiki/Markov_chain_Monte_Carlo">MCMC</a> using <a class="reference external" href="http://pystan.readthedocs.org/en/latest/">PyStan</a>.</p>
<p>For a detailed description of the algorithm implemented by this program, see <a class="reference internal" href="inferdiffprefs_algorithm.html#inferdiffprefs-algorithm"><span class="std std-ref">Algorithm to infer differential preferences</span></a>.</p>
<p>After you install <a class="reference external" href="http://jbloomlab.github.io/dms_tools">dms_tools</a>, this program will be available to run at the command line.</p>
</div>
<div class="section" id="should-you-use-this-program">
<h2><a class="toc-backref" href="#id4">Should you use this program?</a><a class="headerlink" href="#should-you-use-this-program" title="Permalink to this headline">¶</a></h2>
<p>Note that we are no longer certain if the method implemented in this program is superior to the much simpler <a class="reference internal" href="dms_diffselection.html#dms-diffselection"><span class="std std-ref">dms_diffselection</span></a>. So consider trying that program as well.</p>
</div>
<div class="section" id="command-line-usage">
<h2><a class="toc-backref" href="#id5">Command-line usage</a><a class="headerlink" href="#command-line-usage" title="Permalink to this headline">¶</a></h2>
<p>Infer differential preferences for amino acids, nucleotides, or codons. This script is part of dms_tools (version 1.1.19) written by the Bloom Lab (see <a class="reference external" href="https://github.com/jbloomlab/dms_tools/graphs/contributors">https://github.com/jbloomlab/dms_tools/graphs/contributors</a> for all contributors). Detailed documentation is at <a class="reference external" href="http://jbloomlab.github.io/dms_tools/">http://jbloomlab.github.io/dms_tools/</a></p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">dms_inferdiffprefs</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">chartype</span> <span class="p">{</span><span class="n">codon_to_aa</span><span class="p">,</span><span class="n">DNA</span><span class="p">,</span><span class="n">codon</span><span class="p">,</span><span class="n">aa</span><span class="p">}]</span>
                          <span class="p">[</span><span class="o">--</span><span class="n">err</span> <span class="n">ERR</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">excludestop</span><span class="p">]</span>
                          <span class="p">[</span><span class="o">--</span><span class="n">alpha_start</span> <span class="n">ALPHA_START</span><span class="p">]</span>
                          <span class="p">[</span><span class="o">--</span><span class="n">alpha_pis1</span> <span class="n">ALPHA_PIS1</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">alpha_err</span> <span class="n">ALPHA_ERR</span><span class="p">]</span>
                          <span class="p">[</span><span class="o">--</span><span class="n">alpha_deltapi</span> <span class="n">ALPHA_DELTAPI</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">logfile</span> <span class="n">LOGFILE</span><span class="p">]</span>
                          <span class="p">[</span><span class="o">--</span><span class="n">ncpus</span> <span class="n">NCPUS</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">sites</span> <span class="n">SITES</span> <span class="p">[</span><span class="n">SITES</span> <span class="o">...</span><span class="p">]]</span>
                          <span class="p">[</span><span class="o">--</span><span class="n">ratio_estimation</span> <span class="n">MINCOUNTS</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">seed</span> <span class="n">SEED</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">v</span><span class="p">]</span>
                          <span class="n">n_start</span> <span class="n">n_s1</span> <span class="n">n_s2</span> <span class="n">outfile</span>
</pre></div>
</div>
<dl class="docutils">
<dt>Positional arguments:</dt>
<dd><table class="first last docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd>n_start</kbd></td>
<td><p class="first">Existing file with counts from deep sequencing of starting library. For nucleotides, header line should be &#8220;# POSITION WT A C G T&#8221; and then subsequent lines give wildtype and counts for each site (i.e. &#8220;1 G 1013 23 19 47&#8221;); for codons use the 64 codons instead (i.e. &#8220;AAA AAC AAG ...&#8221;).</p>
<p>This file gives the <span class="math">\(n_{r,x}^{\textrm{start}}\)</span> values described in <a class="reference internal" href="inferdiffprefs_algorithm.html#inferdiffprefs-algorithm"><span class="std std-ref">Algorithm to infer differential preferences</span></a>.</p>
<p class="last">See <a class="reference internal" href="fileformats.html#dms-counts"><span class="std std-ref">Deep mutational scanning counts file</span></a> for full specification of the format for this file.</p>
</td></tr>
<tr><td class="option-group">
<kbd>n_s1</kbd></td>
<td><p class="first">Existing file with counts from deep sequencing library subject to selection 1. Typically this is the control selection; differential preferences are for changes in selection 2 relative to selection 1. File has same format as &#8220;n_start&#8221;.</p>
<p>This file gives the <span class="math">\(n_{r,x}^{s1}\)</span> values described in <a class="reference internal" href="inferdiffprefs_algorithm.html#inferdiffprefs-algorithm"><span class="std std-ref">Algorithm to infer differential preferences</span></a>. These are for selection <span class="math">\(s1\)</span>, which is taken as the &#8220;control&#8221; selection.</p>
<p class="last">See <a class="reference internal" href="fileformats.html#dms-counts"><span class="std std-ref">Deep mutational scanning counts file</span></a> for full specification of the format for this file.</p>
</td></tr>
<tr><td class="option-group">
<kbd>n_s2</kbd></td>
<td><p class="first">Existing file with counts from deep sequencing library subject to selection 2. Typically this is the alternate selection; differential preferences are for changes in selection 2 relative to selection 1. File has same format as &#8220;n_start&#8221;.</p>
<p>This file gives the <span class="math">\(n_{r,x}^{s2}\)</span> values described in <a class="reference internal" href="inferdiffprefs_algorithm.html#inferdiffprefs-algorithm"><span class="std std-ref">Algorithm to infer differential preferences</span></a>. These are for selection <span class="math">\(s2\)</span>, which is taken as the &#8220;alternative&#8221; selection to the &#8220;control&#8221; selection <span class="math">\(s1\)</span>.</p>
<p class="last">See <a class="reference internal" href="fileformats.html#dms-counts"><span class="std std-ref">Deep mutational scanning counts file</span></a> for full specification of the format for this file.</p>
</td></tr>
<tr><td class="option-group">
<kbd>outfile</kbd></td>
<td><p class="first">Created output file with differential preferences; overwritten if it already exists.</p>
<p>In addition to the inferred differential preferences <span class="math">\(\Delta\pi_{r,x}\)</span> for each character <span class="math">\(x\)</span> in selection <span class="math">\(s2\)</span> versus selection <span class="math">\(s1\)</span>, this file also gives the root-mean-square (RMS) differential preference for each site, and the posterior probabilities that these preferences are greater than or less than zero.</p>
<p>Note that the preferences are for <span class="math">\(s2\)</span> relative to <span class="math">\(s1\)</span>, so <span class="math">\(\Delta\pi_{r,x} &gt; 0\)</span> means that <span class="math">\(x\)</span> is preferred more at site <span class="math">\(r\)</span> in selection <span class="math">\(s2\)</span> than selection <span class="math">\(s1\)</span>.</p>
<p class="last">See <a class="reference internal" href="fileformats.html#diffpreferences-file"><span class="std std-ref">Differential preferences file</span></a> for full specification of the format for this file.</p>
</td></tr>
</tbody>
</table>
</dd>
<dt>Options:</dt>
<dd><table class="first last docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--chartype=codon_to_aa</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td><p class="first">Characters for which differential preferences are inferred: &#8220;codon_to_aa&#8221; = counts for codons and inferred differential preferences for amino acids; &#8220;DNA&#8221; = counts and inferred differential preference for DNA; &#8220;codon&#8221; = counts and inferred differential preferences for codons; &#8220;aa&#8221; = counts and inferred differential preferences for amino acids (possibly including stop codons, see &#8220;&#8211;excludestop&#8221;).</p>
<p>Note that the <strong>default</strong> value is &#8220;codon_to_aa&#8221;.</p>
<p>A value of &#8220;codon_to_aa&#8221; corresponds to using codon characters for the deep sequencing data, but inferring differential preferences for amino acids, as described in <a class="reference internal" href="inferprefs_algorithm.html#chartype-codon-to-aa"><span class="std std-ref">Characters are codons; preferences are for amino acids.</span></a> This would be the sensible approach if you assume that selection is operating on the amino-acid level. Note that the priors over the mutagenesis rates assume that all codon mutations are made at equal frequencies (<code class="docutils literal"><span class="pre">NNN</span></code> libraries).</p>
<p>A value of &#8220;DNA&#8221; corresponds to using DNA characters for both the deep sequencing data and inferred preferences, as described in <a class="reference internal" href="inferprefs_algorithm.html#chartype-dna"><span class="std std-ref">Characters are nucleotides; preferences are for nucleotides.</span></a></p>
<p>A value of &#8220;codon&#8221; corresponds to using codon characters for both the deep sequencing data and inferred preferences, as described in <a class="reference internal" href="inferprefs_algorithm.html#chartype-codon"><span class="std std-ref">Characters are codons; preferences are for codons.</span></a> You would prefer this option over &#8220;codon_to_aa&#8221; if you thought that there was different selection on different synonymous mutations. Note that the priors over the mutagenesis rates assume that all codon mutations are made at equal frequencies (<code class="docutils literal"><span class="pre">NNN</span></code> libraries).</p>
<p>A value of &#8220;aa&#8221; corresponds to using amino-acid characters for both the deep sequence data and inferred preferences,
as described in <a class="reference internal" href="inferprefs_algorithm.html#chartype-aa"><span class="std std-ref">Characters are amino acids; preferences are for amino acids.</span></a>. You should <strong>only</strong> use this option if your data make it impossible to use &#8220;codon_to_aa&#8221;. Note that the priors for this option assume that all amino-acid mutations are made at equal frequencies.</p>
<p class="last">Possible choices: codon_to_aa, DNA, codon, aa</p>
</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--err=none</span></kbd></td>
<td><p class="first">Control used to estimate errors in counts for all libraries (&#8220;n_start&#8221;, &#8220;n_s1&#8221;, &#8220;n_s2&#8221;). If &#8220;none&#8221; then the counts are assumed to have no errors; otherwise this should be a counts file with the same format as &#8220;n_start&#8221; giving the counts for sequencing unmutated gene.</p>
<p>Setting this option to &#8220;none&#8221; corresponds to setting to zero the error rates <span class="math">\(\xi_{r,x}\)</span> described in <a class="reference internal" href="inferdiffprefs_algorithm.html#inferdiffprefs-algorithm"><span class="std std-ref">Algorithm to infer differential preferences</span></a>.</p>
<p class="last">If you set this option to the name of a file, that file is assumed to give the counts for sequencing the error control (<span class="math">\(n_{r,x}^{\textrm{err}}\)</span>) and is used to estimate the error rate <span class="math">\(\xi_{r,x}\)</span> described in <a class="reference internal" href="inferdiffprefs_algorithm.html#inferdiffprefs-algorithm"><span class="std std-ref">Algorithm to infer differential preferences</span></a>. This file should have the format specified by <a class="reference internal" href="fileformats.html#dms-counts"><span class="std std-ref">Deep mutational scanning counts file</span></a>.</p>
</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--excludestop=False</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td><p class="first">Exclude stop codons as a possible amino acid if using &#8220;&#8211;chartype&#8221; of &#8220;codon_to_aa&#8221; or &#8220;aa&#8221;. Stop codons are only excluded if this option is specified; otherwise they are included, and are required to be specified if &#8220;&#8211;chartype&#8221; is &#8220;aa&#8221;.</p>
<p class="last">By default, when using <code class="docutils literal"><span class="pre">--chartype</span> <span class="pre">codon_to_aa</span></code> or <code class="docutils literal"><span class="pre">--chartype</span> <span class="pre">aa</span></code>, we infer differential preferences for 21 amino acids, with stop codons (denoted by <code class="docutils literal"><span class="pre">*</span></code>) one of the possibilities. If you specify the <code class="docutils literal"><span class="pre">--excludestop</span></code> option, then we constrain the differential preference for stop codons to be zero regardless of whether or not there are counts for these codons in the data, and so only infer differential preferences for the 20 non-stop amino acids.</p>
</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--alpha_start=1.0</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td><p class="first">Concentration parameter for Dirichlet prior over starting frequencies.</p>
<p>This is the parameter <span class="math">\(\alpha_{\textrm{start}}\)</span> described in <a class="reference internal" href="inferdiffprefs_algorithm.html#inferdiffprefs-algorithm"><span class="std std-ref">Algorithm to infer differential preferences</span></a>.</p>
<p class="last">Smaller values favor distributions of starting frequencies that are more unequal among the possible characters at each site; larger values favor distributions of frequencies more equal among the characters at each site.</p>
</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--alpha_pis1=1.0</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td><p class="first">Concentration parameter for Dirichlet prior over preferences for selection 1.</p>
<p>This is the parameter <span class="math">\(\alpha_{\pi_{s1}}\)</span> described in <a class="reference internal" href="inferdiffprefs_algorithm.html#inferdiffprefs-algorithm"><span class="std std-ref">Algorithm to infer differential preferences</span></a>.</p>
<p class="last">Smaller values favor distributions of preferences that are more unequal among the possible characters at each site; larger values favor distributions of preferences more equal among the characters at each site.</p>
</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--alpha_err=1.0</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td><p class="first">Concentration parameter for Dirichlet priors over error rate.</p>
<p>This gives the parameter <span class="math">\(\alpha_{\xi}\)</span> described in <a class="reference internal" href="inferdiffprefs_algorithm.html#inferdiffprefs-algorithm"><span class="std std-ref">Algorithm to infer differential preferences</span></a>.</p>
<p class="last">Smaller values favor distributions of error rates that are more unequal among the possible characters at each site; larger values favor distributions of errors more equal among the characters at each site.</p>
</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--alpha_deltapi=2.0</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td><p class="first">Concentration parameter for Dirichlet priors over differential preferences. Larger values correspond to stronger expectation of differential preferences of zero.</p>
<p>This gives the parameter <span class="math">\(\alpha_{\Delta\pi}\)</span> described in <a class="reference internal" href="inferdiffprefs_algorithm.html#inferdiffprefs-algorithm"><span class="std std-ref">Algorithm to infer differential preferences</span></a>.</p>
<p class="last">Larger values correspond to a stronger prior expectation that the differential preference is zero. You might therefore consider using a value <span class="math">\(&gt; 1\)</span> in order to enforce a prior expectation that there is no difference between the selections.</p>
</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--logfile=Base name of &quot;outfile&quot; with extension &quot;.log&quot;</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Log progress to this file; overwritten if it already exists.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--ncpus=1</span></kbd></td>
<td>Number of CPUs to use; set to -1 to use all available CPUs.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--sites</span></kbd></td>
<td>Only perform the inference for the specified sites, which should be a space separated list such as &#8220;&#8211;sites 1 2 10&#8221;. All of these sites must have data in the counts files.</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--ratio_estimation</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td><p class="first">Rather than use MCMC to estimate the differential preferences using a statistical model, we simply compute them by calculating enrichment ratios relative to wildtype for both selections, then normalizing these ratios to sum to one at each site, and finally taking the difference. The single argument for this option is MINCOUNTS, which is a number &gt; 0. If the counts for a character are less than MINCOUNTS, either due to low counts or error correction, then the counts for that character are changed to MINCOUNTS to avoid estimating ratios of zero, less than zero, or infinity.</p>
<p>A suggested value for <em>mincounts</em> is one. In general, larger values of <em>mincounts</em> conceptually corresponds to a stronger prior favoring all differential preferences being zero.</p>
<p>Differential preferences are computed using ratio estimation as follows. For each site, we set the enrichment ratio of the wildtype character <span class="math">\(\rm{wt}\)</span> equal to one. We do this for each of the two selections <span class="math">\(s1\)</span> and <span class="math">\(s2\)</span>:</p>
<div class="math">
\[\phi_{wt}^{\rm{s1}} = \phi_{wt}^{\rm{s2}} = 1\]</div>
<p>Next, for each non-wildtype character <span class="math">\(x\)</span>, we calculate the enrichment ratio relative to <span class="math">\(\rm{wt}\)</span> for <span class="math">\(s1\)</span> and <span class="math">\(s2\)</span> as:</p>
<div class="math">
\[\phi_{x}^{s1} =
\frac{
  \left(\frac{
    \max\left(\frac{\mathcal{P}}{N_r^{\rm{s1}}}, \frac{n_{r,x}^{\rm{s1}}}{N_r^{\rm{s1}}} - \frac{n_{r,x}^{\rm{err}}}{N_r^{\rm{err}}}\right)
  }{
    \frac{n_{r,\rm{wt}}^{\rm{s1}}}{N_r^{\rm{s1}}} + \delta - \frac{n_{r,\rm{wt}}^{\rm{err}}}{N_r^{\rm{err}}}
  }\right)
}
{
  \left(\frac{
    \max\left(\frac{\mathcal{P}}{N_r^{\rm{pre}}}, \frac{n_{r,x}^{\rm{pre}}}{N_r^{\rm{pre}}} - \frac{n_{r,x}^{\rm{err}}}{N_r^{\rm{err}}}\right)
  }{
    \frac{n_{r,\rm{wt}}^{\rm{pre}}}{N_r^{\rm{pre}}} + \delta - \frac{n_{r,\rm{wt}}^{\rm{err}}}{N_r^{\rm{err}}}
  }\right)
}\]</div>
<p>and</p>
<div class="math">
\[\phi_{x}^{s2} =
\frac{
  \left(\frac{
    \max\left(\frac{\mathcal{P}}{N_r^{\rm{s2}}}, \frac{n_{r,x}^{\rm{s2}}}{N_r^{\rm{s2}}} - \frac{n_{r,x}^{\rm{err}}}{N_r^{\rm{err}}}\right)
  }{
    \frac{n_{r,\rm{wt}}^{\rm{s2}}}{N_r^{\rm{s2}}} + \delta - \frac{n_{r,\rm{wt}}^{\rm{err}}}{N_r^{\rm{err}}}
  }\right)
}
{
  \left(\frac{
    \max\left(\frac{\mathcal{P}}{N_r^{\rm{pre}}}, \frac{n_{r,x}^{\rm{pre}}}{N_r^{\rm{pre}}} - \frac{n_{r,x}^{\rm{err}}}{N_r^{\rm{err}}}\right)
  }{
    \frac{n_{r,\rm{wt}}^{\rm{pre}}}{N_r^{\rm{pre}}} + \delta - \frac{n_{r,\rm{wt}}^{\rm{err}}}{N_r^{\rm{err}}}
  }\right)
}\]</div>
<p>where <span class="math">\(\mathcal{P}\)</span> is the value of <em>mincounts</em>. When <em>error_model</em> is <em>none</em>, then all terms involving the error corrections (with superscript <em>err</em>) are ignored and <span class="math">\(\delta\)</span> is set to zero; otherwise <span class="math">\(\delta\)</span> is one.</p>
<p>Next, for each character <span class="math">\(x\)</span>, including <span class="math">\(\rm{wt}\)</span>, we compute the preferences
as normalized enrichment ratios for both <span class="math">\(s1\)</span> and <span class="math">\(s2\)</span> as:</p>
<blockquote>
<div><div class="math">
\[\pi_{x}^{s1} = \frac{\phi_x^{s1}}{\sum_y \phi_y^{s1}}\]</div>
</div></blockquote>
<p>and</p>
<blockquote>
<div><div class="math">
\[\pi_{x}^{s2} = \frac{\phi_x^{s2}}{\sum_y \phi_y^{s2}}\]</div>
</div></blockquote>
<p>and finally calculate the differential preference as</p>
<blockquote class="last">
<div><div class="math">
\[\Delta\pi_{x} = \pi_{x}^{s2} - \pi_{x}^{s1}\]</div>
</div></blockquote>
</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--seed=1</span></kbd></td>
<td>Random number seed.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">-v</span>, <span class="option">--version</span></kbd></td>
<td>show program&#8217;s version number and exit</td></tr>
</tbody>
</table>
</dd>
</dl>
</div>
<div class="section" id="output">
<h2><a class="toc-backref" href="#id6">Output</a><a class="headerlink" href="#output" title="Permalink to this headline">¶</a></h2>
<p>The inferred differential preferences are in the <a class="reference internal" href="fileformats.html#diffpreferences-file"><span class="std std-ref">Differential preferences file</span></a> specified by <code class="docutils literal"><span class="pre">outfile</span></code>. A summary of the progress is in the file specified by <code class="docutils literal"><span class="pre">logfile</span></code>. In addition, some information will be written to standard output. <a class="reference internal" href="#do-not-worry-about-informational-messages">Do not worry about Informational Messages</a>.</p>
<p>Becase standard output will include lots of informational messages that obscure what is happening, it will be more informative to track the progress by watching <code class="docutils literal"><span class="pre">logfile</span></code>.</p>
</div>
<div class="section" id="do-not-worry-about-informational-messages">
<h2><a class="toc-backref" href="#id7">Do not worry about Informational Messages</a><a class="headerlink" href="#do-not-worry-about-informational-messages" title="Permalink to this headline">¶</a></h2>
<p>When you run <code class="docutils literal"><span class="pre">dms_inferdiffprefs</span></code>, you may get a lot of <code class="docutils literal"><span class="pre">Informational</span> <span class="pre">Message</span></code> output on the screen. These messages look like this:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>Informational Message: The current Metropolis proposal is about to be rejected because of the following issue:
stan::prob::dirichlet_log(N4stan5agrad3varE): prior sample sizes[1]  is 0:0, but must be &gt; 0!
If this warning occurs sporadically, such as for highly constrained variable types like covariance matrices, then the sampler is fine,
but if this warning occurs often then your model may be either severely ill-conditioned or misspecified.
</pre></div>
</div>
<p>You do <strong>not</strong> need to worry about these messages. They are automatically produced by <a class="reference external" href="http://pystan.readthedocs.org/en/latest/">PyStan</a>. Although it seems like there are a lot of these messages, running <code class="docutils literal"><span class="pre">dms_inferdiffprefs</span></code> on a 500-residue gene will invoke at least <span class="math">\(4 \times 500 = 2000\)</span> different <a class="reference external" href="http://en.wikipedia.org/wiki/Markov_chain_Monte_Carlo">MCMC</a> chains, all of which may sporadically produce this warning.</p>
<p><code class="docutils literal"><span class="pre">dms_inferdiffprefs</span></code> automatically tests for convergence and described in <a class="reference internal" href="#runtime-and-convergence">Runtime and convergence</a> and <a class="reference internal" href="inferprefs_algorithm.html#mcmc-inference"><span class="std std-ref">Inferring the preferences by MCMC.</span></a> If the program actually fails to converge for any site, you will not get an <code class="docutils literal"><span class="pre">outfile</span></code> and the program will terminate with an error message.</p>
</div>
<div class="section" id="runtime-and-convergence">
<h2><a class="toc-backref" href="#id8">Runtime and convergence</a><a class="headerlink" href="#runtime-and-convergence" title="Permalink to this headline">¶</a></h2>
<p>Because <code class="docutils literal"><span class="pre">dms_inferdiffprefs</span></code> uses <a class="reference external" href="http://en.wikipedia.org/wiki/Markov_chain_Monte_Carlo">MCMC</a> to infer the preferences as described in <a class="reference internal" href="inferdiffprefs_algorithm.html#inferdiffprefs-algorithm"><span class="std std-ref">Algorithm to infer differential preferences</span></a>, it may take the program a while to run. The program will run fastest if you use multiple CPUs (such as by using the option <code class="docutils literal"><span class="pre">--ncpus</span> <span class="pre">-1</span></code> to use all available CPUs). It should take anywhere from a few minutes to a few hours to complete on a multi-processor machine, depending on the size of the protein, which characters are being used (nucleotides are fastest, codons are slowest), and whether error controls are being included in the analysis (via <code class="docutils literal"><span class="pre">--err</span></code>).</p>
<p>The program automatically checks for <a class="reference external" href="http://en.wikipedia.org/wiki/Markov_chain_Monte_Carlo">MCMC</a> convergence using the criteria described in <a class="reference internal" href="inferprefs_algorithm.html#mcmc-inference"><span class="std std-ref">Inferring the preferences by MCMC.</span></a> The program will terminate with an error if it fails to converge; otherwise it converged properly.</p>
<p>You can make the program run dramatically faster by using the <code class="docutils literal"><span class="pre">--ratio_estimation</span></code> option. This option forgoes the Bayesian estimation and regularizing prior over the differential preferences, but will still be reasonably accurate if you have a fairly large number of counts.</p>
</div>
<div class="section" id="examples">
<h2><a class="toc-backref" href="#id9">Examples</a><a class="headerlink" href="#examples" title="Permalink to this headline">¶</a></h2>
<p>Imagine we have created a library of codon mutants of an influenza gene and selected for functional variants. We deep sequence this starting pool of functional variants and put the codon counts in a file called <code class="docutils literal"><span class="pre">mutvirus_codoncounts.txt</span></code>. We subject the starting pool of functional mutant variants to immune selection by passaging in the presence of antibody, and also do a control passage in the absence of immune selection. The codon counts from deep sequencing these pools are in the files <code class="docutils literal"><span class="pre">immuneselected_codoncounts.txt</span></code> and <code class="docutils literal"><span class="pre">controlselected_codoncounts.txt</span></code>, respectively. We also do control sequencing of unmutated virus to quantify the error rates, and put the codon counts in the file <code class="docutils literal"><span class="pre">unmutatedcontrol_codoncounts.txt</span></code>. We could then compute the differential preference <span class="math">\(\Delta\pi_{r,a}\)</span> for each amino acid <span class="math">\(a\)</span> at each site <span class="math">\(r\)</span> in the immune selected viruses versus the controls using:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">dms_inferdiffprefs</span> <span class="o">--</span><span class="n">err</span> <span class="n">unmutatedcontrol_codoncounts</span><span class="o">.</span><span class="n">txt</span> <span class="n">mutvirus_codoncounts</span><span class="o">.</span><span class="n">txt</span> <span class="n">controlselected_codoncounts</span><span class="o">.</span><span class="n">txt</span> <span class="n">immuneselected_codoncounts</span><span class="o">.</span><span class="n">txt</span> <span class="n">differentialpreferences</span><span class="o">.</span><span class="n">txt</span>
</pre></div>
</div>
<p>The differential preferences are then in the created file called <code class="docutils literal"><span class="pre">differentialpreferences.txt</span></code>, with <span class="math">\(\Delta\pi_{r,a} &gt; 0\)</span> indicating that amino acid <span class="math">\(a\)</span> is preferentially favored at <span class="math">\(r\)</span> in the immune selected versus the control selected sample. A log file summarizing output will be written to <code class="docutils literal"><span class="pre">differentialpreferences.log</span></code>.</p>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#"><code class="docutils literal"><span class="pre">dms_inferdiffprefs</span></code></a><ul>
<li><a class="reference internal" href="#overview">Overview</a></li>
<li><a class="reference internal" href="#should-you-use-this-program">Should you use this program?</a></li>
<li><a class="reference internal" href="#command-line-usage">Command-line usage</a></li>
<li><a class="reference internal" href="#output">Output</a></li>
<li><a class="reference internal" href="#do-not-worry-about-informational-messages">Do not worry about Informational Messages</a></li>
<li><a class="reference internal" href="#runtime-and-convergence">Runtime and convergence</a></li>
<li><a class="reference internal" href="#examples">Examples</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="dms_diffselection.html"
                        title="previous chapter"><code class="docutils literal"><span class="pre">dms_diffselection</span></code></a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="dms_logoplot.html"
                        title="next chapter"><code class="docutils literal"><span class="pre">dms_logoplot</span></code></a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/dms_inferdiffprefs.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="dms_logoplot.html" title="dms_logoplot"
             >next</a> |</li>
        <li class="right" >
          <a href="dms_diffselection.html" title="dms_diffselection"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">dms_tools 1.1.20 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="programs.html" >Programs</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2014, Jesse Bloom.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.4.6.
    </div>
  </body>
</html>