<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>dms_barcodedsubamplicons &mdash; dms_tools 1.1.8 documentation</title>
    
    <link rel="stylesheet" href="_static/default.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.1.8',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="dms_tools 1.1.8 documentation" href="index.html" />
    <link rel="up" title="Programs" href="programs.html" />
    <link rel="next" title="dms_summarizealignments" href="dms_summarizealignments.html" />
    <link rel="prev" title="dms_editsites" href="dms_editsites.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="dms_summarizealignments.html" title="dms_summarizealignments"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="dms_editsites.html" title="dms_editsites"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">dms_tools 1.1.8 documentation</a> &raquo;</li>
          <li><a href="programs.html" accesskey="U">Programs</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="dms-barcodedsubamplicons">
<span id="id1"></span><h1><a class="toc-backref" href="#id2"><tt class="docutils literal"><span class="pre">dms_barcodedsubamplicons</span></tt></a><a class="headerlink" href="#dms-barcodedsubamplicons" title="Permalink to this headline">¶</a></h1>
<div class="contents topic" id="contents">
<p class="topic-title first">Contents</p>
<ul class="simple">
<li><a class="reference internal" href="#dms-barcodedsubamplicons" id="id2"><tt class="docutils literal"><span class="pre">dms_barcodedsubamplicons</span></tt></a><ul>
<li><a class="reference internal" href="#overview" id="id3">Overview</a></li>
<li><a class="reference internal" href="#barcoded-subamplicon-sequencing" id="id4">Barcoded subamplicon sequencing</a></li>
<li><a class="reference internal" href="#algorithm-for-assembling-and-aligning-subamplicons" id="id5">Algorithm for assembling and aligning subamplicons</a></li>
<li><a class="reference internal" href="#command-line-usage" id="id6">Command-line usage</a></li>
<li><a class="reference internal" href="#output-files" id="id7">Output files</a></li>
<li><a class="reference internal" href="#memory-usage" id="id8">Memory usage</a></li>
</ul>
</li>
</ul>
</div>
<div class="section" id="overview">
<h2><a class="toc-backref" href="#id3">Overview</a><a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<p>This program can process the FASTQ reads generated by <a class="reference internal" href="#barcoded-subamplicon-sequencing">Barcoded subamplicon sequencing</a> to count the frequencies of each identity at each site, and to generate summary statistics.</p>
<p>If you do such an analysis for several samples, you can summarize the results with <a class="reference internal" href="dms_summarizealignments.html#dms-summarizealignments"><em>dms_summarizealignments</em></a>.</p>
<p>After you install <a class="reference external" href="http://jbloom.github.io/dms_tools">dms_tools</a>, this program will be available to run at the command line.</p>
</div>
<div class="section" id="barcoded-subamplicon-sequencing">
<h2><a class="toc-backref" href="#id4">Barcoded subamplicon sequencing</a><a class="headerlink" href="#barcoded-subamplicon-sequencing" title="Permalink to this headline">¶</a></h2>
<p>Barcoded subamplicon sequencing involves using PCR to make a series of subamplicons from the gene to be sequenced. Each of these subamplicons contains a random barcode (a string of <tt class="docutils literal"><span class="pre">N</span></tt> nucleotides) at each end that provides a unique identifier. The barcoded subamplicons are then diluted such that the number of unique ssDNA molecules is less than the read depth, and then this dilution is amplified by primers that also add the Illumina adaptors. Finally, these molecules are sequenced.</p>
<p>In analyzing the sequencing, we can then identify barcodes that are sequenced multiple times, and a consensus sequence for the subamplicon can be built from the multiple reads. The <tt class="docutils literal"><span class="pre">dms_barcodedsubamplicons</span></tt> script performs this analysis.</p>
<p>To understand how the sequencing is done in detail, example primers are shown below for sequencing of the hemagglutinin (HA) from A/WSN/1933 (H1N1) influenza:</p>
<div class="highlight-python"><div class="highlight"><pre>Rnd1for1: 5&#39;-CTTTCCCTACACGACGCTCTTCCGATCTNNNNNNNNaagcaggggaaaataaaaacaaccaaa-3&#39;                                                                                                                                                                                                                                                                                                                                                             Rnd1for427: 5&#39;-CTTTCCCTACACGACGCTCTTCCGATCTNNNNNNNNccaaggaaagttcatggcccaac-3&#39;                                                                                                                                                                                                                                                                                                                                                           Rnd1for850: 5&#39;-CTTTCCCTACACGACGCTCTTCCGATCTNNNNNNNNgtttgagtccggcatcatcacc-3&#39;                                                                                                                                                                                                                                                                                                                                              Rnd1for1276: 5&#39;-CTTTCCCTACACGACGCTCTTCCGATCTNNNNNNNNcaacaacttagaaaaaaggatggaaaatttaaataaa-3&#39;
    WSN HA with coding sequence in caps: 5&#39;-agcaaaagcaggggaaaataaaaacaaccaaaATGAAGGCAAAACTACTGGTCCTGTTATATGCATTTGTAGCTACAGATGCAGACACAATATGTATAGGCTACCATGCGAACAACTCAACCGACACTGTTGACACAATACTCGAGAAGAATGTGGCAGTGACACATTCTGTTAACCTGCTCGAAGACAGCCACAACGGGAAACTATGTAAATTAAAAGGAATAGCCCCACTACAATTGGGGAAATGTAACATCACCGGATGGCTCTTGGGAAATCCAGAATGCGACTCACTGCTTCCAGCGAGATCATGGTCCTACATTGTAGAAACACCAAACTCTGAGAATGGAGCATGTTATCCAGGAGATCTCATCGACTATGAGGAACTGAGGGAGCAATTGAGCTCAGTATCATCATTAGAAAGATTCGAAATATTTCCCAAGGAAAGTTCATGGCCCAACCACACATTCAACGGAGTAACAGTATCATGCTCCCATAGGGGAAAAAGCAGTTTTTACAGAAATTTGCTATGGCTGACGAAGAAGGGGGATTCATACCCAAAGCTGACCAATTCCTATGTGAACAATAAAGGGAAAGAAGTCCTTGTACTATGGGGTGTTCATCACCCGTCTAGCAGTGATGAGCAACAGAGTCTCTATAGTAATGGAAATGCTTATGTCTCTGTAGCGTCTTCAAATTATAACAGGAGATTCACCCCGGAAATAGCTGCAAGGCCCAAAGTAAGAGATCAACATGGGAGGATGAACTATTACTGGACCTTGCTAGAACCCGGAGACACAATAATATTTGAGGCAACTGGTAATCTAATAGCACCATGGTATGCTTTCGCACTGAGTAGAGGGTTTGAGTCCGGCATCATCACCTCAAACGCGTCAATGCATGAGTGTAACACGAAGTGTCAAACACCCCAGGGAGCTATAAACAGCAATCTCCCTTTCCAGAATATACACCCAGTCACAATAGGAGAGTGCCCAAAATATGTCAGGAGTACCAAATTGAGGATGGTTACAGGACTAAGAAACATCCCATCCATTCAATACAGAGGTCTATTTGGAGCCATTGCTGGTTTTATTGAGGGGGGATGGACTGGAATGATAGATGGATGGTATGGTTATCATCATCAGAATGAACAGGGATCAGGCTATGCAGCGGATCAAAAAAGCACACAAAATGCCATTAACGGGATTACAAACAAGGTGAACTCTGTTATCGAGAAAATGAACACTCAATTCACAGCTGTGGGTAAAGAATTCAACAACTTAGAAAAAAGGATGGAAAATTTAAATAAAAAAGTTGATGATGGGTTTCTGGACATTTGGACATATAATGCAGAATTGTTAGTTCTACTGGAAAATGAAAGGACTTTGGATTTCCATGACTTAAATGTGAAGAATCTGTACGAGAAAGTAAAAAGCCAATTAAAGAATAATGCCAAAGAAATCGGAAATGGGTGTTTTGAGTTCTACCACAAGTGTGACAATGAATGCATGGAAAGTGTAAGAAATGGGACTTATGATTATCCAAAATATTCAGAAGAATCAAAGTTGAACAGGGAAAAGATAGATGGAGTGAAATTGGAATCAATGGGGGTGTATCAGATTCTGGCGATCTACTCAACTGTCGCCAGTTCACTGGTGCTTTTGGTCTCCCTGGGGGCAATCAGTTTCTGGATGTGTTCTAATGGGTCTTTGCAGTGCAGAATATGCATCTGAgattaggatttcagaaatataaggaaaaacacccttgtttctact-3&#39;
    WSN HA with coding sequence in caps: 3&#39;-tcgttttcgtccccttttatttttgttggtttTACTTCCGTTTTGATGACCAGGACAATATACGTAAACATCGATGTCTACGTCTGTGTTATACATATCCGATGGTACGCTTGTTGAGTTGGCTGTGACAACTGTGTTATGAGCTCTTCTTACACCGTCACTGTGTAAGACAATTGGACGAGCTTCTGTCGGTGTTGCCCTTTGATACATTTAATTTTCCTTATCGGGGTGATGTTAACCCCTTTACATTGTAGTGGCCTACCGAGAACCCTTTAGGTCTTACGCTGAGTGACGAAGGTCGCTCTAGTACCAGGATGTAACATCTTTGTGGTTTGAGACTCTTACCTCGTACAATAGGTCCTCTAGAGTAGCTGATACTCCTTGACTCCCTCGTTAACTCGAGTCATAGTAGTAATCTTTCTAAGCTTTATAAAGGGTTCCTTTCAAGTACCGGGTTGGTGTGTAAGTTGCCTCATTGTCATAGTACGAGGGTATCCCCTTTTTCGTCAAAAATGTCTTTAAACGATACCGACTGCTTCTTCCCCCTAAGTATGGGTTTCGACTGGTTAAGGATACACTTGTTATTTCCCTTTCTTCAGGAACATGATACCCCACAAGTAGTGGGCAGATCGTCACTACTCGTTGTCTCAGAGATATCATTACCTTTACGAATACAGAGACATCGCAGAAGTTTAATATTGTCCTCTAAGTGGGGCCTTTATCGACGTTCCGGGTTTCATTCTCTAGTTGTACCCTCCTACTTGATAATGACCTGGAACGATCTTGGGCCTCTGTGTTATTATAAACTCCGTTGACCATTAGATTATCGTGGTACCATACGAAAGCGTGACTCATCTCCCAAACTCAGGCCGTAGTAGTGGAGTTTGCGCAGTTACGTACTCACATTGTGCTTCACAGTTTGTGGGGTCCCTCGATATTTGTCGTTAGAGGGAAAGGTCTTATATGTGGGTCAGTGTTATCCTCTCACGGGTTTTATACAGTCCTCATGGTTTAACTCCTACCAATGTCCTGATTCTTTGTAGGGTAGGTAAGTTATGTCTCCAGATAAACCTCGGTAACGACCAAAATAACTCCCCCCTACCTGACCTTACTATCTACCTACCATACCAATAGTAGTAGTCTTACTTGTCCCTAGTCCGATACGTCGCCTAGTTTTTTCGTGTGTTTTACGGTAATTGCCCTAATGTTTGTTCCACTTGAGACAATAGCTCTTTTACTTGTGAGTTAAGTGTCGACACCCATTTCTTAAGTTGTTGAATCTTTTTTCCTACCTTTTAAATTTATTTTTTCAACTACTACCCAAAGACCTGTAAACCTGTATATTACGTCTTAACAATCAAGATGACCTTTTACTTTCCTGAAACCTAAAGGTACTGAATTTACACTTCTTAGACATGCTCTTTCATTTTTCGGTTAATTTCTTATTACGGTTTCTTTAGCCTTTACCCACAAAACTCAAGATGGTGTTCACACTGTTACTTACGTACCTTTCACATTCTTTACCCTGAATACTAATAGGTTTTATAAGTCTTCTTAGTTTCAACTTGTCCCTTTTCTATCTACCTCACTTTAACCTTAGTTACCCCCACATAGTCTAAGACCGCTAGATGAGTTGACAGCGGTCAAGTGACCACGAAAACCAGAGGGACCCCCGTTAGTCAAAGACCTACACAAGATTACCCAGAAACGTCACGTCTTATACGTAGACTctaatcctaaagtctttatattcctttttgtgggaacaaagatga-5&#39;
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       Rnd1rev426: 3&#39;-gtgtgtaagttgcctcattgtcatagtacNNNNNNNNTCTAGCCTTCTCGTGTGCAGACTTGAGG-5&#39;                                                                                                                                                                                                                                                                                                                                                    Rnd1rev849: 3&#39;-agtttgcgcagttacgtactcacNNNNNNNNTCTAGCCTTCTCGTGTGCAGACTTGAGG-5&#39;                                                                                                                                                                                                                                                                                                                                                                 Rnd1rev1275: 3&#39;-actactacccaaagacctgtaaaNNNNNNNNTCTAGCCTTCTCGTGTGCAGACTTGAGG-5&#39;                                                                                                                                                                                                                                                                                                                                                    Rnd1rev1698: 3&#39;-ctaatcctaaagtctttatattcctttttgtgggaaNNNNNNNNTCTAGCCTTCTCGTGTGCAGACTTGAGG-5&#39;
</pre></div>
</div>
<p>Briefly, the steps are as follows:</p>
<blockquote>
<div><ol class="arabic">
<li><p class="first">The library of HA genes is subjected to PCR with primers to create subamplicons of <span class="math">\(\approx 425\)</span> internal nucleotides each as well as additional adaptor and primer-binding sequences. In the example above, PCRs are performed with the following primer pairs:</p>
<blockquote>
<div><ul class="simple">
<li><tt class="docutils literal"><span class="pre">Rnd1for1</span></tt> and <tt class="docutils literal"><span class="pre">Rnd1rev426</span></tt>: these primers create a subamplicon spanning nucleotides 1 to 426 (codons 1 to 142) of the WSN HA, which is the part of the sequence shown in caps above.</li>
<li><tt class="docutils literal"><span class="pre">Rnd1for427</span></tt> and <tt class="docutils literal"><span class="pre">Rnd1rev849</span></tt>: these primers create a subamplicon spanning nucleotides 427 to 849 (codons 143 to 283).</li>
<li><tt class="docutils literal"><span class="pre">Rnd1for850</span></tt> and <tt class="docutils literal"><span class="pre">Rnd1rev1275</span></tt>: these primers create a subamplicon spanning nucleotides 850 to 1275 (codons 284 to 425).</li>
<li><tt class="docutils literal"><span class="pre">Rnd1for1276</span></tt> and <tt class="docutils literal"><span class="pre">Rnd1rev1698</span></tt>: these primers create a subamplicon spanning nucleotides 1276 to 1698 (codons 426 to 566).</li>
</ul>
</div></blockquote>
</li>
<li><p class="first">The subamplicons are then diluted so that the total number of ssDNA molecules is less than the sequencing depth. Since each subamplicon molecule has 8 <tt class="docutils literal"><span class="pre">N</span></tt> nucleotides at each end, there are <span class="math">\(4^{16} \approx 4.3 \times 10^9\)</span> unique barcodes. Since the number of unique barcodes is much greater than the total number of molecules in the dilution, each barcode will generally be associated with just one molecule.</p>
</li>
<li><p class="first">The barcoded subamplicons are further amplified by primers that add the Illumina adaptors, and perhaps Illumina indices for multiplexing. For instance, here are two example primers for this second round of PCR:</p>
<div class="highlight-python"><div class="highlight"><pre>&gt;Rnd2forUniversal: forward primer for the round 2 PCRs that adds the Illumina adaptor
AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCC

&gt;Rnd2revIndex: reverse primer for the round 2 PCRs that adds the TruSeq index 1 (in lower case) as well as the Illumina adaptor
CAAGCAGAAGACGGCATACGAGATcgtgatGTGACTGGAGTTCAGACGTGTGCTCTTCC
</pre></div>
</div>
</li>
<li><p class="first">The products of the round 2 PCRs are sequenced with overlapping paired-end reads of sufficient length to cover the entire subamplicon, and the sequencing data are analyzed with the standard Illumina pipeline to generate FASTQ files with the R1 and R2 reads.</p>
</li>
<li><p class="first">The FASTQ files are analyzed with <tt class="docutils literal"><span class="pre">dms_barcodedsubamplicons</span></tt>. For the example here, the command might be:</p>
<div class="highlight-python"><div class="highlight"><pre>dms_barcodedsubamplicons WSN_HA_ WSN-HA.fasta R1.fastq.gz R2.fastq.gz 1,426,36,38 427,849,32,32 850,1275,31,37 1276,1698,46,45
</pre></div>
</div>
<p>The meanings of the arguments are detailed in the <a class="reference internal" href="#command-line-usage">Command-line usage</a>. Briefly:</p>
<blockquote>
<div><ul>
<li><p class="first"><tt class="docutils literal"><span class="pre">WSN_HA_</span></tt> is the prefix attached to the created output files.</p>
</li>
<li><p class="first"><tt class="docutils literal"><span class="pre">WSN-HA.fasta</span></tt> should be the name of a FASTA file containing the part of the gene that we are sequencing. In this example, it would be <strong>only</strong> the capitalized portion of the sequence shown above (as the lowercase parts are simply flanking sequences for primer binding). For instance, here is such a file:</p>
<div class="highlight-python"><div class="highlight"><pre>&gt;WSN HA coding sequence
ATGAAGGCAAAACTACTGGTCCTGTTATATGCATTTGTAGCTACAGATGCAGACACAATATGTATAGGCTACCATGCGAACAACTCAACCGACACTGTTGACACAATACTCGAGAAGAATGTGGCAGTGACACATTCTGTTAACCTGCTCGAAGACAGCCACAACGGGAAACTATGTAAATTAAAAGGAATAGCCCCACTACAATTGGGGAAATGTAACATCACCGGATGGCTCTTGGGAAATCCAGAATGCGACTCACTGCTTCCAGCGAGATCATGGTCCTACATTGTAGAAACACCAAACTCTGAGAATGGAGCATGTTATCCAGGAGATCTCATCGACTATGAGGAACTGAGGGAGCAATTGAGCTCAGTATCATCATTAGAAAGATTCGAAATATTTCCCAAGGAAAGTTCATGGCCCAACCACACATTCAACGGAGTAACAGTATCATGCTCCCATAGGGGAAAAAGCAGTTTTTACAGAAATTTGCTATGGCTGACGAAGAAGGGGGATTCATACCCAAAGCTGACCAATTCCTATGTGAACAATAAAGGGAAAGAAGTCCTTGTACTATGGGGTGTTCATCACCCGTCTAGCAGTGATGAGCAACAGAGTCTCTATAGTAATGGAAATGCTTATGTCTCTGTAGCGTCTTCAAATTATAACAGGAGATTCACCCCGGAAATAGCTGCAAGGCCCAAAGTAAGAGATCAACATGGGAGGATGAACTATTACTGGACCTTGCTAGAACCCGGAGACACAATAATATTTGAGGCAACTGGTAATCTAATAGCACCATGGTATGCTTTCGCACTGAGTAGAGGGTTTGAGTCCGGCATCATCACCTCAAACGCGTCAATGCATGAGTGTAACACGAAGTGTCAAACACCCCAGGGAGCTATAAACAGCAATCTCCCTTTCCAGAATATACACCCAGTCACAATAGGAGAGTGCCCAAAATATGTCAGGAGTACCAAATTGAGGATGGTTACAGGACTAAGAAACATCCCATCCATTCAATACAGAGGTCTATTTGGAGCCATTGCTGGTTTTATTGAGGGGGGATGGACTGGAATGATAGATGGATGGTATGGTTATCATCATCAGAATGAACAGGGATCAGGCTATGCAGCGGATCAAAAAAGCACACAAAATGCCATTAACGGGATTACAAACAAGGTGAACTCTGTTATCGAGAAAATGAACACTCAATTCACAGCTGTGGGTAAAGAATTCAACAACTTAGAAAAAAGGATGGAAAATTTAAATAAAAAAGTTGATGATGGGTTTCTGGACATTTGGACATATAATGCAGAATTGTTAGTTCTACTGGAAAATGAAAGGACTTTGGATTTCCATGACTTAAATGTGAAGAATCTGTACGAGAAAGTAAAAAGCCAATTAAAGAATAATGCCAAAGAAATCGGAAATGGGTGTTTTGAGTTCTACCACAAGTGTGACAATGAATGCATGGAAAGTGTAAGAAATGGGACTTATGATTATCCAAAATATTCAGAAGAATCAAAGTTGAACAGGGAAAAGATAGATGGAGTGAAATTGGAATCAATGGGGGTGTATCAGATTCTGGCGATCTACTCAACTGTCGCCAGTTCACTGGTGCTTTTGGTCTCCCTGGGGGCAATCAGTTTCTGGATGTGTTCTAATGGGTCTTTGCAGTGCAGAATATGCATCTGA
</pre></div>
</div>
</li>
<li><p class="first"><tt class="docutils literal"><span class="pre">R1.fasta.gz</span></tt> and <tt class="docutils literal"><span class="pre">R2.fastq.gz</span></tt> are the FASTQ files with the R1 and R2 reads.</p>
</li>
<li><p class="first">The remaining arguments specify where the subamplicons might align
For instance, <tt class="docutils literal"><span class="pre">1,426,36,38</span></tt> means that for the first subamplicon, the first sequenced nucleotide (first one outside primer binding site) aligns to nucleotide 1 of the reference sequence in <tt class="docutils literal"><span class="pre">WSN-HA.fasta</span></tt>, the last sequence nucleotide aligns to nucleotide 426 of the reference sequence, the first nucleotide in R1 that we want to consider in the alignment is site 36 (there is an 8 nucleotide barcode plus a 27-nucleotide primer-binding site), and the first nucleotide in R2 that we want to consider in the alignment is site 38 (there is an 8-nucleotide barcode plus a 29-nucleotide primer-binding site). The arguments <tt class="docutils literal"><span class="pre">427,849,32,32</span></tt>, <tt class="docutils literal"><span class="pre">850,1275,31,37</span></tt>, and <tt class="docutils literal"><span class="pre">1276,1698,46,45</span></tt> are the similar specifications for the other three subamplicons.</p>
</li>
</ul>
</div></blockquote>
</li>
</ol>
</div></blockquote>
</div>
<div class="section" id="algorithm-for-assembling-and-aligning-subamplicons">
<h2><a class="toc-backref" href="#id5">Algorithm for assembling and aligning subamplicons</a><a class="headerlink" href="#algorithm-for-assembling-and-aligning-subamplicons" title="Permalink to this headline">¶</a></h2>
<p>The algorithm implemented by <tt class="docutils literal"><span class="pre">dms_barcodedsubamplicons</span></tt> is as follows:</p>
<ol class="arabic">
<li><p class="first">Any read pair in which either read fails the Illumina chastity filter is discarded.</p>
</li>
<li><p class="first">All sites in each read pair are classified as &#8220;high quality&#8221; if the site is not an <tt class="docutils literal"><span class="pre">N</span></tt> (ambiguous nucleotide) and has a Q-score of at least <tt class="docutils literal"><span class="pre">--minq</span></tt>, or as &#8220;low quality&#8221; otherwise. For all low-quality sites, the called identity in the read is simply set to <tt class="docutils literal"><span class="pre">N</span></tt> (so we treat a site with a Q-score of less than <tt class="docutils literal"><span class="pre">--minq</span></tt> as equivalent to an <tt class="docutils literal"><span class="pre">N</span></tt>).</p>
</li>
<li><p class="first">We discard any read pairs for which any of the following conditions are met:</p>
<blockquote>
<div><ul class="simple">
<li>There is a low-quality site in the barcode region of either read.</li>
<li>Either read has more than <tt class="docutils literal"><span class="pre">--maxlowqfrac</span></tt> of its nucleotides that are low quality.</li>
</ul>
</div></blockquote>
</li>
<li><p class="first">We then collect all remaining read pairs for each barcode (concatenating the barcodes on the paired reads to give a total barcode length that is twice <tt class="docutils literal"><span class="pre">--barcodelength</span></tt>) and perform all subsequent operations on these barcodes.</p>
</li>
<li><p class="first">Any barcode with less than <tt class="docutils literal"><span class="pre">--minreadsperbarcode</span></tt> read pairs is discarded.</p>
</li>
<li><p class="first">The R1 and R2 reads for each remaining barcode are assembled into consensus sequences. At each site, these consensus sequences have the consensus nucleotide if that nucleotide is found in at least <tt class="docutils literal"><span class="pre">--minreadconcurrence</span></tt> of the reads; otherwise the nucleotide is set to <tt class="docutils literal"><span class="pre">N</span></tt>. During the process of building these consensus sequences, the barcode is discarded if any of the following conditions are met for either the R1 or R2 read:</p>
<blockquote>
<div><ul class="simple">
<li>The reads cannot all be made the same length by trimming no more <tt class="docutils literal"><span class="pre">--maxreadtrim</span></tt> from the 3&#8217; end.</li>
<li>Any of the reads fail to have at least <tt class="docutils literal"><span class="pre">--minreadidentity</span></tt> identical and high-quality (not <tt class="docutils literal"><span class="pre">N</span></tt>) nucleotides when compared pairwise.</li>
</ul>
</div></blockquote>
</li>
<li><p class="first">For each remaining barcode, we attempt to align the consensus sequence to each of the subamplicons specified by <tt class="docutils literal"><span class="pre">alignspecs</span></tt>. The attempted alignment does <strong>not</strong> accommodate gaps; the consensus sequence for the reads is required to align gaplessly at the position specified by <tt class="docutils literal"><span class="pre">alignspecs</span></tt>. In any region where the reads overlap (which may happen near the center of the subamplicon), if both consensus sequences (R1 and R2) report high quality nucleotides, the identity is considered ambiguous if the sequences disagree. An alignment is considered valid only if the following conditions are met:</p>
<blockquote>
<div><ul class="simple">
<li>The total fraction of non-high-quality nucleotides in the aligned region exceeds <tt class="docutils literal"><span class="pre">--maxlowqfrac</span></tt>.</li>
<li>The subamplicon alignment has no more than <tt class="docutils literal"><span class="pre">--maxmuts</span></tt> mutations relative to <tt class="docutils literal"><span class="pre">refseq</span></tt> when mutations are counted in terms of <tt class="docutils literal"><span class="pre">--chartype</span></tt> (so if <tt class="docutils literal"><span class="pre">--chartype</span></tt> is <tt class="docutils literal"><span class="pre">codon</span></tt>, this is the number of <strong>codon</strong> mutations).</li>
</ul>
</div></blockquote>
<p>If the above criteria are met, then the identities (wildtype or mutant) are counted for each site of <tt class="docutils literal"><span class="pre">--chartype</span></tt> where the are no ambiguous nucleotides in the alignment, and this barcoded subamplicon contributes to the totals reported in the <tt class="docutils literal"><span class="pre">counts.txt</span></tt> output file. If the alignment criteria are not met, the barcode is discarded as un-alignable. Note that unless you are working with an extremely repetitive gene or specify an extremely large value of <tt class="docutils literal"><span class="pre">--maxmuts</span></tt> relative to the read length, spurious alignments should not occur.</p>
</li>
</ol>
</div>
<div class="section" id="command-line-usage">
<h2><a class="toc-backref" href="#id6">Command-line usage</a><a class="headerlink" href="#command-line-usage" title="Permalink to this headline">¶</a></h2>
<p>Gathers barcoded subamplicons, aligns to reference sequence, and counts mutations. This script is part of dms_tools (version 1.1.8) written by Jesse D. Bloom. Detailed documentation is at http://jbloom.github.io/dms_tools/</p>
<pre class="literal-block">
usage: dms_barcodedsubamplicons [-h] [--barcodelength BARCODELENGTH]
                                [--chartype {codon}] [--maxmuts MAXMUTS]
                                [--minq MINQ] [--maxlowqfrac MAXLOWQFRAC]
                                [--minreadsperbarcode MINREADSPERBARCODE]
                                [--minreadidentity MINREADIDENTITY]
                                [--minreadconcurrence MINREADCONCURRENCE]
                                [--maxreadtrim MAXREADTRIM]
                                [--R1_is_antisense] [--barcodeinfo]
                                [--purgefrac PURGEFRAC] [--seed SEED] [-v]
                                outprefix refseq r1files r2files alignspecs
                                [alignspecs ...]
</pre>
<dl class="docutils">
<dt>Positional arguments:</dt>
<dd><table class="first last docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd>outprefix</kbd></td>
<td><p class="first">Prefix for output files. The suffixes of the created files are: &#8220;counts.txt&#8221; - character counts at each site; &#8220;summarystats.txt&#8221; - barcode summary stats; &#8221;.log&#8221; - file that logs progress; &#8220;barcodeinfo.txt.gz&#8221; - lists all barcodes if using &#8220;&#8211;barcodeinfo&#8221;. Existing files with these names are removed.</p>
<p class="last">For instance, if <tt class="docutils literal"><span class="pre">outprefix</span></tt> is <tt class="docutils literal"><span class="pre">WSN_HA_</span></tt> then the created output files will be <tt class="docutils literal"><span class="pre">WSN_HA_counts.txt</span></tt>, <tt class="docutils literal"><span class="pre">WSN_HA_summarystats.txt</span></tt>, <tt class="docutils literal"><span class="pre">WSN_HA.log</span></tt>, and <tt class="docutils literal"><span class="pre">WSN_HA_barcodeinfo.txt.gz</span></tt> if you use the <tt class="docutils literal"><span class="pre">--barcodeinfo</span></tt> option. The format of the output files is described in <a class="reference internal" href="#output-files">Output files</a>.</p>
</td></tr>
<tr><td class="option-group">
<kbd>refseq</kbd></td>
<td>Existing FASTA file containing gene to which we are aligning subamplicons and counting mutations.</td></tr>
<tr><td class="option-group">
<kbd>r1files</kbd></td>
<td>Comma-separated list of R1 FASTQ files (no spaces). Files can optionally be gzipped (extension .gz).</td></tr>
<tr><td class="option-group">
<kbd>r2files</kbd></td>
<td>Like &#8220;r1files&#8221; but R2 files. Must be same number of comma-separated entries as for &#8220;r1files&#8221;.</td></tr>
<tr><td class="option-group">
<kbd>alignspecs</kbd></td>
<td><p class="first">This argument is repeated to specify each subamplicon. Each occurrence is four comma-delimited integers (no spaces): &#8220;REFSEQSTART,REFSEQEND,R1START,R2START&#8221;. REFSEQSTART is nucleotide (1, 2, ... numbering) in &#8220;refseq&#8221; where nucleotide R1START in R1 aligns. REFSEQEND is nucleotide in &#8220;refseq&#8221; where nucleotide R2START in R2 aligns.</p>
<p>It is important to set <tt class="docutils literal"><span class="pre">alignspecs</span></tt> so that you don&#8217;t count the part of the subamplicon that is in the primer binding site, since the nucleotide identities in this region come from the primers rather than the templates being sequenced. See <a class="reference internal" href="#barcoded-subamplicon-sequencing">Barcoded subamplicon sequencing</a> for an example.</p>
<p class="last">Note also that the alignments will fail if you don&#8217;t set <tt class="docutils literal"><span class="pre">alignspecs</span></tt> exactly correctly for each subamplicon. So if you have reads failing to align to part of your gene, carefully check <tt class="docutils literal"><span class="pre">alignspecs</span></tt> to make sure it is correct (you can use the example in <a class="reference internal" href="#barcoded-subamplicon-sequencing">Barcoded subamplicon sequencing</a> to make sure you understand things properly).</p>
</td></tr>
</tbody>
</table>
</dd>
<dt>Options:</dt>
<dd><table class="first last docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--barcodelength=8</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td><p class="first">Length of barcodes (NNN... nucleotides) at the beginning of R1 and R2 reads.</p>
<p class="last">This is the length of the barcode on <strong>each</strong> primer. So a value of 8 corresponds to a total of 16 <tt class="docutils literal"><span class="pre">N</span></tt> nucleotides on the two primers. See <a class="reference internal" href="#barcoded-subamplicon-sequencing">Barcoded subamplicon sequencing</a> for an example of primers with 8 <tt class="docutils literal"><span class="pre">N</span></tt> nucleotides.</p>
</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--chartype=codon</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td><p class="first">Character for which we are counting mutations. Currently &#8220;codon&#8221; is only allowed value (in the future &#8220;nucleotide&#8221; might be added).</p>
<p>If <tt class="docutils literal"><span class="pre">chartype</span></tt> is <tt class="docutils literal"><span class="pre">codon</span></tt> then <tt class="docutils literal"><span class="pre">refseq</span></tt> must provide a valid coding sequence (length a multiple of 3).</p>
<p class="last">Possible choices: codon</p>
</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--maxmuts=4</span></kbd></td>
<td>Only align subamplicons (consensus from a barcode) if &lt;= this many mismatches with &#8220;refseq&#8221; counted in terms of &#8220;chartype&#8221;.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--minq=15</span></kbd></td>
<td>Only consider nucleotides with Q scores &gt;= this number.</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--maxlowqfrac=0.075</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Only retain read pairs if no &#8220;N&#8221; or Q &lt; &#8220;minq&#8221; nucleotides in barcodes, and total fraction of such nucleotides is &lt;= this number in each read individually and in the eventual subamplicon built from the barcodes.</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--minreadsperbarcode=2</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Retain only barcodes with &gt;= this many reads that align gaplessly with &gt;= &#8220;minreadidentity&#8221; identical high-quality nucleotides; should be &gt;= 2.</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--minreadidentity=0.9</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Retain only barcodes where all reads have &gt;= this fraction of identical high-quality nucleotides that align gaplessly.</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--minreadconcurrence=0.75</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>For retained barcodes, only make mutation calls when &gt;= this fraction of reads concur.</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--maxreadtrim=3</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td><p class="first">If R1 or R2 reads for same barcode are not all same length, trim up to this many nucleotides from 3&#8217; end; if still not same length then discard.</p>
<p class="last">This option exists because it appears that Illumina runs of a given read length don&#8217;t always return all of the reads with exactly that read length.</p>
</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--R1_is_antisense=False</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Is the R1 read in the antisense direction of &#8220;refseq&#8221;?</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--barcodeinfo=False</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>If you specify this option, create a file with suffix &#8220;barcodeinfo.txt.gz&#8221; containing information for each barcode. This file is quite large, and its creation will about double the program run time.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--purgefrac=0</span></kbd></td>
<td><p class="first">Randomly purge barcodes with this probability, thereby subsampling the data. You might want a value &gt; 0 to estimate how the results depend on the sequencing depth.</p>
<p class="last">Why would you ever want to purge some of the barcodes? You may be trying to determine whether sequencing to a higher depth (which will give you more barcodes with at least <tt class="docutils literal"><span class="pre">--nreadsperbarcode</span></tt> reads) will improve the quality of your results. If you set <tt class="docutils literal"><span class="pre">--purgefrac</span></tt> to a value &gt; 0 (say 0.5), you&#8217;ll see how the results would be affected if you had fewer reads. If these results are noticeably worse by some objective measure, this supports that idea that your read depth might be in regime where greater depth would help.</p>
</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--seed=1</span></kbd></td>
<td>Random number seed used to select reads for purging when using &#8220;&#8211;purgefrac&#8221;.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">-v</span>, <span class="option">--version</span></kbd></td>
<td>show program&#8217;s version number and exit</td></tr>
</tbody>
</table>
</dd>
</dl>
</div>
<div class="section" id="output-files">
<h2><a class="toc-backref" href="#id7">Output files</a><a class="headerlink" href="#output-files" title="Permalink to this headline">¶</a></h2>
<p>The formats of the output files are given below in terms of the file suffix that is appended to <tt class="docutils literal"><span class="pre">--outprefix</span></tt>. The data in the first two output files (<tt class="docutils literal"><span class="pre">counts.txt</span></tt> and <tt class="docutils literal"><span class="pre">summarystats.txt</span></tt>) can be used as input to <a class="reference internal" href="dms_summarizealignments.html#dms-summarizealignments"><em>dms_summarizealignments</em></a> to make plots that summarize the results for several samples.</p>
<ul>
<li><p class="first"><tt class="docutils literal"><span class="pre">counts.txt</span></tt> : This file gives the counts of the different characters of <tt class="docutils literal"><span class="pre">--chartype</span></tt> in the <a class="reference internal" href="fileformats.html#dms-counts"><em>Deep mutational scanning counts file</em></a> format. So if <tt class="docutils literal"><span class="pre">--chartype</span></tt> is <tt class="docutils literal"><span class="pre">codon</span></tt>, then these will be the counts of codon identities. Note that files in <a class="reference internal" href="fileformats.html#dms-counts"><em>Deep mutational scanning counts file</em></a> format can be used directly as input to programs like <a class="reference internal" href="dms_inferprefs.html#dms-inferprefs"><em>dms_inferprefs</em></a> and <a class="reference internal" href="dms_inferdiffprefs.html#dms-inferdiffprefs"><em>dms_inferdiffprefs</em></a>.</p>
</li>
<li><p class="first"><tt class="docutils literal"><span class="pre">summarystats.txt</span></tt> : This text file summarizes the numbers of reads and barcodes parsed. For instance, here is an example:</p>
<div class="highlight-python"><div class="highlight"><pre>total read pairs = 4580258
read pairs that fail Illumina filter = 0
low quality read pairs = 915131
barcodes randomly purged = 0
discarded barcodes with 1 reads = 844768
aligned barcodes with 2 reads = 466387
discarded barcodes with 2 reads = 6457
un-alignable barcodes with 2 reads = 40753
aligned barcodes with 3 reads = 262149
discarded barcodes with 3 reads = 5426
un-alignable barcodes with 3 reads = 24891
aligned barcodes with 4 reads = 120828
discarded barcodes with 4 reads = 3281
un-alignable barcodes with 4 reads = 7675
aligned barcodes with 5 reads = 45342
discarded barcodes with 5 reads = 1516
un-alignable barcodes with 5 reads = 2863
aligned barcodes with 6 reads = 14545
discarded barcodes with 6 reads = 577
un-alignable barcodes with 6 reads = 934
aligned barcodes with 7 reads = 4060
discarded barcodes with 7 reads = 203
un-alignable barcodes with 7 reads = 243
aligned barcodes with 8 reads = 996
discarded barcodes with 8 reads = 65
un-alignable barcodes with 8 reads = 52
aligned barcodes with 9 reads = 250
discarded barcodes with 9 reads = 20
un-alignable barcodes with 9 reads = 11
aligned barcodes with 10 reads = 51
discarded barcodes with 10 reads = 3
un-alignable barcodes with 10 reads = 3
aligned barcodes with 11 reads = 10
discarded barcodes with 11 reads = 1
aligned barcodes with 12 reads = 2
</pre></div>
</div>
</li>
<li><p class="first"><tt class="docutils literal"><span class="pre">.log</span></tt> : This is a text file that logs the progress of the program.</p>
</li>
<li><p class="first"><tt class="docutils literal"><span class="pre">barcodeinfo.txt.gz</span></tt> : This file is only created if the <tt class="docutils literal"><span class="pre">--barcodeinfo</span></tt> option is used. In this case, the file will give all barcodes and their associated reads, and will explain whether or not the barcode was successfully retained during the alignment process.</p>
</li>
</ul>
</div>
<div class="section" id="memory-usage">
<h2><a class="toc-backref" href="#id8">Memory usage</a><a class="headerlink" href="#memory-usage" title="Permalink to this headline">¶</a></h2>
<p><tt class="docutils literal"><span class="pre">dms_barcodedsubamplicons</span></tt> stores all of the reads in the FASTQ files in memory. Therefore, it uses a substantial amount of memory, typically around a gigabyte per million paired-end sequencing reads. However, for typical data sets such memory usage is well within the capacity of modern large-memory nodes.</p>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#"><tt class="docutils literal"><span class="pre">dms_barcodedsubamplicons</span></tt></a><ul>
<li><a class="reference internal" href="#overview">Overview</a></li>
<li><a class="reference internal" href="#barcoded-subamplicon-sequencing">Barcoded subamplicon sequencing</a></li>
<li><a class="reference internal" href="#algorithm-for-assembling-and-aligning-subamplicons">Algorithm for assembling and aligning subamplicons</a></li>
<li><a class="reference internal" href="#command-line-usage">Command-line usage</a></li>
<li><a class="reference internal" href="#output-files">Output files</a></li>
<li><a class="reference internal" href="#memory-usage">Memory usage</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="dms_editsites.html"
                        title="previous chapter"><tt class="docutils literal"><span class="pre">dms_editsites</span></tt></a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="dms_summarizealignments.html"
                        title="next chapter"><tt class="docutils literal"><span class="pre">dms_summarizealignments</span></tt></a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/dms_barcodedsubamplicons.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="dms_summarizealignments.html" title="dms_summarizealignments"
             >next</a> |</li>
        <li class="right" >
          <a href="dms_editsites.html" title="dms_editsites"
             >previous</a> |</li>
        <li><a href="index.html">dms_tools 1.1.8 documentation</a> &raquo;</li>
          <li><a href="programs.html" >Programs</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2014, Jesse Bloom.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3a0.
    </div>
  </body>
</html>