<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>dms_matchsubassembledbarcodes &mdash; dms_tools 1.2.0 documentation</title>
    
    <link rel="stylesheet" href="_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.2.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="dms_tools 1.2.0 documentation" href="index.html" />
    <link rel="up" title="Programs" href="programs.html" />
    <link rel="next" title="dms_summarizealignments" href="dms_summarizealignments.html" />
    <link rel="prev" title="dms_subassemble" href="dms_subassemble.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="dms_summarizealignments.html" title="dms_summarizealignments"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="dms_subassemble.html" title="dms_subassemble"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">dms_tools 1.2.0 documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="programs.html" accesskey="U">Programs</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="dms-matchsubassembledbarcodes">
<span id="id1"></span><h1><a class="toc-backref" href="#id2"><code class="docutils literal"><span class="pre">dms_matchsubassembledbarcodes</span></code></a><a class="headerlink" href="#dms-matchsubassembledbarcodes" title="Permalink to this headline">¶</a></h1>
<div class="contents topic" id="contents">
<p class="topic-title first">Contents</p>
<ul class="simple">
<li><a class="reference internal" href="#dms-matchsubassembledbarcodes" id="id2"><code class="docutils literal"><span class="pre">dms_matchsubassembledbarcodes</span></code></a><ul>
<li><a class="reference internal" href="#overview" id="id3">Overview</a></li>
<li><a class="reference internal" href="#command-line-usage" id="id4">Command-line usage</a></li>
<li><a class="reference internal" href="#output-files" id="id5">Output files</a><ul>
<li><a class="reference internal" href="#single-mutant-counts-file" id="id6">Single-mutant counts file</a></li>
<li><a class="reference internal" href="#all-mutant-counts-file" id="id7">All-mutant counts file</a></li>
<li><a class="reference internal" href="#variant-counts-file" id="id8">Variant counts file</a></li>
<li><a class="reference internal" href="#read-statistics-file" id="id9">Read statistics file</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="section" id="overview">
<h2><a class="toc-backref" href="#id3">Overview</a><a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<p>You would use this program if you have subassembled some sequences using <a class="reference internal" href="dms_subassemble.html#dms-subassemble"><span class="std std-ref">dms_subassemble</span></a>, and then have performed some selections on the subassembled sequences and sequenced the barcodes.</p>
<p>This program assumes that you have sequenced the barcodes with overlapping paired-end reads (the overlap helps with error correction). You give it the R1 and R2 FASTQ files as <code class="docutils literal"><span class="pre">r1files</span></code> and <code class="docutils literal"><span class="pre">r2files</span></code> , and it extracts the barcodes and uses them to determine the counts for individual variants / mutations.</p>
<p>This program assumes that the R1 files (<code class="docutils literal"><span class="pre">r1files</span></code>) are reading the barcodes in the same orientation as they are specified in the <code class="docutils literal"><span class="pre">subassembled</span></code> file created by <a class="reference internal" href="dms_subassemble.html#dms-subassemble"><span class="std std-ref">dms_subassemble</span></a>, and so that the R2 files (<code class="docutils literal"><span class="pre">r2files</span></code> are reading the reverse complement of these barcodes). The options <code class="docutils literal"><span class="pre">--r1start</span></code> and <code class="docutils literal"><span class="pre">--r2end</span></code> determine where exactly in the reads the barcode is found.</p>
<p>The barcodes are only retained if all of the following conditions are met:</p>
<blockquote>
<div><ul class="simple">
<li>Both reads in the pair pass the Illumina filter.</li>
<li>At least one barcode has a quality score <span class="math">\(\ge\)</span> <code class="docutils literal"><span class="pre">--minquality</span></code> for each site.</li>
<li>Both read pairs agree on the identity of all positions in the barcode without any <code class="docutils literal"><span class="pre">N</span></code> nucleotides.</li>
</ul>
</div></blockquote>
</div>
<div class="section" id="command-line-usage">
<h2><a class="toc-backref" href="#id4">Command-line usage</a><a class="headerlink" href="#command-line-usage" title="Permalink to this headline">¶</a></h2>
<p>Extract barcodes after quality filter, match to subassembled variants, write variant counts files, make summary plots. This script is part of dms_tools (version 1.dev2) written by the Bloom Lab (see <a class="reference external" href="https://github.com/jbloomlab/dms_tools/graphs/contributors">https://github.com/jbloomlab/dms_tools/graphs/contributors</a> for all contributors). Detailed documentation is at <a class="reference external" href="http://jbloomlab.github.io/dms_tools/">http://jbloomlab.github.io/dms_tools/</a></p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">dms_matchsubassembledbarcodes</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">minquality</span> <span class="n">MINQUALITY</span><span class="p">]</span>
                                     <span class="p">[</span><span class="o">--</span><span class="n">r1start</span> <span class="n">R1START</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">r2end</span> <span class="n">R2END</span><span class="p">]</span>
                                     <span class="p">[</span><span class="o">--</span><span class="n">purgefrac</span> <span class="n">PURGEFRAC</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">v</span><span class="p">]</span>
                                     <span class="n">outprefix</span> <span class="n">subassembled</span> <span class="n">r1files</span> <span class="n">r2files</span>
</pre></div>
</div>
<dl class="docutils">
<dt>Positional arguments:</dt>
<dd><table class="first last docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd>outprefix</kbd></td>
<td><p class="first">Prefix for output files.</p>
<p class="last">See <a class="reference internal" href="#output-files">Output files</a> for a list of the created files.</p>
</td></tr>
<tr><td class="option-group">
<kbd>subassembled</kbd></td>
<td><p class="first">File output by &#8216;dms_subassemble&#8217; that contains the barcoded subassembled variants.</p>
<p>This program matches the barcodes against barcodes that have been subassembled by <a class="reference internal" href="dms_subassemble.html#dms-subassemble"><span class="std std-ref">dms_subassemble</span></a>. You therefore need to provide it with a <code class="docutils literal"><span class="pre">*_subassembled_variants.txt</span></code> created by <a class="reference internal" href="dms_subassemble.html#dms-subassemble"><span class="std std-ref">dms_subassemble</span></a>. This should be in the <a class="reference internal" href="fileformats.html#subassembly-file"><span class="std std-ref">Subassembly file</span></a> format.</p>
<p class="last">The barcode length is determined from this file, since it will list the barcodes and so specify their length.</p>
</td></tr>
<tr><td class="option-group">
<kbd>r1files</kbd></td>
<td>Comma-separated list of R1 FASTQ files (no spaces) containing barcodes. Files can optionally be gzipped (extension .gz).</td></tr>
<tr><td class="option-group">
<kbd>r2files</kbd></td>
<td>Like &#8216;r1files&#8217; but for R2 read.</td></tr>
</tbody>
</table>
</dd>
<dt>Options:</dt>
<dd><table class="first last docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--minquality=25</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Each site in barcode must have quality &gt;= this for at least one read.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--r1start=1</span></kbd></td>
<td><p class="first">Barcode starts at this position in R1.</p>
<p>This argument specifies where in R1 the barcode starts. By default, we assume that it starts with the first nucleotide of the read.</p>
<p class="last">Must be an integer <span class="math">\(\ge 1\)</span>.</p>
</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--r2end=0</span></kbd></td>
<td><p class="first">Barcode ends this many nucleotides before end of R2.</p>
<p>This argument specifies where in R2 the barcode ends. By default, we assume that it ends with the last nucleotide of the read, and so set a value of zero. If instead the barcode ends two nucleotides before the end of R2, you would set a value of 2.</p>
<p class="last">Must be an integer <span class="math">\(\ge 0\)</span>.</p>
</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--purgefrac=0</span></kbd></td>
<td><p class="first">Randomly purge barcode read pairs in &#8216;r1files&#8217; and &#8216;r2files&#8217; with this probability (subsample the data).</p>
<p>This option is useful if you are trying to figure out whether to sequence to greater depth. You can use it to randomly subsample your data, and so figure out if your results change if you have fewer reads (which is an indication if more reads will be helpful).</p>
<p class="last">The purging here is for the barcode read pairs in <code class="docutils literal"><span class="pre">r1files</span></code> and <code class="docutils literal"><span class="pre">r2files</span></code>, <strong>not</strong> for the subassembled variants in <code class="docutils literal"><span class="pre">subassembled</span></code>.</p>
</td></tr>
<tr><td class="option-group">
<kbd><span class="option">-v</span>, <span class="option">--version</span></kbd></td>
<td>show program&#8217;s version number and exit</td></tr>
</tbody>
</table>
</dd>
</dl>
</div>
<div class="section" id="output-files">
<h2><a class="toc-backref" href="#id5">Output files</a><a class="headerlink" href="#output-files" title="Permalink to this headline">¶</a></h2>
<p>The following output files are created. Each has the prefix specified by <code class="docutils literal"><span class="pre">outprefix</span></code> with the following suffixes.</p>
<div class="section" id="single-mutant-counts-file">
<h3><a class="toc-backref" href="#id6">Single-mutant counts file</a><a class="headerlink" href="#single-mutant-counts-file" title="Permalink to this headline">¶</a></h3>
<p>This file has the prefix specified by <code class="docutils literal"><span class="pre">outprefix</span></code> followed by the suffix <code class="docutils literal"><span class="pre">_singlemutcounts.txt</span></code>. This file is in the format of a <a class="reference internal" href="fileformats.html#dms-counts"><span class="std std-ref">Deep mutational scanning counts file</span></a> files. The mutation counts at each site are <strong>only</strong> for single-mutant variants; any subassembled variants with multiple mutations are ignored in this file. The wildtype counts for each site given by this file are <strong>only</strong> for fully wildtype sequence.</p>
<p>In other words, if we have a sequence with a single mutation <code class="docutils literal"><span class="pre">GGC31AGT</span></code>, it only contributes to the single count of the mutation to <code class="docutils literal"><span class="pre">AGT</span></code> at position 31. But a wildtype sequence contributes to the wildtype counts at all positions. Therefore, the statistic at each site give how that single mutant fared relative to the wildtype sequence.</p>
<p>This file is suitable for use by <a class="reference internal" href="dms_inferprefs.html#dms-inferprefs"><span class="std std-ref">dms_inferprefs</span></a> or <a class="reference internal" href="dms_inferdiffprefs.html#dms-inferdiffprefs"><span class="std std-ref">dms_inferdiffprefs</span></a> regardless of whether your library consisted of only singly mutated genes or singly and multiply mutated genes.</p>
</div>
<div class="section" id="all-mutant-counts-file">
<h3><a class="toc-backref" href="#id7">All-mutant counts file</a><a class="headerlink" href="#all-mutant-counts-file" title="Permalink to this headline">¶</a></h3>
<p>This file has the prefix specified by <code class="docutils literal"><span class="pre">outprefix</span></code> followed by the suffix <code class="docutils literal"><span class="pre">_allmutcounts.txt</span></code>. This file is in the format of a <a class="reference internal" href="fileformats.html#dms-counts"><span class="std std-ref">Deep mutational scanning counts file</span></a> files. The mutation counts here include all mutations at all sites.</p>
<p>In other words, a sequence with a single mutation <code class="docutils literal"><span class="pre">GGC31AGT</span></code> contributes a count of a mutation to <code class="docutils literal"><span class="pre">AGT</span></code> at site 31 <strong>and</strong> a wildtype count at all other positions. A wildtype sequence contributes a wildtype count at all positions. A double mutant <code class="docutils literal"><span class="pre">GAG40ATT,AAA46AAT</span></code> contributes a count of a mutation to <code class="docutils literal"><span class="pre">ATT</span></code> at site 40, a mutation to <code class="docutils literal"><span class="pre">AAT</span></code> at site 46, and a wildtype count at all other positions.</p>
<p>This file is suitable for use by <a class="reference internal" href="dms_inferprefs.html#dms-inferprefs"><span class="std std-ref">dms_inferprefs</span></a> or <a class="reference internal" href="dms_inferdiffprefs.html#dms-inferdiffprefs"><span class="std std-ref">dms_inferdiffprefs</span></a> only if your library consists of singly and multiply mutated clones with a Poisson distribution of the number of mutations per clone. If your library is all single mutants, you should use the <a class="reference internal" href="#single-mutant-counts-file">Single-mutant counts file</a> instead.</p>
</div>
<div class="section" id="variant-counts-file">
<h3><a class="toc-backref" href="#id8">Variant counts file</a><a class="headerlink" href="#variant-counts-file" title="Permalink to this headline">¶</a></h3>
<p>This file has the prefix specified by <code class="docutils literal"><span class="pre">outprefix</span></code> followed by the suffix <code class="docutils literal"><span class="pre">_variantcounts.txt</span></code>.</p>
<p>After a header line beginning with <code class="docutils literal"><span class="pre">#</span></code>, each line has three space-delimited entries:</p>
<blockquote>
<div><ul class="simple">
<li>The number of counts for a barcode (including barcodes with no counts).</li>
<li>The barcode itself.</li>
<li>The mutations separated by a comma but no space, or <em>no_mutations</em> if there are no mutations.</li>
</ul>
</div></blockquote>
<p>Here are a few example lines:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="c1">#COUNTS BARCODE MUTATIONS</span>
<span class="mi">5</span> <span class="n">TTACCATG</span> <span class="n">no_mutations</span>
<span class="mi">3</span> <span class="n">GATACATG</span> <span class="n">GGA31GCT</span>
<span class="mi">1</span> <span class="n">ATACCGAT</span> <span class="n">GGA10AAA</span><span class="p">,</span><span class="n">ATA22GTG</span>
<span class="mi">0</span> <span class="n">CCATAGAT</span> <span class="n">no_mutations</span>
</pre></div>
</div>
</div>
<div class="section" id="read-statistics-file">
<h3><a class="toc-backref" href="#id9">Read statistics file</a><a class="headerlink" href="#read-statistics-file" title="Permalink to this headline">¶</a></h3>
<p>This file has the prefix specified by <code class="docutils literal"><span class="pre">outprefix</span></code> followed by the suffix <code class="docutils literal"><span class="pre">_stats.txt</span></code>.</p>
<p>It gives the fate of all the read pairs in <code class="docutils literal"><span class="pre">r1files</span></code> and <code class="docutils literal"><span class="pre">r2files</span></code>.</p>
<p>It has the following format (although the lines can be in an arbitrary order):</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">nreads</span> <span class="o">=</span> <span class="mi">1000</span>
<span class="n">nfiltered</span> <span class="o">=</span> <span class="mi">10</span>
<span class="n">nlowquality</span> <span class="o">=</span> <span class="mi">47</span>
<span class="n">nmismatches</span> <span class="o">=</span> <span class="mi">50</span>
<span class="n">nunrecognized</span> <span class="o">=</span> <span class="mi">43</span>
<span class="n">nretained</span> <span class="o">=</span> <span class="mi">850</span>
<span class="n">n0mut</span> <span class="o">=</span> <span class="mi">250</span>
<span class="n">n1mut</span> <span class="o">=</span> <span class="mi">309</span>
<span class="n">n2mut</span> <span class="o">=</span> <span class="mi">191</span>
<span class="n">n3mut</span> <span class="o">=</span> <span class="mi">73</span>
<span class="n">n4mut</span> <span class="o">=</span> <span class="mi">20</span>
<span class="n">n5mut</span> <span class="o">=</span> <span class="mi">6</span>
<span class="n">n6mut</span> <span class="o">=</span> <span class="mi">1</span>
</pre></div>
</div>
<p>The meanings are:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">nreads</span></code> is the total number of read pairs</li>
<li><code class="docutils literal"><span class="pre">nfiltered</span></code> is the number removed by the Illumina quality filter</li>
<li><code class="docutils literal"><span class="pre">nlowquality</span></code> is the number that fail to meet the quality filter specified by <code class="docutils literal"><span class="pre">--minquality</span></code></li>
<li><code class="docutils literal"><span class="pre">nmismatches</span></code> is the number that fail because they are mismatched at one or more positions.</li>
<li><code class="docutils literal"><span class="pre">nunrecognized</span></code> is the number of barcodes that don&#8217;t match one of the subassembled ones present in <code class="docutils literal"><span class="pre">subassembled</span></code>.</li>
<li><code class="docutils literal"><span class="pre">nretained</span></code> is the number retained.</li>
<li>The retained reads are then further categorized by how many mutations they have, giving <code class="docutils literal"><span class="pre">n0mut</span></code>, <code class="docutils literal"><span class="pre">n1mut</span></code>, etc for reads that map to unmutated, singly mutated, etc variants.</li>
<li>If <code class="docutils literal"><span class="pre">--purgefrac</span></code> is nonzero, then there is also a key <code class="docutils literal"><span class="pre">nrandomlypurged</span></code> indicating the number of reads purged.</li>
</ul>
</div></blockquote>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#"><code class="docutils literal"><span class="pre">dms_matchsubassembledbarcodes</span></code></a><ul>
<li><a class="reference internal" href="#overview">Overview</a></li>
<li><a class="reference internal" href="#command-line-usage">Command-line usage</a></li>
<li><a class="reference internal" href="#output-files">Output files</a><ul>
<li><a class="reference internal" href="#single-mutant-counts-file">Single-mutant counts file</a></li>
<li><a class="reference internal" href="#all-mutant-counts-file">All-mutant counts file</a></li>
<li><a class="reference internal" href="#variant-counts-file">Variant counts file</a></li>
<li><a class="reference internal" href="#read-statistics-file">Read statistics file</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="dms_subassemble.html"
                        title="previous chapter"><code class="docutils literal"><span class="pre">dms_subassemble</span></code></a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="dms_summarizealignments.html"
                        title="next chapter"><code class="docutils literal"><span class="pre">dms_summarizealignments</span></code></a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/dms_matchsubassembledbarcodes.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="dms_summarizealignments.html" title="dms_summarizealignments"
             >next</a> |</li>
        <li class="right" >
          <a href="dms_subassemble.html" title="dms_subassemble"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">dms_tools 1.2.0 documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="programs.html" >Programs</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2014, Jesse Bloom.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.4.1.
    </div>
  </body>
</html>