<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>dms_summarizealignments &mdash; dms_tools 1.1.10 documentation</title>
    
    <link rel="stylesheet" href="_static/default.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.1.10',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="dms_tools 1.1.10 documentation" href="index.html" />
    <link rel="up" title="Programs" href="programs.html" />
    <link rel="next" title="Data file formats" href="fileformats.html" />
    <link rel="prev" title="dms_barcodedsubamplicons" href="dms_barcodedsubamplicons.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="fileformats.html" title="Data file formats"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="dms_barcodedsubamplicons.html" title="dms_barcodedsubamplicons"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">dms_tools 1.1.10 documentation</a> &raquo;</li>
          <li><a href="programs.html" accesskey="U">Programs</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="dms-summarizealignments">
<span id="id1"></span><h1><a class="toc-backref" href="#id2"><tt class="docutils literal"><span class="pre">dms_summarizealignments</span></tt></a><a class="headerlink" href="#dms-summarizealignments" title="Permalink to this headline">¶</a></h1>
<div class="contents topic" id="contents">
<p class="topic-title first">Contents</p>
<ul class="simple">
<li><a class="reference internal" href="#dms-summarizealignments" id="id2"><tt class="docutils literal"><span class="pre">dms_summarizealignments</span></tt></a><ul>
<li><a class="reference internal" href="#overview" id="id3">Overview</a></li>
<li><a class="reference internal" href="#command-line-usage" id="id4">Command-line usage</a></li>
<li><a class="reference internal" href="#example-usage" id="id5">Example usage</a></li>
<li><a class="reference internal" href="#output-files" id="id6">Output files</a><ul>
<li><a class="reference internal" href="#reads-pdf" id="id7"><tt class="docutils literal"><span class="pre">reads.pdf</span></tt></a></li>
<li><a class="reference internal" href="#barcodes-pdf" id="id8"><tt class="docutils literal"><span class="pre">barcodes.pdf</span></tt></a></li>
<li><a class="reference internal" href="#mutfreqs-pdf" id="id9"><tt class="docutils literal"><span class="pre">mutfreqs.pdf</span></tt></a></li>
<li><a class="reference internal" href="#depth-pdf" id="id10"><tt class="docutils literal"><span class="pre">depth.pdf</span></tt></a></li>
<li><a class="reference internal" href="#mutdepth-pdf" id="id11"><tt class="docutils literal"><span class="pre">mutdepth.pdf</span></tt></a></li>
<li><a class="reference internal" href="#mutcounts-all-pdf" id="id12"><tt class="docutils literal"><span class="pre">mutcounts_all.pdf</span></tt></a></li>
<li><a class="reference internal" href="#mutcounts-multi-nt-pdf" id="id13"><tt class="docutils literal"><span class="pre">mutcounts_multi_nt.pdf</span></tt></a></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="section" id="overview">
<h2><a class="toc-backref" href="#id3">Overview</a><a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<p><tt class="docutils literal"><span class="pre">dms_summarizealignments</span></tt> is a program included with the <a class="reference external" href="http://jbloom.github.io/dms_tools">dms_tools</a> package. It can be used to make plots that summarize the results after examining several samples with <a class="reference internal" href="dms_barcodedsubamplicons.html#dms-barcodedsubamplicons"><em>dms_barcodedsubamplicons</em></a>.</p>
<p>After you install <a class="reference external" href="http://jbloom.github.io/dms_tools">dms_tools</a>, this program will be available to run at the command line.</p>
</div>
<div class="section" id="command-line-usage">
<h2><a class="toc-backref" href="#id4">Command-line usage</a><a class="headerlink" href="#command-line-usage" title="Permalink to this headline">¶</a></h2>
<p>Makes plots that summarize alignmnents for one or more samples. Designed to be run after you have used another program to make your alignments, and now you want to visualize the results. This script is part of dms_tools (version 1.1.10) written by Jesse D. Bloom. Detailed documentation is at http://jbloom.github.io/dms_tools/</p>
<pre class="literal-block">
usage: dms_summarizealignments [-h] [--chartype {codon}]
                               [--maxmutcounts MAXMUTCOUNTS]
                               [--maxperbarcode MAXPERBARCODE]
                               [--writemutfreqs] [-v]
                               outprefix {barcodedsubamplicons} alignments
                               [alignments ...]
</pre>
<dl class="docutils">
<dt>Positional arguments:</dt>
<dd><table class="first last docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd>outprefix</kbd></td>
<td><p class="first">Prefix for the output files. Suffixes are: &#8216;mutfreqs.pdf&#8217; - average per-codon mutation frequencies; &#8216;mutcounts_all.pdf&#8217; - distribution of number of times each mutation is counted; &#8216;mutcounts_multi_nt.pdf&#8217; - distribution of number of times each multi-nucleotide mutation is counted; &#8216;depth.pdf&#8217; - sequencing depth along gene; &#8216;mutdepth.pdf&#8217; - frequencies of mutations along gene; &#8216;reads.pdf&#8217; - number of reads and their usability; &#8216;barcodes.pdf&#8217; - number of barcodes and their usability. Any existing files with these names are overwritten.</p>
<p class="last">See <a class="reference internal" href="#output-files">Output files</a> for a detailed description of the created files and examples.</p>
</td></tr>
<tr><td class="option-group">
<kbd>alignment_type</kbd></td>
<td><p class="first">The type of alignments that are being summarized. Use &#8216;barcodedsubamplicons&#8217; if the alignments were built with &#8216;dms_barcodedsubamplicons&#8217;.</p>
<p>Currently, <tt class="docutils literal"><span class="pre">dms_summarizealignments</span></tt> assumes that the alignments were created using <a class="reference internal" href="dms_barcodedsubamplicons.html#dms-barcodedsubamplicons"><em>dms_barcodedsubamplicons</em></a>.</p>
<p class="last">Possible choices: barcodedsubamplicons</p>
</td></tr>
<tr><td class="option-group">
<kbd>alignments</kbd></td>
<td><p class="first">This argument is repeated to specify each alignment that is being summarized. Each repetition is two comma-delimited strings (no spaces): ALIGNPREFIX,NAME. ALIGNPREFIX is the value of &#8216;&#8211;outprefix&#8217; used when calling the alignment program and NAME is the name assigned to that sample in the summaries. We expect to find all of the alignment output files with ALIGNPREFIX; these are the input data.</p>
<p class="last">Essentially, <tt class="docutils literal"><span class="pre">ALIGNPREFIX</span></tt> should be the value of <tt class="docutils literal"><span class="pre">--outprefix</span></tt> used when running <a class="reference internal" href="dms_barcodedsubamplicons.html#dms-barcodedsubamplicons"><em>dms_barcodedsubamplicons</em></a>. This program expects to find the <tt class="docutils literal"><span class="pre">counts.txt</span></tt> and <tt class="docutils literal"><span class="pre">summarystats.txt</span></tt> files that are created by running <a class="reference internal" href="dms_barcodedsubamplicons.html#dms-barcodedsubamplicons"><em>dms_barcodedsubamplicons</em></a> with <tt class="docutils literal"><span class="pre">--outprefix</span></tt> as <tt class="docutils literal"><span class="pre">ALIGNPREFIX</span></tt>. <tt class="docutils literal"><span class="pre">NAME</span></tt> is simply the name given to each sample in the <a class="reference internal" href="#output-files">Output files</a>.</p>
</td></tr>
</tbody>
</table>
</dd>
<dt>Options:</dt>
<dd><table class="first last docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--chartype=codon</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td><p class="first">Character type used for the alignments / mutation counting.</p>
<p class="last">Possible choices: codon</p>
</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--maxmutcounts=25</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Maximum x-value for &#8216;mutcounts_all.pdf&#8217; and &#8216;mutcounts_multi_nt.pdf&#8217; plots.</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--maxperbarcode=3</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>In &#8216;barcodes.pdf&#8217; plot, group all barcodes with &gt;= this many reads.</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--writemutfreqs=False</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Write a file &#8216;mutfreqs.txt&#8217; that gives the numerical values plotted in &#8216;mutfreqs.pdf&#8217;?</td></tr>
<tr><td class="option-group">
<kbd><span class="option">-v</span>, <span class="option">--version</span></kbd></td>
<td>show program&#8217;s version number and exit</td></tr>
</tbody>
</table>
</dd>
</dl>
</div>
<div class="section" id="example-usage">
<h2><a class="toc-backref" href="#id5">Example usage</a><a class="headerlink" href="#example-usage" title="Permalink to this headline">¶</a></h2>
<p>Imagine that you have six samples, <em>DNA</em>, <em>mutDNA</em>, <em>virus-p1</em>, <em>mutvirus-p1</em>, <em>virus-p2</em>, and <em>mutvirus-p2</em>, and furthermore imagine that for each sample you ran <a class="reference internal" href="dms_barcodedsubamplicons.html#dms-barcodedsubamplicons"><em>dms_barcodedsubamplicons</em></a> with its <tt class="docutils literal"><span class="pre">--outprefix</span></tt> option to values of <tt class="docutils literal"><span class="pre">DNA/DNA_</span></tt>, <tt class="docutils literal"><span class="pre">mutDNA/mutDNA_</span></tt>, etc. You could then summarize the results with the following command:</p>
<div class="highlight-python"><div class="highlight"><pre>dms_summarizealignments alignmentsummary_ barcodedsubamplicons DNA/DNA_,DNA mutDNA/mutDNA_,mutDNA virus-p1/virus-p1_,virus-p1 mutvirus-p1/mutvirus-p1_,mutvirus-p1 virus-p2/virus-p2_,virus-p2 mutvirus-p2/mutvirus-p2_,mutvirus-p2
</pre></div>
</div>
<p>Because this command uses <tt class="docutils literal"><span class="pre">outprefix</span></tt> of <tt class="docutils literal"><span class="pre">alignmentsummary_</span></tt>, the <a class="reference internal" href="#output-files">Output files</a> will have names like <tt class="docutils literal"><span class="pre">alignmentsummary_mutfreqs.pdf</span></tt>, <tt class="docutils literal"><span class="pre">alignmentsummary_mutcounts_all.pdf</span></tt>, etc. The <tt class="docutils literal"><span class="pre">alignments</span></tt> option specifies the prefix for each of the six samples, and then gives them the same names as above (<em>DNA</em>, <em>mutDNA</em>, etc).</p>
</div>
<div class="section" id="output-files">
<h2><a class="toc-backref" href="#id6">Output files</a><a class="headerlink" href="#output-files" title="Permalink to this headline">¶</a></h2>
<p>Here are the output files, with the names that would be produced if running <tt class="docutils literal"><span class="pre">dms_summarizealignments</span></tt> as in the <a class="reference internal" href="#example-usage">Example usage</a>.</p>
<div class="section" id="reads-pdf">
<h3><a class="toc-backref" href="#id7"><tt class="docutils literal"><span class="pre">reads.pdf</span></tt></a><a class="headerlink" href="#reads-pdf" title="Permalink to this headline">¶</a></h3>
<p>This plot shows the number of read pairs that failed the Illumina chastity filter, were filtered by <a class="reference internal" href="dms_barcodedsubamplicons.html#dms-barcodedsubamplicons"><em>dms_barcodedsubamplicons</em></a> as being of low quality, or were successfully processed and assigned to a barcode.</p>
<a class="reference internal image-reference" href="_images/alignmentsummary_reads.pdf"><img alt="alignmentsummary_reads.pdf" class="align-center" src="_images/alignmentsummary_reads.pdf" style="width: 60%;" /></a>
</div>
<div class="section" id="barcodes-pdf">
<h3><a class="toc-backref" href="#id8"><tt class="docutils literal"><span class="pre">barcodes.pdf</span></tt></a><a class="headerlink" href="#barcodes-pdf" title="Permalink to this headline">¶</a></h3>
<p>For all barcodes for which read were assigned, this plot shows the number barcodes with one, two, or at least three read pairs. For each category, it also indicates whether it was possible to build a consensus and align it to the reference sequence, whether the consensus was not alignable to the reference sequence, or whether the barcode simply had to be discarded due to insufficient matching reads.</p>
<a class="reference internal image-reference" href="_images/alignmentsummary_barcodes.pdf"><img alt="alignmentsummary_barcodes.pdf" class="align-center" src="_images/alignmentsummary_barcodes.pdf" style="width: 60%;" /></a>
</div>
<div class="section" id="mutfreqs-pdf">
<h3><a class="toc-backref" href="#id9"><tt class="docutils literal"><span class="pre">mutfreqs.pdf</span></tt></a><a class="headerlink" href="#mutfreqs-pdf" title="Permalink to this headline">¶</a></h3>
<p>This plot shows the average per-codon mutation rate across the gene as determined from the aligned barcodes. Each bar is split: the left half categories mutations as to whether they are synonymous, nonsynonymous, or to stop codons; the right categorizes mutations by the number of nucleotide changes to that codon.</p>
<a class="reference internal image-reference" href="_images/alignmentsummary_mutfreqs.pdf"><img alt="alignmentsummary_mutfreqs.pdf" class="align-center" src="_images/alignmentsummary_mutfreqs.pdf" style="width: 60%;" /></a>
</div>
<div class="section" id="depth-pdf">
<h3><a class="toc-backref" href="#id10"><tt class="docutils literal"><span class="pre">depth.pdf</span></tt></a><a class="headerlink" href="#depth-pdf" title="Permalink to this headline">¶</a></h3>
<p>This plot shows the per-codon sequencing depth as a function of the codon position in the primary sequence.</p>
<a class="reference internal image-reference" href="_images/alignmentsummary_depth.pdf"><img alt="alignmentsummary_depth.pdf" class="align-center" src="_images/alignmentsummary_depth.pdf" style="width: 60%;" /></a>
</div>
<div class="section" id="mutdepth-pdf">
<h3><a class="toc-backref" href="#id11"><tt class="docutils literal"><span class="pre">mutdepth.pdf</span></tt></a><a class="headerlink" href="#mutdepth-pdf" title="Permalink to this headline">¶</a></h3>
<p>This plot shows the per-codon mutation rate as a function of the codon position in the primary sequence.</p>
<a class="reference internal image-reference" href="_images/alignmentsummary_mutdepth.pdf"><img alt="alignmentsummary_mutdepth.pdf" class="align-center" src="_images/alignmentsummary_mutdepth.pdf" style="width: 60%;" /></a>
</div>
<div class="section" id="mutcounts-all-pdf">
<h3><a class="toc-backref" href="#id12"><tt class="docutils literal"><span class="pre">mutcounts_all.pdf</span></tt></a><a class="headerlink" href="#mutcounts-all-pdf" title="Permalink to this headline">¶</a></h3>
<p>This plot shows the number of times each of the possible codon mutations and each of the possible synonymous codon mutations was observed. For each number of counts, it indicates the fraction of all of these possible mutations that were observed at least that many times.</p>
<a class="reference internal image-reference" href="_images/alignmentsummary_mutcounts_all.pdf"><img alt="alignmentsummary_mutcounts_all.pdf" class="align-center" src="_images/alignmentsummary_mutcounts_all.pdf" style="width: 60%;" /></a>
</div>
<div class="section" id="mutcounts-multi-nt-pdf">
<h3><a class="toc-backref" href="#id13"><tt class="docutils literal"><span class="pre">mutcounts_multi_nt.pdf</span></tt></a><a class="headerlink" href="#mutcounts-multi-nt-pdf" title="Permalink to this headline">¶</a></h3>
<p>This plot shows the number of times each codon mutation was observed <strong>only</strong> for mutations that involve multi-nucleotide changes to the codon. This plot might be preferable to the one above in some cases because single-nucleotide codon changes can arise due to errors, but multi-nucleotide changes typically do not.</p>
<a class="reference internal image-reference" href="_images/alignmentsummary_mutcounts_multi_nt.pdf"><img alt="alignmentsummary_mutcounts_multi_nt.pdf" class="align-center" src="_images/alignmentsummary_mutcounts_multi_nt.pdf" style="width: 60%;" /></a>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#"><tt class="docutils literal"><span class="pre">dms_summarizealignments</span></tt></a><ul>
<li><a class="reference internal" href="#overview">Overview</a></li>
<li><a class="reference internal" href="#command-line-usage">Command-line usage</a></li>
<li><a class="reference internal" href="#example-usage">Example usage</a></li>
<li><a class="reference internal" href="#output-files">Output files</a><ul>
<li><a class="reference internal" href="#reads-pdf"><tt class="docutils literal"><span class="pre">reads.pdf</span></tt></a></li>
<li><a class="reference internal" href="#barcodes-pdf"><tt class="docutils literal"><span class="pre">barcodes.pdf</span></tt></a></li>
<li><a class="reference internal" href="#mutfreqs-pdf"><tt class="docutils literal"><span class="pre">mutfreqs.pdf</span></tt></a></li>
<li><a class="reference internal" href="#depth-pdf"><tt class="docutils literal"><span class="pre">depth.pdf</span></tt></a></li>
<li><a class="reference internal" href="#mutdepth-pdf"><tt class="docutils literal"><span class="pre">mutdepth.pdf</span></tt></a></li>
<li><a class="reference internal" href="#mutcounts-all-pdf"><tt class="docutils literal"><span class="pre">mutcounts_all.pdf</span></tt></a></li>
<li><a class="reference internal" href="#mutcounts-multi-nt-pdf"><tt class="docutils literal"><span class="pre">mutcounts_multi_nt.pdf</span></tt></a></li>
</ul>
</li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="dms_barcodedsubamplicons.html"
                        title="previous chapter"><tt class="docutils literal"><span class="pre">dms_barcodedsubamplicons</span></tt></a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="fileformats.html"
                        title="next chapter">Data file formats</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/dms_summarizealignments.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="fileformats.html" title="Data file formats"
             >next</a> |</li>
        <li class="right" >
          <a href="dms_barcodedsubamplicons.html" title="dms_barcodedsubamplicons"
             >previous</a> |</li>
        <li><a href="index.html">dms_tools 1.1.10 documentation</a> &raquo;</li>
          <li><a href="programs.html" >Programs</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2014, Jesse Bloom.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3a0.
    </div>
  </body>
</html>