<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Algorithm to infer site-specific preferences &mdash; dms_tools 1.1.6 documentation</title>
    
    <link rel="stylesheet" href="_static/default.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.1.6',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="dms_tools 1.1.6 documentation" href="index.html" />
    <link rel="up" title="Algorithms" href="algorithms.html" />
    <link rel="next" title="Algorithm to infer differential preferences" href="inferdiffprefs_algorithm.html" />
    <link rel="prev" title="Algorithms" href="algorithms.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="inferdiffprefs_algorithm.html" title="Algorithm to infer differential preferences"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="algorithms.html" title="Algorithms"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">dms_tools 1.1.6 documentation</a> &raquo;</li>
          <li><a href="algorithms.html" accesskey="U">Algorithms</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="algorithm-to-infer-site-specific-preferences">
<span id="inferprefs-algorithm"></span><h1><a class="toc-backref" href="#id1">Algorithm to infer site-specific preferences</a><a class="headerlink" href="#algorithm-to-infer-site-specific-preferences" title="Permalink to this headline">¶</a></h1>
<p>Here is a description of the algorithm that <a class="reference internal" href="dms_inferprefs.html#dms-inferprefs"><em>dms_inferprefs</em></a> uses to infer the site-specific preferences.</p>
<div class="contents topic" id="contents">
<p class="topic-title first">Contents</p>
<ul class="simple">
<li><a class="reference internal" href="#algorithm-to-infer-site-specific-preferences" id="id1">Algorithm to infer site-specific preferences</a><ul>
<li><a class="reference internal" href="#definition-of-the-site-specific-preferences" id="id2">Definition of the site-specific preferences.</a></li>
<li><a class="reference internal" href="#the-experimental-data-from-deep-sequencing" id="id3">The experimental data from deep sequencing.</a></li>
<li><a class="reference internal" href="#assumption-that-mutations-and-errors-are-rare" id="id4">Assumption that mutations and errors are rare.</a></li>
<li><a class="reference internal" href="#actual-frequencies-and-likelihoods-of-observing-experimental-data" id="id5">Actual frequencies and likelihoods of observing experimental data.</a></li>
<li><a class="reference internal" href="#priors-over-the-unknown-parameters" id="id6">Priors over the unknown parameters.</a><ul>
<li><a class="reference internal" href="#characters-are-nucleotides-preferences-are-for-nucleotides" id="id7">Characters are nucleotides; preferences are for nucleotides.</a></li>
<li><a class="reference internal" href="#characters-are-codons-preferences-are-for-codons" id="id8">Characters are codons; preferences are for codons.</a></li>
<li><a class="reference internal" href="#characters-are-codons-preferences-are-for-amino-acids" id="id9">Characters are codons; preferences are for amino acids.</a></li>
<li><a class="reference internal" href="#characters-are-amino-acids-preferences-are-for-amino-acids" id="id10">Characters are amino acids; preferences are for amino acids.</a></li>
</ul>
</li>
<li><a class="reference internal" href="#inferring-the-preferences-by-mcmc" id="id11">Inferring the preferences by MCMC.</a></li>
</ul>
</li>
</ul>
</div>
<div class="section" id="definition-of-the-site-specific-preferences">
<h2><a class="toc-backref" href="#id2">Definition of the site-specific preferences.</a><a class="headerlink" href="#definition-of-the-site-specific-preferences" title="Permalink to this headline">¶</a></h2>
<p>We assume each site <span class="math">\(r\)</span> in our gene has a preference
<span class="math">\(\pi_{r,x} \ge 0\)</span> for each character <span class="math">\(x\)</span>, where the
characters can be nucleotides, codons, or amino acids. The preferences
are defined as follows: if <span class="math">\(\mu_{r,x}\)</span> is the frequency of
<span class="math">\(x\)</span> at <span class="math">\(r\)</span> in the mutant library pre-selection and
<span class="math">\(f_{r,x}\)</span> is the frequency of <span class="math">\(x\)</span> in the library
post-selection, then</p>
<div class="math" id="equation-pirx">
<span class="eqno">(1)</span>\[f_{r,x} = \frac{\mu_{r,x} \times \pi_{r,x}}{\sum_y \mu_{r,y} \times \pi_{r,y}}.\]</div>
<p>This equation implies that <span class="math">\(1 = \sum_x \pi_{r,x}\)</span>.</p>
</div>
<div class="section" id="the-experimental-data-from-deep-sequencing">
<h2><a class="toc-backref" href="#id3">The experimental data from deep sequencing.</a><a class="headerlink" href="#the-experimental-data-from-deep-sequencing" title="Permalink to this headline">¶</a></h2>
<p>We create mutant libraries such that each site <span class="math">\(r\)</span> is potentially
mutated from the wildtype identity &nbsp;to any of the other possible
characters. We use deep sequencing to count the appearances of each
character <span class="math">\(x\)</span> at each site <span class="math">\(r\)</span> in this mutant library; since
this sequencing is performed prior to selection for gene function, we
refer to it as <em>pre</em>-selection. Let <span class="math">\(N_r^{\textrm{pre}}\)</span>&nbsp;be the total number of sequencing
reads covering <span class="math">\(r\)</span>, and let <span class="math">\(n_{r,x}^{\textrm{pre}}\)</span> &nbsp;be the number that report <span class="math">\(x\)</span>
(note that
<span class="math">\(\mbox{$N_r^{\textrm{pre}}$}= \sum_x \mbox{$n_{r,x}^{\textrm{pre}}$}\)</span>).</p>
<p>We impose an experimental selection on the mutant library with some
biologically relevant pressure that favors some mutations and disfavor
others. We then use deep sequencing to count the characters in this
selected library; since this sequencing is performed after selection, we
refer to it as <em>post</em>-selection. Let <span class="math">\(N_r^{\textrm{post}}\)</span>&nbsp;be the total number of sequencing
reads covering <span class="math">\(r\)</span>, and let <span class="math">\(n_{r,x}^{\textrm{post}}\)</span>&nbsp;be the number that report <span class="math">\(x\)</span>
(note that
<span class="math">\(\mbox{$N_r^{\textrm{post}}$}= \sum_x \mbox{$n_{r,x}^{\textrm{post}}$}\)</span>).</p>
<p>We allow for the possibility that the deep sequencing is not completely
accurate; for instance, perhaps some of the reads that report a mutant
character reflect a sequencing error rather than a true mutation. The
rates of such errors can be quantified by control sequencing of the
unmutated gene, so that any counts at <span class="math">\(r\)</span> of
<span class="math">\(x \ne \mbox{$\operatorname{wt}\left(r\right)$}\)</span> reflect sequencing errors. It
is possible that the error rates for sequencing the pre-selection and
post-selection libraries are different, as for instance would arise in
sequencing an RNA virus for which the post-selection libraries must be
reverse-transcribed to DNA prior to sequencing. Let <span class="math">\(N_r^{\textrm{err,pre}}\)</span>&nbsp;be the total number
of sequencing reads covering <span class="math">\(r\)</span> in the pre-selection error
control, and let <span class="math">\(n_{r,x}^{\textrm{err,pre}}\)</span>&nbsp;be the number that report <span class="math">\(x\)</span> (note that
<span class="math">\(\mbox{$N_r^{\textrm{err,pre}}$}= \sum_x \mbox{$n_{r,x}^{\textrm{err,pre}}$}\)</span>).
Make similar definitions of <span class="math">\(N_r^{\textrm{err,post}}\)</span>&nbsp;and <span class="math">\(n_{r,x}^{\textrm{err,post}}\)</span>&nbsp;for the post-selection error control.</p>
<p>For notational compactness, we define vectors that contain the counts of
all characters at each site <span class="math">\(r\)</span> for each sample:
<span class="math">\(\mbox{$\mathbf{n_r^{\textbf{pre}}}$}= \left(\cdots, \mbox{$n_{r,x}^{\textrm{pre}}$}, \cdots\right)\)</span>,
<span class="math">\(\mbox{$\mathbf{n_r^{\textbf{post}}}$}= \left(\cdots, \mbox{$n_{r,x}^{\textrm{post}}$}, \cdots\right)\)</span>,
<span class="math">\(\mbox{$\mathbf{n_r^{\textbf{err,pre}}}$}= \left(\cdots, \mbox{$n_{r,x}^{\textrm{err,pre}}$}, \cdots\right)\)</span>,
and
<span class="math">\(\mbox{$\mathbf{n_r^{\textbf{err,post}}}$}= \left(\cdots, \mbox{$n_{r,x}^{\textrm{err,post}}$}, \cdots\right)\)</span>.</p>
</div>
<div class="section" id="assumption-that-mutations-and-errors-are-rare">
<h2><a class="toc-backref" href="#id4">Assumption that mutations and errors are rare.</a><a class="headerlink" href="#assumption-that-mutations-and-errors-are-rare" title="Permalink to this headline">¶</a></h2>
<p>The samples described above allow for the possibility of errors
as well as the actual mutations.
We assume that the the mutagenesis and error rates rate at each site are
sufficiently low that most characters are wildtype, so for instance
<span class="math">\(\mbox{$N_r^{\textrm{pre}}$}\sim \mbox{$n_{r,\operatorname{wt}\left(r\right)}^{\textrm{pre}}$}\gg \mbox{$n_{r,x}^{\textrm{pre}}$}\)</span>
for all <span class="math">\(x \ne \mbox{$\operatorname{wt}\left(r\right)$}\)</span>, and
<span class="math">\(\mbox{$N_r^{\textrm{err,pre}}$}\sim \mbox{$n_{r,\operatorname{wt}\left(r\right)}^{\textrm{err,pre}}$}\gg \mbox{$n_{r,x}^{\textrm{err,pre}}$}\)</span>
for all <span class="math">\(x \ne \mbox{$\operatorname{wt}\left(r\right)$}\)</span>. This allows us to
ignore as negligibly rare the possibility that the same site experiences
both a mutation and an error in a single molecule, which simplifies the
analysis below.</p>
</div>
<div class="section" id="actual-frequencies-and-likelihoods-of-observing-experimental-data">
<h2><a class="toc-backref" href="#id5">Actual frequencies and likelihoods of observing experimental data.</a><a class="headerlink" href="#actual-frequencies-and-likelihoods-of-observing-experimental-data" title="Permalink to this headline">¶</a></h2>
<p>We have defined <span class="math">\(\mu_{r,x}\)</span> and <span class="math">\(f_{r,x}\)</span> as the frequencies
of <span class="math">\(x\)</span> at site <span class="math">\(r\)</span> in the pre-selection and post-selection
libraries, respectively. Also define <span class="math">\(\epsilon_{r,x}\)</span> to be the
frequency at which <span class="math">\(r\)</span> is identified as <span class="math">\(x\)</span> in sequencing
the error control for the pre-selection library, and let
<span class="math">\(\rho_{r,x}\)</span> be the frequency at which <span class="math">\(r\)</span> is identified as
<span class="math">\(x\)</span> in sequencing the error control for the post-selection
library. For notational compactness, we define vectors of these
frequencies for all characters at each site <span class="math">\(r\)</span>:
<span class="math">\(\mbox{$\boldsymbol{\mathbf{\mu_r}}$}= \left(\cdots, \mu_{r,x}, \cdots\right)\)</span>,
<span class="math">\(\mbox{$\boldsymbol{\mathbf{f_r}}$}= \left(\cdots, f_{r,x}, \cdots\right)\)</span>,
<span class="math">\(\mbox{$\boldsymbol{\mathbf{\epsilon_r}}$}= \left(\cdots, \epsilon_{r,x}, \cdots\right)\)</span>,
and
<span class="math">\(\mbox{$\boldsymbol{\mathbf{\rho_r}}$}= \left(\cdots, \rho_{r,x}, \cdots\right)\)</span>.
We also define
<span class="math">\(\mbox{$\boldsymbol{\mathbf{\pi_r}}$}= \left(\cdots, \pi_{r,x}, \cdots\right)\)</span>.
Note that Equation <a href="#equation-pirx">(1)</a> implies that</p>
<div class="math" id="equation-pir">
<span class="eqno">(2)</span>\[\mbox{$\boldsymbol{\mathbf{f_r}}$}= \frac{\mbox{$\boldsymbol{\mathbf{\mu_r}}$}\circ \mbox{$\boldsymbol{\mathbf{\pi_r}}$}}{\mbox{$\boldsymbol{\mathbf{\mu_r}}$}\cdot \mbox{$\boldsymbol{\mathbf{\pi_r}}$}}\]</div>
<p>where <span class="math">\(\circ\)</span> is the <a class="reference external" href="http://en.wikipedia.org/wiki/Hadamard_product_(matrices)">Hadamard product</a>.</p>
<p>Unless we exhaustively sequence every molecule in each library, the
observed counts will not precisely reflect their actual frequencies in
the libraries, since the deep sequencing only observes a sample of the
molecules. If the deep sequencing represents a random sampling of a
small fraction of a large library of molecules, then the likelihood of
observing some specific set of counts will be multinomially distributed
around the actual frequencies. So</p>
<div class="math" id="equation-pr_nrpre">
<span class="eqno">(3)</span>\[\Pr\left(\mbox{$\mathbf{n_r^{\textbf{pre}}}$}\mid \mbox{$N_r^{\textrm{pre}}$}, \mbox{$\boldsymbol{\mathbf{\mu_r}}$}, \mbox{$\boldsymbol{\mathbf{\epsilon_r}}$}\right) = \operatorname{Multinomial}\left(\mbox{$\mathbf{n_r^{\textbf{pre}}}$}; \mbox{$N_r^{\textrm{pre}}$}, \mbox{$\boldsymbol{\mathbf{\mu_r}}$}+ \mbox{$\boldsymbol{\mathbf{\epsilon_r}}$}- \mbox{$\boldsymbol{\mathbf{\delta_r}}$}\right)\]</div>
<p>where <span class="math">\(\operatorname{Multinomial}\)</span> denotes the <a class="reference external" href="http://en.wikipedia.org/wiki/Multinomial_distribution">Multinomial</a> distribution,
<span class="math">\(\mbox{$\boldsymbol{\mathbf{\delta_r}}$}= \left(\cdots, \delta_{x, \operatorname{wt}\left(r\right)}, \cdots\right)\)</span>
is a vector for which the element corresponding to <span class="math">\(\operatorname{wt}\left(r\right)\)</span>&nbsp;is one and all other
elements are zero (<span class="math">\(\delta_{xy}\)</span> is the <a class="reference external" href="http://en.wikipedia.org/wiki/Kronecker_delta">Kronecker delta</a>), and we
have assumed that the probability that a site experiences both a
mutation and an error is negligibly small. Similarly,</p>
<div class="math" id="equation-pr_nrerrpre">
<span class="eqno">(4)</span>\[\Pr\left(\mbox{$\mathbf{n_r^{\textbf{err,pre}}}$}\mid \mbox{$N_r^{\textrm{err,pre}}$}, \mbox{$\boldsymbol{\mathbf{\epsilon_r}}$}\right) = \operatorname{Multinomial}\left(\mbox{$\mathbf{n_r^{\textbf{err,pre}}}$}; \mbox{$N_r^{\textrm{err,pre}}$}, \mbox{$\boldsymbol{\mathbf{\epsilon_r}}$}\right)\]</div>
<div class="math" id="equation-pr_nrerrpost">
<span class="eqno">(5)</span>\[\Pr\left(\mbox{$\mathbf{n_r^{\textbf{err,post}}}$}\mid \mbox{$N_r^{\textrm{err,post}}$}, \mbox{$\boldsymbol{\mathbf{\rho_r}}$}\right) = \operatorname{Multinomial}\left(\mbox{$\mathbf{n_r^{\textbf{err,post}}}$}; \mbox{$N_r^{\textrm{err,post}}$}, \mbox{$\boldsymbol{\mathbf{\rho_r}}$}\right)\]</div>
<div class="math" id="equation-pr_nrpost">
<span class="eqno">(6)</span>\[\begin{split}\begin{eqnarray}
\Pr\left(\mbox{$\mathbf{n_r^{\textbf{post}}}$}\mid \mbox{$N_r^{\textrm{post}}$}, \mbox{$\boldsymbol{\mathbf{\mu_r}}$}, \mbox{$\boldsymbol{\mathbf{\pi_r}}$}, \mbox{$\boldsymbol{\mathbf{\rho_r}}$}\right) &amp;=&amp; \operatorname{Multinomial}\left(\mbox{$\mathbf{n_r^{\textbf{post}}}$}; \mbox{$N_r^{\textrm{post}}$}, \mbox{$\boldsymbol{\mathbf{f_r}}$}+ \mbox{$\boldsymbol{\mathbf{\rho_r}}$}- \mbox{$\boldsymbol{\mathbf{\delta_r}}$}\right) \\
&amp;=&amp; \operatorname{Multinomial}\left(\mbox{$\mathbf{n_r^{\textbf{post}}}$}; \mbox{$N_r^{\textrm{post}}$}, \frac{\mbox{$\boldsymbol{\mathbf{\mu_r}}$}\circ \mbox{$\boldsymbol{\mathbf{\pi_r}}$}}{\mbox{$\boldsymbol{\mathbf{\mu_r}}$}\cdot \mbox{$\boldsymbol{\mathbf{\pi_r}}$}} + \mbox{$\boldsymbol{\mathbf{\rho_r}}$}- \mbox{$\boldsymbol{\mathbf{\delta_r}}$}\right).
\end{eqnarray}\end{split}\]</div>
</div>
<div class="section" id="priors-over-the-unknown-parameters">
<h2><a class="toc-backref" href="#id6">Priors over the unknown parameters.</a><a class="headerlink" href="#priors-over-the-unknown-parameters" title="Permalink to this headline">¶</a></h2>
<p>We specify <a class="reference external" href="http://en.wikipedia.org/wiki/Dirichlet_distribution">Dirichlet</a> priors over the four parameter vectors for each
site <span class="math">\(r\)</span>:</p>
<div class="math" id="equation-pr_pir">
<span class="eqno">(7)</span>\[\begin{split}\Pr\left(\boldsymbol{\mathbf{\pi_r}}\right) &amp;=&amp; \operatorname{Dirichlet}\left(\boldsymbol{\mathbf{\pi_r}}; \alpha_{\pi} \times \mathbf{1}\right)\end{split}\]</div>
<div class="math" id="equation-pr_mur">
<span class="eqno">(8)</span>\[\begin{split}\Pr\left(\boldsymbol{\mathbf{\mu_r}}\right) &amp;=&amp; \operatorname{Dirichlet}\left(\boldsymbol{\mathbf{\mu_r}}; \alpha_{\mu} \times \mathcal{N}_x \times \boldsymbol{\mathbf{a_{r,\mu}}}\right)\end{split}\]</div>
<div class="math" id="equation-pr_epsilonr">
<span class="eqno">(9)</span>\[\begin{split}\Pr\left(\boldsymbol{\mathbf{\epsilon_r}}\right) &amp;=&amp; \operatorname{Dirichlet}\left(\boldsymbol{\mathbf{\epsilon_r}}; \alpha_{\epsilon} \times \mathcal{N}_x \times \boldsymbol{\mathbf{a_{r,\epsilon}}}\right)\end{split}\]</div>
<div class="math" id="equation-pr_rhor">
<span class="eqno">(10)</span>\[\begin{split}\Pr\left(\boldsymbol{\mathbf{\rho_r}}\right) &amp;=&amp; \operatorname{Dirichlet}\left(\boldsymbol{\mathbf{\rho_r}}; \alpha_{\rho} \times \mathcal{N}_x \times \boldsymbol{\mathbf{a_{r,\rho}}}\right)\end{split}\]</div>
<p>where <span class="math">\(\mathbf{1}\)</span> is a vector of ones, <span class="math">\(\mathcal{N}_x\)</span> is
the number of characters (i.e. 64 for codons, 20 for amino acids, 4 for
nucleotides), the <span class="math">\(\alpha\)</span> parameters (i.e. <span class="math">\(\alpha_{\pi}\)</span>,
<span class="math">\(\alpha_{\mu}\)</span>, <span class="math">\(\alpha_{\epsilon}\)</span>, and
<span class="math">\(\alpha_{\rho}\)</span>) are scalar concentration parameters with values
<span class="math">\(&gt; 0\)</span> specified by the user, and the <span class="math">\(\mathbf{a_r}\)</span> vectors
have entries <span class="math">\(&gt; 0\)</span> that sum to one.</p>
<p>We specify the prior vectors <span class="math">\(\mathbf{a_{r,\mu}}\)</span>, <span class="math">\(\mathbf{a_{r,\epsilon}}\)</span>, and <span class="math">\(\mathbf{a_{r,\rho}}\)</span>&nbsp;in terms of the average per-site mutation or error rates over the entire library.</p>
<p>Our prior assumption is that the rate of sequencing errors depends on the number of nucleotides being changed &#8211; for nucleotide characters all mutations have only one nucleotide changed, but for codon characters there can be one, two, or three nucleotides changed.
Specifically, the average per-site error rate for mutations with <span class="math">\(m\)</span> nucleotide changes <span class="math">\(\overline{\epsilon_m}\)</span> and <span class="math">\(\overline{\rho_m}\)</span> in the pre-selection and post-selection controls, respectively, are defined as</p>
<div class="math" id="equation-avgepsilonm">
<span class="eqno">(11)</span>\[\begin{split}\overline{\epsilon_m} &amp;=&amp; \frac{1}{L}\sum\limits_r \frac{1}{\mbox{$N_r^{\textrm{err,pre}}$}}\sum\limits_{x} \mbox{$n_{r,x}^{\textrm{err,pre}}$}\times \delta_{m,D_{x,\operatorname{wt}\left(r\right)}}\end{split}\]</div>
<div class="math" id="equation-avgrhom">
<span class="eqno">(12)</span>\[\begin{split}\overline{\rho_m} &amp;=&amp; \frac{1}{L}\sum\limits_r \frac{1}{\mbox{$N_r^{\textrm{err,post}}$}}\sum\limits_{x} \mbox{$n_{r,x}^{\textrm{err,post}}$}\times \delta_{m,D_{x,\operatorname{wt}\left(r\right)}}\end{split}\]</div>
<p>where <span class="math">\(L\)</span> is the number of sites with deep mutational scanning data, <span class="math">\(r\)</span> ranges over all of these sites, <span class="math">\(x\)</span> ranges over all characters (nucleotides or codons), and <span class="math">\(D_{x,\operatorname{wt}\left(r\right)}\)</span> is the number of nucleotide differences between <span class="math">\(x\)</span> and the wildtype character <span class="math">\(\operatorname{wt}\left(r\right)\)</span>. Note that <span class="math">\(1 = \sum\limits_m \overline{\epsilon_m} = \sum\limits_m \overline{\rho_m}\)</span>.</p>
<p>Given these definitions, we define the prior vectors for the error rates as</p>
<div class="math" id="equation-arepsilon">
<span class="eqno">(13)</span>\[\begin{split}\mbox{$\boldsymbol{\mathbf{a_{r,\epsilon}}}$}&amp;=&amp; \left(\cdots, \sum\limits_m \frac{\overline{\epsilon}_{m}}{\mathcal{C}_m} \times \delta_{m,D_{x,\operatorname{wt}\left(r\right)}},\cdots\right),\end{split}\]</div>
<div class="math" id="equation-arrho">
<span class="eqno">(14)</span>\[\begin{split}\mbox{$\boldsymbol{\mathbf{a_{r,\rho}}}$}&amp;=&amp; \left(\cdots, \sum\limits_m \frac{\overline{\rho}_{m}}{\mathcal{C}_m} \times \delta_{m,D_{x,\operatorname{wt}\left(r\right)}}, \cdots\right)\end{split}\]</div>
<p>where <span class="math">\(\mathcal{C}_m\)</span> is the number of mutant characters with <span class="math">\(m\)</span> changes relative to the wildtype (so for nucleotides <span class="math">\(\mathcal{C}_0 = 1\)</span> and <span class="math">\(\mathcal{C}_1 = 3\)</span>, while for codons <span class="math">\(\mathcal{C}_0 = 1\)</span>, <span class="math">\(\mathcal{C}_1 = 9\)</span>, <span class="math">\(\mathcal{C}_2 = \mathcal{C}_3 = 27\)</span>) and <span class="math">\(\delta_{xy}\)</span> is again the <a class="reference external" href="http://en.wikipedia.org/wiki/Kronecker_delta">Kronecker delta</a>.</p>
<p>Our prior assumption is that the mutagenesis is done so that all mutant characters are introduced at equal frequency &#8211; note that this assumption is only true for codon characters if the mutagenesis is done at the codon level. In estimating the mutagenesis rate, we need to account for the facts that the observed counts of non-wildtype characters in the pre-selection mutant library will be inflated by the error rate represented by <span class="math">\(\overline{\epsilon_m}\)</span>. We therefore estimate the per-site mutagenesis rate as</p>
<div class="math" id="equation-avgmu">
<span class="eqno">(15)</span>\[\overline{\mu} = \left(\frac{1}{L}\sum\limits_r \frac{1}{\mbox{$N_r^{\textrm{pre}}$}}\sum\limits_{x\ne \operatorname{wt}\left(r\right)} \mbox{$n_{r,x}^{\textrm{pre}}$}\right) - \sum\limits_{m \ge 1} \overline{\epsilon_m}.\]</div>
<p>Note that a value of <span class="math">\(\overline{\mu} \le 0\)</span> would suggest that mutations are no more prevalent (or actually less prevalent) in the pre-selection mutant library then the pre-selection error control. Such a situation would violate the assumptions of the experiment, and so the algorithm will halt if this is the case. The prior vector for the mutagenesis rate is then</p>
<div class="math" id="equation-armu">
<span class="eqno">(16)</span>\[\begin{split}\mbox{$\boldsymbol{\mathbf{a_{r,\mu}}}$}&amp;=&amp; \left(\cdots, \frac{\overline{\mu}}{\mathcal{N}_x - 1} + \delta_{x,\operatorname{wt}\left(r\right)} \times \left[1 - \overline{\mu}\right] ,\cdots\right).\end{split}\]</div>
<div class="section" id="characters-are-nucleotides-preferences-are-for-nucleotides">
<span id="chartype-dna"></span><h3><a class="toc-backref" href="#id7">Characters are nucleotides; preferences are for nucleotides.</a><a class="headerlink" href="#characters-are-nucleotides-preferences-are-for-nucleotides" title="Permalink to this headline">¶</a></h3>
<p>One scenario is that the deep sequencing counts are for nucleotide
characters <span class="math">\(x\)</span>, and that we want to determine the preference
<span class="math">\(\pi_{r,x}\)</span> for each nucleotide at each site <span class="math">\(r\)</span>. This is
the most natural approach if the mutant library is generated by a
nucleotide-level mutagenesis process, such as error-prone PCR. Note that
in this case the priors correspond to the assumption that all single-nucleotide
mutations and errors are equally likely. In this scenario, we use the prior
vectors specified by Equations <a href="#equation-arepsilon">(13)</a>, <a href="#equation-arrho">(14)</a>, and <a href="#equation-armu">(16)</a>,
with the summations over <span class="math">\(x\)</span> covering the four nucleotide identities
(<tt class="docutils literal"><span class="pre">A</span></tt>, <tt class="docutils literal"><span class="pre">C</span></tt>, <tt class="docutils literal"><span class="pre">G</span></tt>, and <tt class="docutils literal"><span class="pre">T</span></tt>) and the summations over <span class="math">\(m\)</span> covering
the two possible number of nucleotide changes to a nucleotide character (0 and 1).</p>
</div>
<div class="section" id="characters-are-codons-preferences-are-for-codons">
<span id="chartype-codon"></span><h3><a class="toc-backref" href="#id8">Characters are codons; preferences are for codons.</a><a class="headerlink" href="#characters-are-codons-preferences-are-for-codons" title="Permalink to this headline">¶</a></h3>
<p>A second
scenario is that the deep sequencing counts are for codon
characters <span class="math">\(x\)</span>, and that we want to determine the preference
<span class="math">\(\pi_{r,x}\)</span> for each codon at each site <span class="math">\(r\)</span>. This is a
natural approach if the mutant library is generated by introducing
all possible codon mutations at each site, as is done by techniques
like the one described by <a class="reference external" href="http://www.plosone.org/article/info%3Adoi%2F10.1371%2Fjournal.pone.0052031">Firnberg2012</a>. The priors above based on the assumption
that all possible codon mutations are made at equal frequency, so
these priors are only appropriate if codons are mutagenized to
<tt class="docutils literal"><span class="pre">NNN</span></tt> (they would need to be adjusted if codons are only mutated to
a subset of possibilities such as <tt class="docutils literal"><span class="pre">NNK</span></tt>). In this scenario, we use the prior
vectors specified by Equations <a href="#equation-arepsilon">(13)</a>, <a href="#equation-arrho">(14)</a>, and <a href="#equation-armu">(16)</a>,
with the summations over <span class="math">\(x\)</span> covering the 64 codon identities
(<tt class="docutils literal"><span class="pre">AAA</span></tt>, <tt class="docutils literal"><span class="pre">AAC</span></tt>, <tt class="docutils literal"><span class="pre">AAG</span></tt>, <tt class="docutils literal"><span class="pre">AAT</span></tt>, <tt class="docutils literal"><span class="pre">ACA</span></tt>, etc) and the summations over <span class="math">\(m\)</span> covering
the four possible number of nucleotide changes to a codon character (0, 1, 2, and 3).</p>
</div>
<div class="section" id="characters-are-codons-preferences-are-for-amino-acids">
<span id="chartype-codon-to-aa"></span><h3><a class="toc-backref" href="#id9">Characters are codons; preferences are for amino acids.</a><a class="headerlink" href="#characters-are-codons-preferences-are-for-amino-acids" title="Permalink to this headline">¶</a></h3>
<p>A third
possibility is that the deep sequencing counts are for codon
characters <span class="math">\(x\)</span>, but that we want to determine the preferences
<span class="math">\(\pi_{r,a}\)</span> for amino-acid characters <span class="math">\(a\)</span>. This would be
the appropriate approach if the mutant library is generated by
introducing all possible codon mutations at each site and we are
assuming that all selection acts at the protein level (so all codons
for the same amino-acid should have the same preference). In this
case, we define the vector-valued function <span class="math">\(\mathbf{A}\)</span> to map
from codons to amino acids, so that</p>
<div class="math" id="equation-A">
<span class="eqno">(17)</span>\[\mathbf{A}\left(\mathbf{w}\right) = \left(\cdots, \sum\limits_x \delta_{a,\mathcal{A}\left(x\right)} \times w_x, \cdots\right)\]</div>
<p>where <span class="math">\(\mathbf{w}\)</span> is a 64-element vector giving the values for
each codon <span class="math">\(x\)</span>, <span class="math">\(\mathbf{A}\left(\mathbf{w}\right)\)</span> is a
20-element vector giving the values for each amino acid <span class="math">\(a\)</span> (or
a 21-element vector if stop codons are considered a possible
amino-acid identity), and <span class="math">\(\mathcal{A}\left(x\right)\)</span> is the
amino acid encoded by codon <span class="math">\(x\)</span>. The parameter vectors <span class="math">\(\boldsymbol{\mathbf{\pi_r}}\)</span>, <span class="math">\(\boldsymbol{\mathbf{\mu_r}}\)</span>, <span class="math">\(\boldsymbol{\mathbf{\epsilon_r}}\)</span>,
and <span class="math">\(\boldsymbol{\mathbf{\rho_r}}\)</span> are then all of length 20 (or 21) for the amino acid vectors.
The first of these vectors is still a symmetric <a class="reference external" href="http://en.wikipedia.org/wiki/Dirichlet_distribution">Dirichlet</a>, and the
priors for the remaining three are
<span class="math">\(\mathbf{A}\left(\mbox{$\boldsymbol{\mathbf{a_{r,\mu}}}$}\right)\)</span>,
<span class="math">\(\mathbf{A}\left(\mbox{$\boldsymbol{\mathbf{a_{r,\epsilon}}}$}\right)\)</span>,
and
<span class="math">\(\mathbf{A}\left(\mbox{$\boldsymbol{\mathbf{a_{r,\rho}}}$}\right)\)</span>
for the <span class="math">\(\boldsymbol{\mathbf{a_{r,\mu}}}\)</span>, <span class="math">\(\boldsymbol{\mathbf{a_{r,\epsilon}}}\)</span>, and <span class="math">\(\boldsymbol{\mathbf{a_{r,\rho}}}\)</span>&nbsp;vectors defined for codons immediately above. The
likelihoods are computed by similarly transforming the count vectors
in Equations <a href="#equation-pr_nrpre">(3)</a>, <a href="#equation-pr_nrpost">(6)</a>,
<a href="#equation-pr_nrerrpre">(4)</a>, and <a href="#equation-pr_nrerrpost">(5)</a> to
<span class="math">\(\mathbf{A}\left(\mbox{$\mathbf{n_r^{\textbf{pre}}}$}\right)\)</span>,
<span class="math">\(\mathbf{A}\left(\mbox{$\mathbf{n_r^{\textbf{post}}}$}\right)\)</span>,
<span class="math">\(\mathbf{A}\left(\mbox{$\mathbf{n_r^{\textbf{err,pre}}}$}\right)\)</span>,
and
<span class="math">\(\mathbf{A}\left(\mbox{$\mathbf{n_r^{\textbf{err,post}}}$}\right)\)</span>.</p>
</div>
<div class="section" id="characters-are-amino-acids-preferences-are-for-amino-acids">
<span id="chartype-aa"></span><h3><a class="toc-backref" href="#id10">Characters are amino acids; preferences are for amino acids.</a><a class="headerlink" href="#characters-are-amino-acids-preferences-are-for-amino-acids" title="Permalink to this headline">¶</a></h3>
<p>A fourth scenario is that the deep sequencing counts are for amino-acid
characters <span class="math">\(x\)</span>, and that we want to determine the preference
<span class="math">\(\pi_{r,x}\)</span> for each amino acid at each site <span class="math">\(r\)</span>. This scenario
would only be used if there has already been some post-processing of the deep mutational
scanning data, since the actual sequence data will always report either
nucleotides or codons. If at all possible, you should prefer to analyze the data
using the approach <a class="reference internal" href="#chartype-codon-to-aa"><em>Characters are codons; preferences are for amino acids.</em></a>. However, in some cases you may receive third-party
data that has already been processed to amino acids. In this scenario,
we use the prior assumption that all amino acid mutations are introduced at equal frequency,
and that all amino-acid errors happen at equal frequency. This first assumption will
only be true for certain mutagenesis strategies. The second assumption is very unlikely
to be true &#8211; but we can&#8217;t do better without analyzing the data at the codon level.
So in this scenario, we use the prior
vectors specified by Equations <a href="#equation-arepsilon">(13)</a>, <a href="#equation-arrho">(14)</a>, and <a href="#equation-armu">(16)</a>,
with the summations over <span class="math">\(x\)</span> covering the 20 amino acids (or 21 if stop codons
are allowed),
and the summations over <span class="math">\(m\)</span> covering
the two possible number of amino-acid changes to an amino-acid character (0 and 1).</p>
</div>
</div>
<div class="section" id="inferring-the-preferences-by-mcmc">
<span id="mcmc-inference"></span><h2><a class="toc-backref" href="#id11">Inferring the preferences by MCMC.</a><a class="headerlink" href="#inferring-the-preferences-by-mcmc" title="Permalink to this headline">¶</a></h2>
<p>Given the sequencing data and the parameters that specify the priors,
the posterior probability of any given set of the parameters <span class="math">\(\boldsymbol{\mathbf{\pi_r}}\)</span>, <span class="math">\(\boldsymbol{\mathbf{\mu_r}}\)</span>, <span class="math">\(\boldsymbol{\mathbf{\epsilon_r}}\)</span>,
and <span class="math">\(\boldsymbol{\mathbf{\rho_r}}\)</span>
is given by the product of Equations <a href="#equation-pr_nrpre">(3)</a>,
<a href="#equation-pr_nrpost">(6)</a>, <a href="#equation-pr_nrerrpre">(4)</a>,
<a href="#equation-pr_nrerrpost">(5)</a>, <a href="#equation-pr_pir">(7)</a>, <a href="#equation-pr_mur">(8)</a>,
<a href="#equation-pr_epsilonr">(9)</a>, and <a href="#equation-pr_rhor">(10)</a>. We use <a class="reference external" href="http://en.wikipedia.org/wiki/Markov_chain_Monte_Carlo">MCMC</a>
implemented by <a class="reference external" href="http://pystan.readthedocs.org/en/latest/">PyStan</a>
to sample from this posterior distribution. We
monitor for convergence of the sampling of the site-specific preferences
&nbsp;by running four chains, and ensuring that the mean over all <span class="math">\(\pi_{r,x}\)</span>
values of the potential scale
reduction statistic <span class="math">\(\hat{R}\)</span> of <a class="reference external" href="http://www.jstor.org/stable/2246093">GelmanRubin1992</a> is <span class="math">\(\le 1.1\)</span> and that the
mean effective sample size is <span class="math">\(\ge 100\)</span>, or that <span class="math">\(\hat{R} \le 1.15\)</span> and the effective sample size is <span class="math">\(\ge 300\)</span>. We repeatedly increase the number of steps
until convergence occurs. The inferred preferences are
summarized by the posterior mean and median-centered 95% credible
interval for each <span class="math">\(\pi_{r,x}\)</span>.</p>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Algorithm to infer site-specific preferences</a><ul>
<li><a class="reference internal" href="#definition-of-the-site-specific-preferences">Definition of the site-specific preferences.</a></li>
<li><a class="reference internal" href="#the-experimental-data-from-deep-sequencing">The experimental data from deep sequencing.</a></li>
<li><a class="reference internal" href="#assumption-that-mutations-and-errors-are-rare">Assumption that mutations and errors are rare.</a></li>
<li><a class="reference internal" href="#actual-frequencies-and-likelihoods-of-observing-experimental-data">Actual frequencies and likelihoods of observing experimental data.</a></li>
<li><a class="reference internal" href="#priors-over-the-unknown-parameters">Priors over the unknown parameters.</a><ul>
<li><a class="reference internal" href="#characters-are-nucleotides-preferences-are-for-nucleotides">Characters are nucleotides; preferences are for nucleotides.</a></li>
<li><a class="reference internal" href="#characters-are-codons-preferences-are-for-codons">Characters are codons; preferences are for codons.</a></li>
<li><a class="reference internal" href="#characters-are-codons-preferences-are-for-amino-acids">Characters are codons; preferences are for amino acids.</a></li>
<li><a class="reference internal" href="#characters-are-amino-acids-preferences-are-for-amino-acids">Characters are amino acids; preferences are for amino acids.</a></li>
</ul>
</li>
<li><a class="reference internal" href="#inferring-the-preferences-by-mcmc">Inferring the preferences by MCMC.</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="algorithms.html"
                        title="previous chapter">Algorithms</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="inferdiffprefs_algorithm.html"
                        title="next chapter">Algorithm to infer differential preferences</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/inferprefs_algorithm.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="inferdiffprefs_algorithm.html" title="Algorithm to infer differential preferences"
             >next</a> |</li>
        <li class="right" >
          <a href="algorithms.html" title="Algorithms"
             >previous</a> |</li>
        <li><a href="index.html">dms_tools 1.1.6 documentation</a> &raquo;</li>
          <li><a href="algorithms.html" >Algorithms</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2014, Jesse Bloom.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3a0.
    </div>
  </body>
</html>