<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Python API &mdash; dms_tools 1.1.6 documentation</title>
    
    <link rel="stylesheet" href="_static/default.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.1.6',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="dms_tools 1.1.6 documentation" href="index.html" />
    <link rel="next" title="Citations and Acknowledgments" href="citations.html" />
    <link rel="prev" title="Algorithm to infer differential preferences" href="inferdiffprefs_algorithm.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="citations.html" title="Citations and Acknowledgments"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="inferdiffprefs_algorithm.html" title="Algorithm to infer differential preferences"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">dms_tools 1.1.6 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="python-api">
<span id="pythonapi"></span><h1>Python API<a class="headerlink" href="#python-api" title="Permalink to this headline">¶</a></h1>
<p><a class="reference external" href="http://jbloom.github.io/dms_tools">dms_tools</a> is written in <a class="reference external" href="https://www.python.org">Python</a>, so if you install the package you can access the following API for the package via <a class="reference external" href="https://www.python.org">Python</a>.</p>
<span class="target" id="module-__init__"></span><div class="section" id="dms-tools-package">
<h2><tt class="docutils literal"><span class="pre">dms_tools</span></tt> package<a class="headerlink" href="#dms-tools-package" title="Permalink to this headline">¶</a></h2>
<p>This package contains utilities for analyzing deep mutational scanning (DMS) data.</p>
<p>The package webpage is <a class="reference external" href="https://github.com/jbloom/dms_tools">https://github.com/jbloom/dms_tools</a></p>
<p>The package is typically run via a series of command line scripts. However,
it also provides the following Python API.</p>
<div class="section" id="modules-in-the-package">
<h3>Modules in the package<a class="headerlink" href="#modules-in-the-package" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><em>mcmc</em> : Perform MCMC inference</li>
<li><em>parsearguments</em> : Parses command line arguments to scripts.</li>
<li><em>file_io</em> : File input / output in <tt class="docutils literal"><span class="pre">dms_tools</span></tt> formats.</li>
<li><em>plot</em> : make plots</li>
<li><em>weblogo</em> : use <tt class="docutils literal"><span class="pre">weblogolib</span></tt> to make logo plots.</li>
<li><em>utils</em> : assorted utilities for <tt class="docutils literal"><span class="pre">dms_tools</span></tt></li>
<li><em>simulate</em> : simulate data to test inferences.</li>
</ul>
</div>
<div class="section" id="package-level-variables">
<h3>Package-level variables<a class="headerlink" href="#package-level-variables" title="Permalink to this headline">¶</a></h3>
<p>The following variables are defined for this package:</p>
<ul class="simple">
<li><em>aminoacids</em> : list of the 20 amino acids in one-letter upper-case code.</li>
<li><em>aminoacids_nostop</em> : another name for <em>aminoacids</em>.</li>
<li><em>aminoacids_withstop</em> : like <em>aminoacids</em> but also includes the stop codon (<tt class="docutils literal"><span class="pre">*</span></tt>).</li>
<li><em>nts</em> : a list of the four upper-case DNA nucleotides.</li>
<li><em>codons</em> : a list of the 64 codons as upper-case three-letter strings.</li>
<li><em>codon_to_aa</em> : dictionary keyed by codons in <em>codons</em>, values are amino acids.</li>
<li><em>aa_to_codons</em> : dictionary keyed by all amino acids in <em>aminoacids</em>, values list of all encoding codons.</li>
</ul>
</div>
</div>
<span class="target" id="module-mcmc"></span><div class="section" id="mcmc-module">
<h2><tt class="docutils literal"><span class="pre">mcmc</span></tt> module<a class="headerlink" href="#mcmc-module" title="Permalink to this headline">¶</a></h2>
<p>Module that performs MCMC to infer preferences and differential preferences.</p>
<p>MCMC done using <tt class="docutils literal"><span class="pre">Stan</span></tt> (<a class="reference external" href="http://mc-stan.org/">http://mc-stan.org/</a>) via <tt class="docutils literal"><span class="pre">pystan</span></tt>.</p>
<p>Written by Jesse Bloom.</p>
<div class="section" id="functions-in-this-module">
<h3>Functions in this module<a class="headerlink" href="#functions-in-this-module" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><em>InferSitePreferences</em> : use MCMC to infer site-specific preferences.</li>
<li><em>InferSitePreferencesFromEnrichmentRatios</em> :infers site-specific preferences directly from enrichment ratios.</li>
<li><em>InferSiteDiffPreferencesFromEnrichmentRatios</em> :infers site-specific differential preferences directly from enrichment ratios.</li>
<li><em>InferSiteDiffPrefs</em> : use MCMC to infer site-specific differential preferences.</li>
<li><em>StanModelNoneErr</em> : <tt class="docutils literal"><span class="pre">pystan</span></tt> model used by <em>InferSitePreferences</em>.</li>
<li><em>StanModelSameErr</em> : <tt class="docutils literal"><span class="pre">pystan</span></tt> model used by <em>InferSitePreferences</em>.</li>
<li><em>StanModelDifferentErr</em> : <tt class="docutils literal"><span class="pre">pystan</span></tt> model used by <em>InferSitePreferences</em>.</li>
<li><em>StanModelDiffPrefNoErr</em> : <tt class="docutils literal"><span class="pre">pystan</span></tt> model used by <em>InferSiteDiffPrefs</em>.</li>
<li><em>StanModelDiffPrefSameErr</em> : <tt class="docutils literal"><span class="pre">pystan</span></tt> model used by <em>InferSiteDiffPrefs</em>.</li>
</ul>
</div>
<div class="section" id="function-documentation">
<h3>Function documentation<a class="headerlink" href="#function-documentation" title="Permalink to this headline">¶</a></h3>
<dl class="function">
<dt id="mcmc.InferSiteDiffPreferencesFromEnrichmentRatios">
<tt class="descclassname">mcmc.</tt><tt class="descname">InferSiteDiffPreferencesFromEnrichmentRatios</tt><big>(</big><em>characterlist</em>, <em>wtchar</em>, <em>error_model</em>, <em>counts</em>, <em>pseudocounts=1</em><big>)</big><a class="headerlink" href="#mcmc.InferSiteDiffPreferencesFromEnrichmentRatios" title="Permalink to this definition">¶</a></dt>
<dd><p>Infers site-specific differential preferences from enrichment ratios.</p>
<p>This function mirrors the operations performed by <em>InferSiteDiffPrefs</em>, expect the preferences
are calculated directly from enrichment ratios.</p>
<p><em>characterlist</em>, <em>wtchar</em>, <em>error_model</em>, and <em>counts</em> have the same meaning as for <em>InferSitePreferences</em>.</p>
<p><em>pseudocounts</em> is a number &gt; 0 giving the pseudocounts added to each count in <em>counts</em>.</p>
<p>The return value is: 
<em>(converged, deltapi_means, pr_deltapi_gt0, pr_deltapi_lt0, logstring)</em>, where the tuple
entries have the same meaning as for <em>InferSiteDiffPrefs</em> except that <em>pr_deltapi_gt0</em> 
and <em>pr_deltapi_lt0</em> are <em>None</em> since they cannot
be estimated from direct enrichment ratio calculation
as it is not a statistical model, and <em>converged</em> is <em>True</em> since this calculation
always converges.</p>
</dd></dl>

<dl class="function">
<dt id="mcmc.InferSiteDiffPrefs">
<tt class="descclassname">mcmc.</tt><tt class="descname">InferSiteDiffPrefs</tt><big>(</big><em>characterlist</em>, <em>wtchar</em>, <em>error_model</em>, <em>counts</em>, <em>priors</em>, <em>seed=1</em>, <em>niter=10000</em>, <em>n_jobs=1</em>, <em>r_max=1.1</em>, <em>neff_min=100</em>, <em>nchains=4</em>, <em>increasefactor=2</em>, <em>increasetries=6</em><big>)</big><a class="headerlink" href="#mcmc.InferSiteDiffPrefs" title="Permalink to this definition">¶</a></dt>
<dd><p>Infers site-specific differntial preferences by MCMC for a specific site.</p>
<p>The algorithm is explained in the documentation for the <tt class="docutils literal"><span class="pre">dms_tools</span></tt> package.</p>
<p>The MCMC automatically tries to guarantee convergence via the parameters specified
by <em>r_max</em>, <em>neff_min</em>, <em>nchains</em>, <em>niter</em>, <em>increasefactor</em>, and <em>increasetries</em>.</p>
<p>Here are the calling variables:</p>
<blockquote>
<div><ul>
<li><p class="first"><em>characterlist</em> is a list of all valid characters; entries must be unique.
Typically a list of all amino acids, codons, or nucleotides. Characters
are case sensitive.</p>
</li>
<li><p class="first"><em>wtchar</em> is a character that is in <em>characterlist</em> and defines the wildtype
character at the site.</p>
</li>
<li><p class="first"><em>error_model</em> specifies how the errors are estimated. It can
be one of the following strings:</p>
<blockquote>
<div><ul class="simple">
<li><em>none</em> : no errors</li>
<li><em>same</em> : same error rates for starting library and selections <span class="math">\(s1\)</span>
and <span class="math">\(s2\)</span>.</li>
</ul>
</div></blockquote>
<p>Alternatively, it could be an instance of <em>StanModelDiffPrefNoErr</em>
or <em>StanModelDiffPrefSameErr</em>. These are the
<tt class="docutils literal"><span class="pre">pystan</span></tt> models that implement the three strings described above.
If you pass these model objects, then they will not need to be
compiled by this function (strings require models to be compiled).
So it is advantageous to pass the <tt class="docutils literal"><span class="pre">pystan</span></tt> models rather than
the strings if calling this function repeatedly.</p>
</li>
<li><p class="first"><em>counts</em> is a dictionary specifying the deep sequencing counts.
Each string key should specify a dictionary keyed by all characters
and with values giving the integer counts for that character. Keys:</p>
<blockquote>
<div><ul class="simple">
<li><em>nrstart</em> : counts for starting library</li>
<li><em>nrs1</em> : counts for selection <span class="math">\(s1\)</span></li>
<li><em>nrs2</em> : counts for selection <span class="math">\(s2\)</span></li>
<li><em>nrerr</em> : only required if <em>error_model</em> is
<em>same</em>, specifis counts for error control.</li>
</ul>
</div></blockquote>
</li>
<li><p class="first"><em>priors</em> is a dictionary specifying the parameter vectors for the
Dirichlet priors. Each string key should specify a dictionary keyed by
all characters and with values giving the prior for that character. Keys:</p>
<blockquote>
<div><ul class="simple">
<li><em>pirs1_prior_params</em> : specifies Dirichlet prior vector for <span class="math">\(\pi_r^{s1}\)</span></li>
<li><em>frstart_prior_params</em> : specifies Dirichlet prior vector for 
<span class="math">\(f_r^{\textrm{start}}\)</span></li>
<li><em>deltapi_concentration</em> : specifies floating number giving concentration
parameter for <span class="math">\(\Delta\pi_r\)</span></li>
<li><em>xir_prior_params</em> : only required if <em>error_model</em> is <em>same</em>,
specified Dirichlet prior vector for error rates <span class="math">\(\xi_r\)</span></li>
</ul>
</div></blockquote>
<p>Any values less than <em>PRIOR_MIN_VALUE</em> (a constant specified in this module)
are automatically set to <em>PRIOR_MIN_VALUE</em>.</p>
</li>
<li><p class="first"><em>seed</em> : integer seed for MCMC. Calling multiple times with same <em>seed</em>
and data will give identical results provided architecture for floating
point math on the computer is also the same.</p>
</li>
<li><p class="first"><em>n_jobs</em> : The number of CPUs to use. If -1, use all available CPUs.</p>
</li>
<li><p class="first">The following parameters specify how the MCMC tries to guarantee convergence.
They all have reasonable default values, so you probably don&#8217;t need to change them.
The MCMC is considered to have converged if the mean Gelman-Rubin R
statistic (<a class="reference external" href="http://www.jstor.org/stable/2246093">http://www.jstor.org/stable/2246093</a>) over all <span class="math">\(\pi_{r,x}\)</span>
values is less than or equal to <em>r_max</em>, and the mean effective sample size
is greater than or equal to <em>neff_min</em>. The MCMC first runs with <em>nchains</em>
chains and <em>niter</em> iterations. If it fails to converge, it then increases
the number of iterations by a factor of <em>increasefactor</em> and tries again,
and repeats until it converges or until it has tried <em>increasetries</em>
times to increase the number of iterations.</p>
</li>
</ul>
</div></blockquote>
<p>The return value is: <em>(converged, deltapi_means, pr_deltapi_gt0, pr_deltapi_lt0, logstring)</em>.
The entries in this tuple have the following meanings:</p>
<blockquote>
<div><ul class="simple">
<li><em>converged</em> is <em>True</em> if the MCMC converges and <em>False</em> otherwise.</li>
<li><em>deltapi_means</em> is a dictionary keyed by all characters in <em>characterlist</em>
with the value for each character giving the <span class="math">\(\Delta\pi_{r,a}\)</span> value.</li>
<li><em>pr_deltapi_gt0</em> is a dictionary keyed by all characters in <em>characterlist</em>
with the value being the posterior probability that <span class="math">\(\Delta\pi_{r,a} &gt; 0\)</span>.</li>
<li><em>pr_deltapi_lt0</em> is a dictionary keyed by all characters in <em>characterlist</em>
with the value being the posterior probability that <span class="math">\(\Delta\pi_{r,a} &lt; 0\)</span>.</li>
<li><em>logstring</em> is a string that can be printed to give summary information
about the MCMC run and its convergence.</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="mcmc.InferSitePreferences">
<tt class="descclassname">mcmc.</tt><tt class="descname">InferSitePreferences</tt><big>(</big><em>characterlist</em>, <em>wtchar</em>, <em>error_model</em>, <em>counts</em>, <em>priors</em>, <em>seed=1</em>, <em>niter=10000</em>, <em>increasetries=6</em>, <em>n_jobs=1</em>, <em>r_max=1.1</em>, <em>neff_min=100</em>, <em>nchains=4</em>, <em>increasefactor=2</em><big>)</big><a class="headerlink" href="#mcmc.InferSitePreferences" title="Permalink to this definition">¶</a></dt>
<dd><p>Infers site-specific preferences by MCMC for a specific site.</p>
<p>Uses MCMC to infer the site-specific preferences <span class="math">\(\pi_{r,a}\)</span> for some site
<span class="math">\(r\)</span> for each character <span class="math">\(a\)</span> by integrating over the posterior defined by the product 
of the following priors and likelihoods, where for instance <span class="math">\(\boldsymbol{\mathbf{\pi_r}}\)</span>
is used to indicate the vector of <span class="math">\(\pi_{r,a}\)</span> values for all characters:</p>
<div class="math">
\[\Pr\left(\boldsymbol{\mathbf{\pi_r}}\right) = \mathrm{Dirichlet}\left(\boldsymbol{\mathbf{a_{\pi,r}}}\right)\]\[\Pr\left(\boldsymbol{\mathbf{\mu_r}}\right) = \mathrm{Dirichlet}\left(\boldsymbol{\mathbf{a_{\mu,r}}}\right)\]\[\Pr\left(\boldsymbol{\mathbf{\epsilon_r}}\right) = \mathrm{Dirichlet}\left(\boldsymbol{\mathbf{a_{\epsilon,r}}}\right)\]\[\Pr\left(\boldsymbol{\mathbf{\rho_r}}\right) = \mathrm{Dirichlet}\left(\boldsymbol{\mathbf{a_{\rho,r}}}\right)\]\[\Pr\left(\boldsymbol{\mathbf{n_r^{\rm{pre}}}} \mid \boldsymbol{\mathbf{\mu_r}}, \boldsymbol{\mathbf{\epsilon_r}}\right) = \mathrm{Multinomial}\left(\boldsymbol{\mathbf{n_r^{\rm{pre}}}}; \boldsymbol{\mathbf{\mu_r}} + \boldsymbol{\mathbf{\epsilon_r}} - \boldsymbol{\mathbf{\delta_r}}\right)\]\[\Pr\left(\boldsymbol{\mathbf{n_r^{\rm{post}}}} \mid \boldsymbol{\mathbf{\mu_r}}, \boldsymbol{\mathbf{\epsilon_r}}\right) = \mathrm{Multinomial}\left(\boldsymbol{\mathbf{n_r^{\rm{post}}}}; \frac{\boldsymbol{\mathbf{\mu_r}} \circ \boldsymbol{\mathbf{\pi_r}}}{\boldsymbol{\mathbf{\mu_r}} \cdot \boldsymbol{\mathbf{\pi_r}}} + \boldsymbol{\mathbf{\rho_r}} - \boldsymbol{\mathbf{\delta_r}}\right)\]\[\Pr\left(\boldsymbol{\mathbf{n_r^{\rm{err,pre}}}} \mid \boldsymbol{\mathbf{\epsilon_r}}\right) = \mathrm{Multinomial}\left(\boldsymbol{\mathbf{n_r^{\rm{err,pre}}}}; \boldsymbol{\mathbf{\epsilon_r}}\right)\]\[\Pr\left(\boldsymbol{\mathbf{n_r^{\rm{err,post}}}} \mid \boldsymbol{\mathbf{\rho_r}}\right) = \mathrm{Multinomial}\left(\boldsymbol{\mathbf{n_r^{\rm{err,post}}}}; \boldsymbol{\mathbf{\rho_r}}\right)\]</div>
<p>where <span class="math">\(\boldsymbol{\mathbf{\delta_r}}\)</span> is a vector that is zero except for a one at the element
corresponding to the wildtype character.</p>
<p>The MCMC automatically tries to guarantee convergence via the parameters specified
by <em>r_max</em>, <em>neff_min</em>, <em>nchains</em>, <em>niter</em>, <em>increasefactor</em>, and <em>increasetries</em>.</p>
<p>Here are the calling variables:</p>
<blockquote>
<div><ul>
<li><dl class="first docutils">
<dt><em>characterlist</em> is a list of all valid characters; entries must be unique. </dt>
<dd><p class="first last">Typically a list of all amino acids, codons, or nucleotides. Characters
are case sensitive.</p>
</dd>
</dl>
</li>
<li><p class="first"><em>wtchar</em> is a character that is in <em>characterlist</em> and defines the wildtype
character at the site.</p>
</li>
<li><p class="first"><em>error_model</em> specifies how the errors are estimated. It might
be one of the following strings:</p>
<blockquote>
<div><ul class="simple">
<li><em>none</em> : no errors (<span class="math">\(\boldsymbol{\mathbf{\epsilon_r}} = \boldsymbol{\mathbf{\rho_r}} = \mathbf{0}\)</span>)</li>
<li><em>same</em> : same error rates for pre and post-selection 
(<span class="math">\(\boldsymbol{\mathbf{\epsilon_r}} = \boldsymbol{\mathbf{\rho_r}}\)</span>).</li>
<li><em>different</em> : different error rates for pre and post
selection (<span class="math">\(\boldsymbol{\mathbf{\epsilon_r}} \ne \boldsymbol{\mathbf{\rho_r}}\)</span>).</li>
</ul>
</div></blockquote>
<p>Alternatively, it could be an instance of <em>StanModelNoneErr</em>, 
<em>StanModelSameErr</em>, or <em>StanModelDifferentErr</em>. These are the
<tt class="docutils literal"><span class="pre">pystan</span></tt> models that implement the three strings described above.
If you pass these model objects, then they will not need to be
compiled by this function (strings require models to be compiled).
So it is advantageous to pass the <tt class="docutils literal"><span class="pre">pystan</span></tt> models rather than
the strings if calling this function repeatedly.</p>
</li>
<li><p class="first"><em>counts</em> is a dictionary specifying the deep sequencing counts.
Each string key should specify a dictionary keyed by all characters
and with values giving the integer counts for that character. Keys:</p>
<blockquote>
<div><ul class="simple">
<li><em>nrpre</em> : specifies <span class="math">\(\boldsymbol{\mathbf{n_r^{\rm{pre}}}}\)</span></li>
<li><em>nrpost</em> : specifies <span class="math">\(\boldsymbol{\mathbf{n_r^{\rm{post}}}}\)</span></li>
<li><em>nrerr</em> : only required if <em>error_model</em> is <em>same</em>, specifies
<span class="math">\(\boldsymbol{\mathbf{n_r^{\rm{err,pre}}}} = \boldsymbol{\mathbf{n_r^{\rm{err,post}}}}\)</span></li>
<li><em>nrerrpre</em> and <em>nrerrpost</em> : only required if <em>error_model</em> is
<em>different</em>, specify <span class="math">\(\boldsymbol{\mathbf{n_r^{\rm{err,pre}}}}\)</span> and
<span class="math">\(\boldsymbol{\mathbf{n_r^{\rm{err,post}}}}\)</span>.</li>
</ul>
</div></blockquote>
</li>
<li><p class="first"><em>priors</em> is a dictionary specifying the parameter vectors for the
Dirichlet priors. Each string key should specify a dictionary keyed by
all characters and with values giving the prior for that character. Keys:</p>
<blockquote>
<div><ul class="simple">
<li><em>pir_prior_params</em> : specifies <span class="math">\(\boldsymbol{\mathbf{a_{\pi,r}}}\)</span></li>
<li><em>mur_prior_params</em> : specifies <span class="math">\(\boldsymbol{\mathbf{a_{\mu,r}}}\)</span></li>
<li><em>epsilonr_prior_params</em> : only required if <em>error_model</em> is <em>same</em>
or <em>different</em>, specifies <span class="math">\(\boldsymbol{\mathbf{a_{\epsilon,r}}}\)</span></li>
<li><em>rhor_prior_params</em> : only required if <em>error_model</em> is <em>different</em>,
specifies <span class="math">\(\boldsymbol{\mathbf{a_{\rho,r}}}\)</span></li>
</ul>
</div></blockquote>
<p>Any values less than <em>PRIOR_MIN_VALUE</em> (a constant specified in this module)
are automatically set to <em>PRIOR_MIN_VALUE</em>.</p>
</li>
<li><p class="first"><em>seed</em> : integer seed for MCMC. Calling multiple times with same <em>seed</em>
and data will give identical results provided architecture for floating
point math on the computer is also the same.</p>
</li>
<li><p class="first"><em>n_jobs</em> : The number of CPUs to use. If -1, use all available CPUs.</p>
</li>
<li><p class="first">The following parameters specify how the MCMC tries to guarantee convergence.
They all have reasonable default values, so you probably don&#8217;t need to change them.
The MCMC is considered to have converged if the mean Gelman-Rubin R
statistic (<a class="reference external" href="http://www.jstor.org/stable/2246093">http://www.jstor.org/stable/2246093</a>) over all <span class="math">\(\pi_{r,x}\)</span>
values is less than or equal to <em>r_max</em>, and the mean effective sample size
is greater than or equal to <em>neff_min</em>. The MCMC first runs with <em>nchains</em>
chains and <em>niter</em> iterations. If it fails to converge, it then increases
the number of iterations by a factor of <em>increasefactor</em> and tries again,
and repeats until it converges or until it has tried <em>increasetries</em>
times to increase the number of iterations. Additionally, if
the effective sample size exceeds 3 times <em>neff_min</em> then we allow
R to be <em>1 + 1.5 (r_min - 1)</em>.</p>
</li>
</ul>
</div></blockquote>
<p>The return value is: <em>(converged, pi_means, pi_95credint, logstring)</em>.
The entries in this tuple have the following meanings:</p>
<blockquote>
<div><ul class="simple">
<li><em>converged</em> is <em>True</em> if the MCMC converges and <em>False</em> otherwise.</li>
<li><em>pi_means</em> is a dictionary keyed by all characters in <em>characterlist</em>
with the value for each character giving the <span class="math">\(\pi_{r,a}\)</span> value.</li>
<li><em>pi_95credint</em> is a dictionary keyed by all characters in <em>characterlist</em>
with the value being a 2-tuple giving the lower and upper limits on the
median-centered credible interval for <span class="math">\(\pi_{r,a}\)</span>.</li>
<li><em>logstring</em> is a string that can be printed to give summary information
about the MCMC run and its convergence.</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="mcmc.InferSitePreferencesFromEnrichmentRatios">
<tt class="descclassname">mcmc.</tt><tt class="descname">InferSitePreferencesFromEnrichmentRatios</tt><big>(</big><em>characterlist</em>, <em>wtchar</em>, <em>error_model</em>, <em>counts</em>, <em>pseudocounts=1</em><big>)</big><a class="headerlink" href="#mcmc.InferSitePreferencesFromEnrichmentRatios" title="Permalink to this definition">¶</a></dt>
<dd><p>Infers site-specific preferences from enrichment ratios.</p>
<p>This function mirrors the operations performed by <em>InferSitePreferences</em>, expect the preferences
are calculated directly from enrichment ratios.</p>
<p><em>characterlist</em>, <em>wtchar</em>, <em>error_model</em>, and <em>counts</em> have the same meaning as for <em>InferSitePreferences</em>.</p>
<p><em>pseudocounts</em> is a number &gt; 0 giving the pseudocounts added to each count in <em>counts</em>.</p>
<p>Briefly, for each character <span class="math">\(x\)</span>, we calculate the enrichment relative to the wildtype
character <span class="math">\(\rm{wt}\)</span> at this site as</p>
<div class="math">
\[\phi_{x} = \frac{\left(\frac{\max\left(\mathcal{P}, n_x^{\rm{post}} - n_x^{\rm{err,post}} + \mathcal{P}\right)}{\max\left(\mathcal{P}, n_{\rm{wt}}^{\rm{post}} - n_{\rm{wt}}^{\rm{err,post}} + \mathcal{P}\right)}\right)}{\left(\frac{\max\left(\mathcal{P}, n_x^{\rm{pre}} - n_x^{\rm{err,pre}} + \mathcal{P}\right)}{\max\left(\mathcal{P}, n_{\rm{wt}}^{\rm{pre}} - n_{\rm{wt}}^{\rm{err,pre}} + \mathcal{P}\right)}\right)}\]</div>
<p>where <span class="math">\(\mathcal{P}\)</span> is the value of <em>pseudocounts</em>. When <em>error_model</em> is <em>none</em>, then 
<span class="math">\(0 = n_x^{\rm{post}} = n_x^{\rm{pre}}\)</span> and the above equation reduces to a simple ratio
of post- and pre-selection. With <span class="math">\(\mathcal{P} &gt; 0\)</span>, this enrichment is always <span class="math">\(&gt; 0\)</span> and 
<span class="math">\(&lt; \infty\)</span>.</p>
<p>We then calculate the preference as</p>
<div class="math">
\[\pi_x = \frac{\phi_x}{\sum_y \phi_y}.\]</div>
<p>The return value is: <em>(converged, pi_means, pi_95credint, logstring)</em>, where the tuple
entries have the same meaning as for <em>InferSitePreferences</em> except that <em>pi_95credint</em> is
<em>None</em> since no credible intervals can be estimated from direct enrichment ratio calculation
as it is not a statistical model, and <em>converged</em> is <em>True</em> since this calculation
always converges.</p>
</dd></dl>

<dl class="class">
<dt id="mcmc.StanModelDiffPrefNoErr">
<em class="property">class </em><tt class="descclassname">mcmc.</tt><tt class="descname">StanModelDiffPrefNoErr</tt><a class="headerlink" href="#mcmc.StanModelDiffPrefNoErr" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <tt class="xref py py-class docutils literal"><span class="pre">object</span></tt></p>
<p>PyStan model for <em>error_model</em> of &#8216;none&#8217; for <em>InferSiteDiffPrefs</em>.</p>
</dd></dl>

<dl class="class">
<dt id="mcmc.StanModelDiffPrefSameErr">
<em class="property">class </em><tt class="descclassname">mcmc.</tt><tt class="descname">StanModelDiffPrefSameErr</tt><a class="headerlink" href="#mcmc.StanModelDiffPrefSameErr" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <tt class="xref py py-class docutils literal"><span class="pre">object</span></tt></p>
<p>PyStan model for <em>error_model</em> of &#8216;same&#8217; for <em>InferSiteDiffPrefs</em>.</p>
</dd></dl>

<dl class="class">
<dt id="mcmc.StanModelDifferentErr">
<em class="property">class </em><tt class="descclassname">mcmc.</tt><tt class="descname">StanModelDifferentErr</tt><a class="headerlink" href="#mcmc.StanModelDifferentErr" title="Permalink to this definition">¶</a></dt>
<dd><p>PyStan model for <em>error_model</em> of &#8216;different&#8217; for <em>InferSitePreferences</em>.</p>
</dd></dl>

<dl class="class">
<dt id="mcmc.StanModelNoneErr">
<em class="property">class </em><tt class="descclassname">mcmc.</tt><tt class="descname">StanModelNoneErr</tt><a class="headerlink" href="#mcmc.StanModelNoneErr" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <tt class="xref py py-class docutils literal"><span class="pre">object</span></tt></p>
<p>PyStan model for <em>error_model</em> of &#8216;none&#8217; for <em>InferSitePreferences</em>.</p>
</dd></dl>

<dl class="class">
<dt id="mcmc.StanModelSameErr">
<em class="property">class </em><tt class="descclassname">mcmc.</tt><tt class="descname">StanModelSameErr</tt><a class="headerlink" href="#mcmc.StanModelSameErr" title="Permalink to this definition">¶</a></dt>
<dd><p>PyStan model for <em>error_model</em> of &#8216;same&#8217; for <em>InferSitePreferences</em>.</p>
</dd></dl>

</div>
</div>
<span class="target" id="module-file_io"></span><div class="section" id="file-io-module">
<h2><tt class="docutils literal"><span class="pre">file_io</span></tt> module<a class="headerlink" href="#file-io-module" title="Permalink to this headline">¶</a></h2>
<p>Module that reads and writes files.</p>
<p>Written by Jesse Bloom.</p>
<div class="section" id="id1">
<h3>Functions in this module<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><em>Versions</em> : Returns string giving versions of utilized packages.</li>
<li><em>ReadDMSCounts</em> : reads deep mutational scanning counts.</li>
<li><em>WriteDMSCounts</em> : writes deep mutational scanning counts.</li>
<li><em>ReadMultipleDMSCountsFiles</em> : reads multiple counts files for the same sequence.</li>
<li><em>WritePreferences</em> : writes site-specific preferences.</li>
<li><em>ReadPreferences</em> : reads the preferences file written by <em>WritePreferences</em>.</li>
<li><em>WriteDiffPrefs</em> : writes site-specific differential preferences.</li>
<li><em>ReadDiffPrefs</em> : reads differential preferences file written by <em>WriteDiffPrefs</em>.</li>
<li><em>ReadMultiPrefOrDiffPref</em> : reads multiple preferences or differential preferences files.</li>
<li><em>IteratePairedFASTQ</em> : iterate over paired-read FASTQ files.</li>
<li><em>ReadSummaryStats</em> : reads alignment summary statistics.</li>
</ul>
</div>
<div class="section" id="id2">
<h3>Function documentation<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h3>
<dl class="function">
<dt id="file_io.IteratePairedFASTQ">
<tt class="descclassname">file_io.</tt><tt class="descname">IteratePairedFASTQ</tt><big>(</big><em>r1files</em>, <em>r2files</em>, <em>gzipped</em>, <em>applyfilter</em>, <em>usegzip=False</em><big>)</big><a class="headerlink" href="#file_io.IteratePairedFASTQ" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterates over FASTQ file pairs for paired-end sequencing reads.</p>
<p>This function has been tested for its ability to read FASTQ files
generated by the Illumina Casava 1.8 pipeline for paired-end reads.</p>
<p>CALLING VARIABLES:</p>
<p><cite>r1files</cite> : list of file name(s) containing the R1 reads.</p>
<p><cite>r2files</cite> : list of file name(s) containing the R2 reads. The files
specified here must specify the exact same number of reads as
those in <cite>r1files</cite>, and they must be in the same sequence (i.e.
the first R1 read found going through <cite>r1files</cite> in order must
match the first R2 read found going through <cite>r2files</cite> in order,
etc).</p>
<p><cite>gzipped</cite> : Boolean switch specifying whether the FASTQ files
specified by <cite>r1files</cite> and <cite>r2files</cite> are gzipped. If set to <cite>True</cite>,
they are gzipped. This iterator will then read them without
unzipping the files.</p>
<p><cite>applyfilter</cite> : Boolean switch specifying whether we remove read
pairs in which one or more of the reads failed the Illumina
chastity filter. If <cite>True</cite>, the iterator simply returns <cite>False</cite>
each time it encounters a read pair where the chastity filter is
failed. The chastity filter is indicated by Y or N in the
sequence header &#8211; values of Y indicate a failed filter.</p>
<p><cite>usegzip</cite> : Boolean switch that is meaningful only if <cite>gzipped</cite>
is <cite>True</cite>. This specifies that we use the Python <tt class="docutils literal"><span class="pre">gzip</span></tt> module to 
unzip the reads. Otherwise we use the system <tt class="docutils literal"><span class="pre">gzip</span> <span class="pre">-cd</span> <span class="pre">command</span></tt>,
which may be faster. You should definitely specify
<cite>usegzip</cite> to <cite>True</cite> if you are only reading a few lines, as with
<cite>usegzip</cite> set to <cite>False</cite> the entire file is unzipped before anything happens.
<cite>False</cite> by default.</p>
<p>RESULT OF THIS FUNCTION:</p>
<p>The function will iterate over all read pairs in the indicated
files. For each iteration, it will return:</p>
<ul class="simple">
<li>The return variable is <cite>False</cite> if <tt class="docutils literal"><span class="pre">applyfilter</span> <span class="pre">==</span> <span class="pre">True</span></tt> and one of the
reads in the pair failed the Illumina chastity filter.</li>
<li>Otherwise return variable is the following tuple:
<cite>(name, r1, r2, q1, q2)</cite> where:<ul>
<li><cite>name</cite> : name of the read pair (a string)</li>
<li><cite>r1</cite> : sequence of the first read R1 (a string)</li>
<li><cite>r2</cite> : sequence of the second read R2 (a string)</li>
<li><cite>q1</cite> : string of Q-scores for the R1 read</li>
<li><cite>q2</cite> : string of Q-scores for the R2 read</li>
</ul>
</li>
</ul>
</dd></dl>

<dl class="function">
<dt id="file_io.ReadDMSCounts">
<tt class="descclassname">file_io.</tt><tt class="descname">ReadDMSCounts</tt><big>(</big><em>f</em>, <em>chartype</em><big>)</big><a class="headerlink" href="#file_io.ReadDMSCounts" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads deep mutational scanning counts.</p>
<p><em>f</em> is a readable file-like object that contains the counts, or 
a string that gives the name of a file.</p>
<p><em>chartype</em> is the type of character. Valid values are the following strings:</p>
<blockquote>
<div><ul class="simple">
<li><em>DNA</em> : DNA nucleotides, those listed in the <em>nts</em> variable
defined in this module.</li>
<li><em>codon</em> : DNA codons, those listed in the <em>codons</em> variable
defined in this module.</li>
<li><em>aminoacids_nostop</em> : amino acids, not including stop codons.</li>
<li><em>aminoacids_withstop</em> : amino acids, including stop codons (<tt class="docutils literal"><span class="pre">*</span></tt>).</li>
</ul>
</div></blockquote>
<p>The counts are returned in the dictionary <em>counts</em>. The dictionary
is keyed by <strong>strings</strong> giving the position (e.g. &#8216;1&#8217;, &#8216;2&#8217;, or &#8216;5A&#8217;).
For each position <em>r</em>, <em>strings[r]</em> is a dictionary keyed by the
string <em>WT</em> and all characters (nucleotides or codons) as specified
by <em>chartype</em>, in upper case. <em>counts[r][&#8216;WT&#8217;]</em> is the wildtype identity
at the site; <em>counts[r][x]</em> is the number of counts for character <em>x</em>.</p>
<p>The specifications on the format of the contents of <em>f</em> are as follows:</p>
<blockquote>
<div><ul>
<li><p class="first">Any lines beginning with <em>#</em> are treated as comments or headers, 
and do not contain data.</p>
</li>
<li><p class="first">Columns are space delimited. They can separated by one or more 
spaces, or by one or more tabs.</p>
</li>
<li><p class="first">Before the first data line (a line <strong>not</strong> beginning with <em>#</em>) 
there must be a header line beginning with <em># POSITION WT</em> and 
then followed in arbitrary order by entries specifying valid 
DNA nucleotides or valid DNA codons. The order of the entries 
in this header line establish the order of the data in subsequent lines.</p>
</li>
<li><p class="first">The case does matter, all nucleotides and codons <strong>must be upper case</strong>.</p>
</li>
<li><p class="first">It is OK to intersperse columns that do not represent valid codons 
or nucleotides as long as all of the codons or nucleotides are present. 
This might be useful if you file is using <em>N</em> to denote ambiguous 
nucleotides (which are ignored as not being definitive counts), or 
is using lower-case letters (such as <em>a</em> to denote unsure calls).</p>
</li>
<li><p class="first">For every data line, the first two entries will be <em>POSITION</em> and <em>WT</em>:</p>
<blockquote>
<div><ul>
<li><p class="first"><em>POSITION</em> is the position number. These do not have to be 
consecutive integers, so it is OK to skip sites (e.g. <em>1</em>, <em>2</em>, <em>5</em>) 
or have negative site numbers (e.g. <em>-1</em>, <em>0</em>, <em>1</em>) or have sites 
with letter suffixes (e.g. <em>1</em>, <em>2</em>, <em>2A</em>, <em>3</em>). All of these 
options are useful when referring to numbering schemes that are 
non-sequential, as is sometimes the case in gene numbering. Note 
however that each position number must be unique.</p>
</li>
<li><dl class="first docutils">
<dt><em>WT</em> must give a valid wildtype identity of the site as a </dt>
<dd><p class="first last">nucleotide or codon.</p>
</dd>
</dl>
</li>
</ul>
</div></blockquote>
</li>
</ul>
</div></blockquote>
<p>Example of parsing nucleotide counts files with minimal required format and
extra lines / columns.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">f</span> <span class="o">=</span> <span class="n">cStringIO</span><span class="o">.</span><span class="n">StringIO</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;# POSITION WT A T C G</span><span class="se">\n</span><span class="s">&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;1 A 310818 13 0 37</span><span class="se">\n</span><span class="s">&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;2 T 16 311782 37 4</span><span class="se">\n</span><span class="s">&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;2A G 27 27 11 312520</span><span class="se">\n</span><span class="s">&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;3 A 313647 13 4 31</span><span class="se">\n</span><span class="s">&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">.</span><span class="n">seek</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f2</span> <span class="o">=</span> <span class="n">cStringIO</span><span class="o">.</span><span class="n">StringIO</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f2</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;# extra comment line</span><span class="se">\n</span><span class="s">&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f2</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;# POSITION WT N A T G C a</span><span class="se">\n</span><span class="s">&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f2</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;1 A 1 310818  13 37 0 15</span><span class="se">\n</span><span class="s">&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f2</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;2 T 15 16 311782   4 37</span><span class="se">\t</span><span class="s"> 0</span><span class="se">\n</span><span class="s">&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f2</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;# another comment line</span><span class="se">\n</span><span class="s">&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f2</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;2A G 7 27 27 312520 11 1</span><span class="se">\n</span><span class="s">&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f2</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;3 A 0 313647 13 31 4 2</span><span class="se">\n</span><span class="s">&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f2</span><span class="o">.</span><span class="n">seek</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">ReadDMSCounts</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="s">&#39;DNA&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d2</span> <span class="o">=</span> <span class="n">ReadDMSCounts</span><span class="p">(</span><span class="n">f2</span><span class="p">,</span> <span class="s">&#39;DNA&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">==</span> <span class="n">d2</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="p">[</span><span class="s">&#39;1&#39;</span><span class="p">][</span><span class="s">&#39;WT&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s">&#39;A&#39;</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="p">[</span><span class="s">&#39;2A&#39;</span><span class="p">][</span><span class="s">&#39;WT&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s">&#39;G&#39;</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="p">[</span><span class="s">&#39;2&#39;</span><span class="p">][</span><span class="s">&#39;T&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">311782</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="p">[</span><span class="s">&#39;2&#39;</span><span class="p">][</span><span class="s">&#39;C&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">37</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="mi">1</span> <span class="ow">in</span> <span class="n">d</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="s">&#39;N&#39;</span> <span class="ow">in</span> <span class="n">d2</span><span class="p">[</span><span class="s">&#39;1&#39;</span><span class="p">]</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="s">&#39;a&#39;</span> <span class="ow">in</span> <span class="n">d2</span><span class="p">[</span><span class="s">&#39;2&#39;</span><span class="p">]</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">.</span><span class="n">seek</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ReadDMSCounts</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="s">&#39;codon&#39;</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
    <span class="o">...</span>
<span class="gr">ValueError</span>: <span class="n">file does not have a valid header for codon</span>
</pre></div>
</div>
<p>Example for codons</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">nts</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">codons</span> <span class="o">=</span> <span class="p">[]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">nt1</span> <span class="ow">in</span> <span class="n">nts</span><span class="p">:</span>
<span class="gp">... </span>    <span class="k">for</span> <span class="n">nt2</span> <span class="ow">in</span> <span class="n">nts</span><span class="p">:</span>
<span class="gp">... </span>        <span class="k">for</span> <span class="n">nt3</span> <span class="ow">in</span> <span class="n">nts</span><span class="p">:</span>
<span class="gp">... </span>            <span class="n">codons</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s">&#39;</span><span class="si">%s%s%s</span><span class="s">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">nt1</span><span class="p">,</span> <span class="n">nt2</span><span class="p">,</span> <span class="n">nt3</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span> <span class="o">=</span> <span class="n">cStringIO</span><span class="o">.</span><span class="n">StringIO</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;# POSITION WT </span><span class="si">%s</span><span class="se">\n</span><span class="s">&#39;</span> <span class="o">%</span> <span class="s">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">codons</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;1 AAA </span><span class="si">%s</span><span class="se">\n</span><span class="s">&#39;</span> <span class="o">%</span> <span class="s">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">codons</span><span class="p">))]))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">.</span><span class="n">seek</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">ReadDMSCounts</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="s">&#39;codon&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="p">[</span><span class="s">&#39;1&#39;</span><span class="p">][</span><span class="s">&#39;AAC&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="p">[</span><span class="s">&#39;1&#39;</span><span class="p">][</span><span class="s">&#39;WT&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s">&#39;AAA&#39;</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="p">[</span><span class="s">&#39;1&#39;</span><span class="p">][</span><span class="s">&#39;WT&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s">&#39;AAa&#39;</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">codons</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span> <span class="o">=</span> <span class="n">codons</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f2</span> <span class="o">=</span> <span class="n">cStringIO</span><span class="o">.</span><span class="n">StringIO</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f2</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;# POSITION WT </span><span class="si">%s</span><span class="se">\n</span><span class="s">&#39;</span> <span class="o">%</span> <span class="s">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">codons</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f2</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;1 AAA </span><span class="si">%s</span><span class="se">\n</span><span class="s">&#39;</span> <span class="o">%</span> <span class="s">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">codons</span><span class="p">))]))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f2</span><span class="o">.</span><span class="n">seek</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">ReadDMSCounts</span><span class="p">(</span><span class="n">f2</span><span class="p">,</span> <span class="s">&#39;codon&#39;</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
    <span class="o">...</span>
<span class="gr">ValueError</span>: <span class="n">file does not have a valid header for codon</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="file_io.ReadDiffPrefs">
<tt class="descclassname">file_io.</tt><tt class="descname">ReadDiffPrefs</tt><big>(</big><em>f</em><big>)</big><a class="headerlink" href="#file_io.ReadDiffPrefs" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads the differential preferences written by <em>WriteDiffPrefs</em>.</p>
<p><em>f</em> is the name of an existing file or a readable file-like object.</p>
<p>The return value is the tuple: 
<em>(sites, wts, deltapi_means, pr_deltapi_lt0, pr_deltapi_gt0, rms)</em>
where <em>sites</em>, <em>wts</em>, <em>deltapi_means</em>, <em>pr_deltapi_lt0</em>, and 
<em>pr_deltapi_gt0</em> all have the same values used to write the file
with <em>WriteDiffPrefs</em> and <em>rms</em> is a dictionary with <em>rms[r]</em> giving
the root-mean-square differential preference for each <em>r</em> in 
<em>sites</em>.</p>
<p>See docstring of <em>WriteDiffPrefs</em> for example usage.</p>
</dd></dl>

<dl class="function">
<dt id="file_io.ReadMultiPrefOrDiffPref">
<tt class="descclassname">file_io.</tt><tt class="descname">ReadMultiPrefOrDiffPref</tt><big>(</big><em>infiles</em>, <em>removestop=False</em><big>)</big><a class="headerlink" href="#file_io.ReadMultiPrefOrDiffPref" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads multiple preferences or differential preferences files.</p>
<p><em>infiles</em> is a list of preferences or differential preferences files in
the formats read by <em>ReadPreferences</em> or <em>ReadDiffPrefs</em>. The files can
be a mix of these two formats. Each of the files must have data for the
same set of sites or a <em>ValueError</em> is raised. In addition, the
files must use the same sets of characters or a <em>ValueError</em> is raised.
Valid character sets are:</p>
<blockquote>
<div><ul class="simple">
<li>nucleotides</li>
<li>codons</li>
<li>amino acids without stop codons</li>
<li>amino acids with stop codons</li>
<li>a mix of amino acids with and without stop codons if <em>removestop</em>
is <em>True</em>.</li>
</ul>
</div></blockquote>
<p><em>removestop</em> specifies whether we remove stop codons from amino acids. If
this is done, all preferences are renormalized to sum to one by scaling
them, and all differential preferences are renormalized to sum to zero
by adding or subtracting a fixed amount to each.</p>
<p>The return value is as follows: 
<em>(sites, characters, wts, databyfile)</em>. In this tuple:</p>
<blockquote>
<div><p><em>sites</em> is a list of all site numbers (as strings) for which
we have preferences or differential preferences.</p>
<p><em>characters</em> is a list of the characters that are being used
(i.e. nucleotides, codons, or amino acids with or without
stop codons).</p>
<p><em>wts</em> is a dictionary with <em>wts[r]</em> giving the wildtype character
at site <em>r</em> for all <em>r</em> in <em>sites</em>. If the files do not all have
the same wildtype at <em>r</em>, then <em>wts[r] = &#8216;?&#8217;</em>.</p>
<p><em>databyfile</em> is a dictionary keyed by each file name in <em>infiles</em>
If that file is a preferences file, then the value is the dictionary
<em>pi_means</em> that would be returned by <em>ReadPreferences</em>. If that file
is a differential preferences file, then the value is the 
dictionary <em>deltapi_means</em> that would be returned by <em>ReadDiffPrefs</em>.</p>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="file_io.ReadMultipleDMSCountsFiles">
<tt class="descclassname">file_io.</tt><tt class="descname">ReadMultipleDMSCountsFiles</tt><big>(</big><em>filelist</em>, <em>chartype</em><big>)</big><a class="headerlink" href="#file_io.ReadMultipleDMSCountsFiles" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads multiple deep mutational scanning counts for the same sequence.</p>
<p><em>filelist</em> is a list of names of files that can be read by <em>ReadDMSCounts</em>
using <em>chartype</em>.</p>
<p>This function reads the counts for all files in <em>filelist</em>. It confirms
that each file specifies counts for the same set of sites, and that
the wildtype sequence specified in each file is the same. It then returns
the following tuple: <em>(sites, wts, counts)</em>. In this tuple:</p>
<blockquote>
<div><ul class="simple">
<li><em>sites</em> is a list of strings giving the sites, sorted so that the
numeric portion of the string is interpreted as a number.</li>
<li><em>wts</em> is a dictionary, with <em>wts[r]</em> giving the wildtype character
for site <em>r</em> for each <em>r</em> in <em>sites</em>.</li>
<li><em>counts</em> is a dictionary keyed by each file in <em>filelist</em>, with the value
for each file name being the counts dictonary for that file in the format
returned by <em>ReadDMSCounts</em>.</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="file_io.ReadPreferences">
<tt class="descclassname">file_io.</tt><tt class="descname">ReadPreferences</tt><big>(</big><em>f</em><big>)</big><a class="headerlink" href="#file_io.ReadPreferences" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads the site-specific preferences written by <em>WritePreferences</em>.</p>
<p><em>f</em> is the name of an existing file or a readable file-like object.</p>
<p>The return value is the tuple: <em>(sites, wts, pi_means, pi_95credint, h)</em>
where <em>sites</em>, <em>wts</em>, <em>pi_means</em>, and <em>pi_95credint</em> will all
have the same values used to write the file with <em>WritePreferences</em>,
and <em>h</em> is a dictionary with <em>h[r]</em> giving the site entropy (log base
2) for each <em>r</em> in <em>sites</em>.</p>
<p>See docstring of <em>WritePreferences</em> for example usage.</p>
</dd></dl>

<dl class="function">
<dt id="file_io.ReadSummaryStats">
<tt class="descclassname">file_io.</tt><tt class="descname">ReadSummaryStats</tt><big>(</big><em>f</em><big>)</big><a class="headerlink" href="#file_io.ReadSummaryStats" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads alignment summary statistics.</p>
<p><em>f</em> should be either a readable file-like object or a string giving
the name of an existing file that contains alignment summary statistics.
For instance, this might be the <tt class="docutils literal"><span class="pre">summarystats.txt</span></tt> file produced by
<tt class="docutils literal"><span class="pre">dms_barcodedsubamplicons</span></tt>. Here is example content from such
a file:</p>
<div class="highlight-python"><div class="highlight"><pre>total read pairs = 4580258
read pairs that fail Illumina filter = 0
low quality read pairs = 915131
barcodes randomly purged = 0
discarded barcodes with 1 reads = 844768
aligned barcodes with 2 reads = 466387
discarded barcodes with 2 reads = 6457
un-alignable barcodes with 2 reads = 40753
aligned barcodes with 3 reads = 262149
discarded barcodes with 3 reads = 5426
un-alignable barcodes with 3 reads = 24891
aligned barcodes with 4 reads = 120828
discarded barcodes with 4 reads = 3281
un-alignable barcodes with 4 reads = 7675
aligned barcodes with 5 reads = 45342
discarded barcodes with 5 reads = 1516
un-alignable barcodes with 5 reads = 2863
</pre></div>
</div>
<p>The return value is the 2-tuple <em>(keylist, stats)</em>.
<em>keylist</em> is a list of all of the keys in <em>f</em> (these
are the strings before the equals sign) in the same
order as they appear in the file. <em>stats</em> is a dictionary
keyed by each key in <em>keylist</em> and with the value being
an integer giving the number of counts for that key.
The reason for returning <em>keylist</em> as well as <em>stats</em>
is that <em>keylist</em> preserves the order of the entries in <em>f</em>,
whereas the dictionary <em>stats</em> will have an arbitrary order.</p>
<p>Any lines in <em>f</em> that begin with <tt class="docutils literal"><span class="pre">#</span></tt> are treated
as comments and are not processed into the return 
variables.</p>
</dd></dl>

<dl class="function">
<dt id="file_io.Versions">
<tt class="descclassname">file_io.</tt><tt class="descname">Versions</tt><big>(</big><big>)</big><a class="headerlink" href="#file_io.Versions" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a string with version information.</p>
<p>You would call this function if you want a string giving detailed information
on the version of <tt class="docutils literal"><span class="pre">dms_tools</span></tt> and the associated packages that it uses.
This string might be useful if you want to print this information at the beginning
of a script.</p>
</dd></dl>

<dl class="function">
<dt id="file_io.WriteDMSCounts">
<tt class="descclassname">file_io.</tt><tt class="descname">WriteDMSCounts</tt><big>(</big><em>f</em>, <em>counts</em><big>)</big><a class="headerlink" href="#file_io.WriteDMSCounts" title="Permalink to this definition">¶</a></dt>
<dd><p>Writes deep mutational scanning counts file.</p>
<p>This function is the inverse of <em>ReadDMSCounts</em>.</p>
<p><em>f</em> is a writable file-like object, or a string that gives the name of
a file that is created.</p>
<p><em>counts</em> is a dictionary of the same format returned by <em>ReadDMSCounts</em>.
These counts are written to <em>f</em>.</p>
<p>There is no <em>chartype</em> argument &#8211; that is automatically determined from
<em>counts</em>.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;1&#39;</span><span class="p">:{</span><span class="s">&#39;A&#39;</span><span class="p">:</span><span class="mi">2</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">:</span><span class="mi">3</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">:</span><span class="mi">4</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">:</span><span class="mi">1</span><span class="p">,</span> <span class="s">&#39;WT&#39;</span><span class="p">:</span><span class="s">&#39;A&#39;</span><span class="p">},</span> <span class="s">&#39;2&#39;</span><span class="p">:{</span><span class="s">&#39;A&#39;</span><span class="p">:</span><span class="mi">1</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">:</span><span class="mi">1</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">:</span><span class="mi">3</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">:</span><span class="mi">4</span><span class="p">,</span> <span class="s">&#39;WT&#39;</span><span class="p">:</span><span class="s">&#39;C&#39;</span><span class="p">}}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span> <span class="o">=</span> <span class="n">cStringIO</span><span class="o">.</span><span class="n">StringIO</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">WriteDMSCounts</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="n">counts</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">.</span><span class="n">seek</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">readcounts</span> <span class="o">=</span> <span class="n">ReadDMSCounts</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="s">&#39;DNA&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">==</span> <span class="n">readcounts</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="file_io.WriteDiffPrefs">
<tt class="descclassname">file_io.</tt><tt class="descname">WriteDiffPrefs</tt><big>(</big><em>f</em>, <em>sites</em>, <em>wts</em>, <em>deltapi_means</em>, <em>pr_deltapi_lt0</em>, <em>pr_deltapi_gt0</em><big>)</big><a class="headerlink" href="#file_io.WriteDiffPrefs" title="Permalink to this definition">¶</a></dt>
<dd><p>Writes differential preferences to a file.</p>
<dl class="docutils">
<dt><em>f</em> is a writeable file-like object to which we write the differential</dt>
<dd>preferences, or a string giving the name of a file that we create.</dd>
<dt><em>sites</em> is a list of all site numbers (as strings) for which</dt>
<dd>we have preferences.</dd>
<dt><em>wts</em> is a dictionary with <em>wts[r]</em> giving the wildtype character</dt>
<dd>at site <em>r</em> for all <em>r</em> in <em>sites</em>.</dd>
<dt><em>deltapi_means</em> is a dictionary keyed by all sites (as strings). For each </dt>
<dd>site <em>r</em>, <em>deltapi_means[r]</em> is a dictionary where <em>deltapi_means[r][x]</em>
is the preference of <em>r</em> for character <em>x</em>, where the
characters are nucleotide, codons, or amino acids.</dd>
<dt><em>pr_deltapi_gt0</em> and <em>pr_deltapi_lt0</em> can either both be <em>None</em>,</dt>
<dd>or can both be dictionaries keyed by all sites in <em>deltapi_means</em>
that have as their values the posterior probability that particular
differential preference is greater than or less than zero, 
respectively.</dd>
</dl>
<p>The written file has the following format:</p>
<div class="highlight-python"><div class="highlight"><pre># POSITION WT RMS_dPI dPI_A dPI_C dPI_G dPI_T Pr_dPI_A_lt0 Pr_dPI_A_gt0 Pr_dPI_C_lt0 Pr_dPI_C_gt0 Pr_dPI_G_lt0 Pr_dPI_G_gt0 Pr_dPI_T_lt0 Pr_dPI_T_gt0
1 G 0.241 0.02 -0.4 0.2 0.18 0.41 0.59 0.90 0.10 0.2 0.8 0.3 0.7
</pre></div>
</div>
<p>In this format, the first header line begins with <tt class="docutils literal"><span class="pre">#</span></tt> and indicates the
characters as the suffix to <tt class="docutils literal"><span class="pre">dPI_</span></tt>. The posterior probabilities
are indicated by columns of the form <tt class="docutils literal"><span class="pre">Pr_dPI_A_gt0</span></tt> for greater than zero,
and <tt class="docutils literal"><span class="pre">Pr_dPI_A_lt0</span></tt> for less than zero.
The <tt class="docutils literal"><span class="pre">RMS_dPI</span></tt> column gives the root-mean-square differential preference
as <span class="math">\(\textrm{RMS}_{\Delta\pi_r} = \sqrt{\frac{1}{\mathcal{N}_x}\sum_x \left(\Delta\pi_{r,x}\right)^2}\)</span></p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">sites</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;1&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">deltapi_means</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;1&#39;</span><span class="p">:{</span><span class="s">&#39;A&#39;</span><span class="p">:</span><span class="o">-</span><span class="mf">0.2</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">:</span><span class="o">-</span><span class="mf">0.3</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">:</span><span class="mf">0.4</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">:</span><span class="mf">0.1</span><span class="p">}}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">wts</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;1&#39;</span><span class="p">:</span><span class="s">&#39;G&#39;</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span> <span class="o">=</span> <span class="n">cStringIO</span><span class="o">.</span><span class="n">StringIO</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">WriteDiffPrefs</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="n">sites</span><span class="p">,</span> <span class="n">wts</span><span class="p">,</span> <span class="n">deltapi_means</span><span class="p">,</span> <span class="bp">None</span><span class="p">,</span> <span class="bp">None</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">.</span><span class="n">seek</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lines</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lines</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s">&#39;# POSITION WT RMS_dPI dPI_A dPI_C dPI_G dPI_T</span><span class="se">\n</span><span class="s">&#39;</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">entries</span> <span class="o">=</span> <span class="n">lines</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">entries</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="nb">str</span><span class="p">(</span><span class="n">sites</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">entries</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="n">wts</span><span class="p">[</span><span class="n">sites</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="mf">1.0e-6</span> <span class="o">&gt;</span> <span class="nb">abs</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">entries</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span> <span class="o">-</span> <span class="n">math</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="nb">sum</span><span class="p">([</span><span class="n">x</span><span class="o">**</span><span class="mi">2</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">deltapi_means</span><span class="p">[</span><span class="n">sites</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span><span class="o">.</span><span class="n">values</span><span class="p">()])</span> <span class="o">/</span> <span class="nb">float</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">deltapi_means</span><span class="p">[</span><span class="n">sites</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span><span class="o">.</span><span class="n">values</span><span class="p">()))))</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="mf">1.0e-6</span> <span class="o">&gt;</span> <span class="nb">abs</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">entries</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span> <span class="o">-</span> <span class="n">deltapi_means</span><span class="p">[</span><span class="n">sites</span><span class="p">[</span><span class="mi">0</span><span class="p">]][</span><span class="s">&#39;A&#39;</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f2</span> <span class="o">=</span> <span class="n">cStringIO</span><span class="o">.</span><span class="n">StringIO</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pr_deltapi_gt0</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;1&#39;</span><span class="p">:{</span><span class="s">&#39;A&#39;</span><span class="p">:</span><span class="mf">0.95</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">:</span><span class="mf">0.5</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">:</span><span class="mf">0.7</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">:</span><span class="mf">0.5</span><span class="p">}}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pr_deltapi_lt0</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;1&#39;</span><span class="p">:{</span><span class="s">&#39;A&#39;</span><span class="p">:</span><span class="mf">0.05</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">:</span><span class="mf">0.5</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">:</span><span class="mf">0.3</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">:</span><span class="mf">0.5</span><span class="p">}}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">WriteDiffPrefs</span><span class="p">(</span><span class="n">f2</span><span class="p">,</span> <span class="n">sites</span><span class="p">,</span> <span class="n">wts</span><span class="p">,</span> <span class="n">deltapi_means</span><span class="p">,</span> <span class="n">pr_deltapi_lt0</span><span class="p">,</span> <span class="n">pr_deltapi_gt0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f2</span><span class="o">.</span><span class="n">seek</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lines</span> <span class="o">=</span> <span class="n">f2</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lines</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s">&#39;# POSITION WT RMS_dPI dPI_A dPI_C dPI_G dPI_T Pr_dPI_A_lt0 Pr_dPI_A_gt0 Pr_dPI_C_lt0 Pr_dPI_C_gt0 Pr_dPI_G_lt0 Pr_dPI_G_gt0 Pr_dPI_T_lt0 Pr_dPI_T_gt0</span><span class="se">\n</span><span class="s">&#39;</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">.</span><span class="n">seek</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tup</span> <span class="o">=</span> <span class="n">ReadDiffPrefs</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">tup</span><span class="p">)</span> <span class="o">==</span> <span class="mi">6</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tup</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="n">sites</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tup</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="n">wts</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tup</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">==</span> <span class="n">deltapi_means</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tup</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span> <span class="o">==</span> <span class="n">tup</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span> <span class="o">==</span> <span class="bp">None</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="mf">1.0e-6</span> <span class="o">&gt;</span> <span class="n">tup</span><span class="p">[</span><span class="mi">5</span><span class="p">][</span><span class="s">&#39;1&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="nb">float</span><span class="p">(</span><span class="n">entries</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f2</span><span class="o">.</span><span class="n">seek</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tup</span> <span class="o">=</span> <span class="n">ReadDiffPrefs</span><span class="p">(</span><span class="n">f2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">tup</span><span class="p">)</span> <span class="o">==</span> <span class="mi">6</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tup</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="n">sites</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tup</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="n">wts</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tup</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">==</span> <span class="n">deltapi_means</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tup</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span> <span class="o">==</span> <span class="n">pr_deltapi_lt0</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tup</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span> <span class="o">==</span> <span class="n">pr_deltapi_gt0</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="mf">1.0e-6</span> <span class="o">&gt;</span> <span class="n">tup</span><span class="p">[</span><span class="mi">5</span><span class="p">][</span><span class="s">&#39;1&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="nb">float</span><span class="p">(</span><span class="n">entries</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f2</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="file_io.WritePreferences">
<tt class="descclassname">file_io.</tt><tt class="descname">WritePreferences</tt><big>(</big><em>f</em>, <em>sites</em>, <em>wts</em>, <em>pi_means</em>, <em>pi_95credint</em><big>)</big><a class="headerlink" href="#file_io.WritePreferences" title="Permalink to this definition">¶</a></dt>
<dd><p>Writes site-specific preferences to a file.</p>
<dl class="docutils">
<dt><em>f</em> is a writeable file-like object to which we write the preferences,</dt>
<dd>or a string giving the name of a file that we create.</dd>
<dt><em>sites</em> is a list of all site numbers (as strings) for which</dt>
<dd>we have preferences.</dd>
<dt><em>wts</em> is a dictionary with <em>wts[r]</em> giving the wildtype character</dt>
<dd>at site <em>r</em> for all <em>r</em> in <em>sites</em>.</dd>
<dt><em>pi_means</em> is a dictionary keyed by all sites (as strings). For each </dt>
<dd>site <em>r</em>, <em>pi_means[r]</em> is a dictionary where <em>pi_means[r][x]</em>
is the preference of <em>r</em> for character <em>x</em>, where the
characters are nucleotide, codons, or amino acids.</dd>
<dt><em>pi_95credint</em> can either be <em>None</em> or a dictionary keyed by all</dt>
<dd>sites that also key <em>pi_means</em>, and has as values 2-tuples giving
the upper and lower bounds of the 95% credible interval. (If it is
a dictionary with any value of <em>None</em>, that is the same as setting the
overall variable to <em>None</em>).</dd>
</dl>
<p>The written file has the following format:</p>
<div class="highlight-python"><div class="highlight"><pre># POSITION WT SITE_ENTROPY PI_A PI_C PI_G PI_T PI_A_95 PI_C_95 PI_G_95 PI_T_95
1 G 1.61 0.21 0.09 0.58 0.12 0.17,0.28 0.04,0.12 0.51,0.69 0.09,0.15
</pre></div>
</div>
<p>In this format, the first header line begins with <tt class="docutils literal"><span class="pre">#</span></tt> and indicates the
characters as the suffix to <tt class="docutils literal"><span class="pre">PI_</span></tt>. The 95% credible intervals (if given)
are indicated by columns beginning with <tt class="docutils literal"><span class="pre">PI_</span></tt> and ending with <tt class="docutils literal"><span class="pre">_95</span></tt>.
The <tt class="docutils literal"><span class="pre">SITE_ENTROPY</span></tt> column gives the site entropy in bits (log base 2)
as <span class="math">\(h_r = \sum_x \pi_{r,x} \times \log_2 \pi_{r,x}\)</span>.</p>
<p>The <tt class="docutils literal"><span class="pre">POSITION</span></tt> header can also be <tt class="docutils literal"><span class="pre">SITE</span></tt>.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">sites</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;1&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pi_means</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;1&#39;</span><span class="p">:{</span><span class="s">&#39;A&#39;</span><span class="p">:</span><span class="mf">0.2</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">:</span><span class="mf">0.3</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">:</span><span class="mf">0.4</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">:</span><span class="mf">0.1</span><span class="p">}}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">wts</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;1&#39;</span><span class="p">:</span><span class="s">&#39;G&#39;</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pi_95credint</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;1&#39;</span><span class="p">:{</span><span class="s">&#39;A&#39;</span><span class="p">:(</span><span class="mf">0.1</span><span class="p">,</span> <span class="mf">0.3</span><span class="p">),</span> <span class="s">&#39;C&#39;</span><span class="p">:(</span><span class="mf">0.25</span><span class="p">,</span> <span class="mf">0.35</span><span class="p">),</span> <span class="s">&#39;G&#39;</span><span class="p">:(</span><span class="mf">0.3</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">),</span> <span class="s">&#39;T&#39;</span><span class="p">:(</span><span class="mf">0.05</span><span class="p">,</span> <span class="mf">0.15</span><span class="p">)}}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span> <span class="o">=</span> <span class="n">cStringIO</span><span class="o">.</span><span class="n">StringIO</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">WritePreferences</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="n">sites</span><span class="p">,</span> <span class="n">wts</span><span class="p">,</span> <span class="n">pi_means</span><span class="p">,</span> <span class="n">pi_95credint</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">.</span><span class="n">seek</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lines</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lines</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s">&#39;# POSITION WT SITE_ENTROPY PI_A PI_C PI_G PI_T PI_A_95 PI_C_95 PI_G_95 PI_T_95</span><span class="se">\n</span><span class="s">&#39;</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">.</span><span class="n">seek</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tup</span> <span class="o">=</span> <span class="n">ReadPreferences</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tup</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="n">sites</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tup</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span> <span class="o">==</span> <span class="n">pi_95credint</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f2</span> <span class="o">=</span> <span class="n">cStringIO</span><span class="o">.</span><span class="n">StringIO</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">WritePreferences</span><span class="p">(</span><span class="n">f2</span><span class="p">,</span> <span class="n">sites</span><span class="p">,</span> <span class="n">wts</span><span class="p">,</span> <span class="n">pi_means</span><span class="p">,</span> <span class="bp">None</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f2</span><span class="o">.</span><span class="n">seek</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lines</span> <span class="o">=</span> <span class="n">f2</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lines</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s">&#39;# POSITION WT SITE_ENTROPY PI_A PI_C PI_G PI_T</span><span class="se">\n</span><span class="s">&#39;</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f2</span><span class="o">.</span><span class="n">seek</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">(</span><span class="n">sites2</span><span class="p">,</span> <span class="n">wts2</span><span class="p">,</span> <span class="n">pi_means2</span><span class="p">,</span> <span class="n">pi_95credint2</span><span class="p">,</span> <span class="n">h</span><span class="p">)</span> <span class="o">=</span> <span class="n">ReadPreferences</span><span class="p">(</span><span class="n">f2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sites</span> <span class="o">==</span> <span class="n">sites2</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">wts</span> <span class="o">==</span> <span class="n">wts2</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pi_95credint2</span> <span class="o">==</span> <span class="bp">None</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">r</span> <span class="o">=</span> <span class="n">sites</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">([</span><span class="nb">abs</span><span class="p">(</span><span class="n">pi_means</span><span class="p">[</span><span class="n">r</span><span class="p">][</span><span class="n">x</span><span class="p">]</span> <span class="o">-</span> <span class="n">pi_means2</span><span class="p">[</span><span class="n">r</span><span class="p">][</span><span class="n">x</span><span class="p">])</span> <span class="o">&lt;</span> <span class="mf">1.0e-6</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">pi_means</span><span class="p">[</span><span class="n">r</span><span class="p">]</span><span class="o">.</span><span class="n">keys</span><span class="p">()])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="s">&quot;</span><span class="si">%.3f</span><span class="s">&quot;</span> <span class="o">%</span> <span class="n">h</span><span class="p">[</span><span class="n">r</span><span class="p">]</span>
<span class="go">1.846</span>
</pre></div>
</div>
</dd></dl>

</div>
</div>
<span class="target" id="module-utils"></span><div class="section" id="utils-module">
<h2><tt class="docutils literal"><span class="pre">utils</span></tt> module<a class="headerlink" href="#utils-module" title="Permalink to this headline">¶</a></h2>
<p>Module with utilities for manipulating data related to <tt class="docutils literal"><span class="pre">dms_tools</span></tt>.</p>
<p>Written by Jesse Bloom.</p>
<div class="section" id="id3">
<h3>Functions in this module<a class="headerlink" href="#id3" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><em>NaturalSort</em> : Performs a natural sort (alphanumeric sort) on a list.</li>
<li><em>AvgMutRate</em> : Returns the average mutation rate across all sites.</li>
<li><em>SumCodonToAA</em> : Sums all codon entries for each amino acid.</li>
<li><em>SiteEntropy</em> : Computes site entropy.</li>
<li><em>RMS</em> : Computes root-mean-square value.</li>
<li><em>RemoveStopFromPreferences</em> : removes stop codon</li>
<li><em>RemoveStopFromDiffPrefs</em> : removes stop codon</li>
<li><em>SumPrefsDiffPrefs</em> : sums preferences and differential preferences</li>
<li><em>Pref_or_DiffPref</em> : determines whether data represent preferences or differential preferences.</li>
<li><em>PrefsToEnrichments</em> : converts preferences to enrichment ratios.</li>
<li><em>AdjustErrorCounts</em> : adjust error counts so they don&#8217;t exceed actual counts.</li>
<li><em>CheckReadQuality</em> : checks quality of pair of reads.</li>
<li><em>BuildReadConsensus</em> : builds consensus of reads if they are sufficiently similar.</li>
<li><em>AlignSubamplicon</em> : attempt to align subamplicon at specified position.</li>
<li><em>ClassifyCodonCounts</em> : classifies codon mutations.</li>
<li><em>CodonMutsCumulFracs</em> : cumulative counts of codon mutations.</li>
</ul>
</div>
<div class="section" id="id4">
<h3>Function documentation<a class="headerlink" href="#id4" title="Permalink to this headline">¶</a></h3>
<dl class="function">
<dt id="utils.AdjustErrorCounts">
<tt class="descclassname">utils.</tt><tt class="descname">AdjustErrorCounts</tt><big>(</big><em>wt</em>, <em>counts</em>, <em>maxexcess</em><big>)</big><a class="headerlink" href="#utils.AdjustErrorCounts" title="Permalink to this definition">¶</a></dt>
<dd><p>Adjust error counts so that they don&#8217;t exceed counts of interest.</p>
<p>This deals with the case when the error control estimates a higher
rate of a mutation than is found in the actual sample. This will confound
the MCMC inference by <tt class="docutils literal"><span class="pre">dms_inferprefs</span></tt> and <tt class="docutils literal"><span class="pre">dms_inferdiffprefs</span></tt>.
So this script adjusts down the error counts until they are compatbile
with the counts in the actual sample.</p>
<p>Returns a dictionary like <em>counts</em> but with adjusted counts.</p>
<p>This is a function utilized to clean counts data for <tt class="docutils literal"><span class="pre">dms_inferprefs</span></tt>
and <tt class="docutils literal"><span class="pre">dms_inferdiffprefs</span></tt>.</p>
<p>Essentially, it checks if the mutation rate in the error control
is higher than that of the sample for which it is supposed to control.
If so, it reduces the mutation counts for the error control so that
they only exceed the controlled for sample by <em>maxexcess</em>.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">wt</span> <span class="o">=</span> <span class="s">&#39;A&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;nrpre&#39;</span><span class="p">:{</span><span class="s">&#39;A&#39;</span><span class="p">:</span><span class="mi">500</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">:</span><span class="mi">10</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">:</span><span class="mi">40</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">:</span><span class="mi">20</span><span class="p">},</span> <span class="s">&#39;nrerrpre&#39;</span><span class="p">:{</span><span class="s">&#39;A&#39;</span><span class="p">:</span><span class="mi">250</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">:</span><span class="mi">1</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">:</span><span class="mi">30</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">:</span><span class="mi">10</span><span class="p">},</span> <span class="s">&#39;nrpost&#39;</span><span class="p">:{</span><span class="s">&#39;A&#39;</span><span class="p">:</span><span class="mi">500</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">:</span><span class="mi">20</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">:</span><span class="mi">20</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">:</span><span class="mi">20</span><span class="p">},</span> <span class="s">&#39;nrerrpost&#39;</span><span class="p">:{</span><span class="s">&#39;A&#39;</span><span class="p">:</span><span class="mi">1000</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">:</span><span class="mi">30</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">:</span><span class="mi">20</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">:</span><span class="mi">50</span><span class="p">}}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adjusted_counts</span> <span class="o">=</span> <span class="n">AdjustErrorCounts</span><span class="p">(</span><span class="n">wt</span><span class="p">,</span> <span class="n">counts</span><span class="p">,</span> <span class="n">maxexcess</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="p">[</span><span class="s">&#39;nrpre&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">adjusted_counts</span><span class="p">[</span><span class="s">&#39;nrpre&#39;</span><span class="p">]</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="p">[</span><span class="s">&#39;nrpost&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">adjusted_counts</span><span class="p">[</span><span class="s">&#39;nrpost&#39;</span><span class="p">]</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adjusted_counts</span><span class="p">[</span><span class="s">&#39;nrerrpre&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="p">{</span><span class="s">&#39;A&#39;</span><span class="p">:</span><span class="mi">250</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">:</span><span class="mi">1</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">:</span><span class="mi">21</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">:</span><span class="mi">10</span><span class="p">}</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adjusted_counts</span><span class="p">[</span><span class="s">&#39;nrerrpost&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="p">{</span><span class="s">&#39;A&#39;</span><span class="p">:</span><span class="mi">1000</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">:</span><span class="mi">30</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">:</span><span class="mi">20</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">:</span><span class="mi">41</span><span class="p">}</span>
<span class="go">True</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">wt</span> <span class="o">=</span> <span class="s">&#39;A&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;nrpre&#39;</span><span class="p">:{</span><span class="s">&#39;A&#39;</span><span class="p">:</span><span class="mi">500</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">:</span><span class="mi">10</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">:</span><span class="mi">40</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">:</span><span class="mi">20</span><span class="p">},</span> <span class="s">&#39;nrpost&#39;</span><span class="p">:{</span><span class="s">&#39;A&#39;</span><span class="p">:</span><span class="mi">500</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">:</span><span class="mi">20</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">:</span><span class="mi">20</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">:</span><span class="mi">20</span><span class="p">},</span> <span class="s">&#39;nrerr&#39;</span><span class="p">:{</span><span class="s">&#39;A&#39;</span><span class="p">:</span><span class="mi">1000</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">:</span><span class="mi">30</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">:</span><span class="mi">20</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">:</span><span class="mi">50</span><span class="p">}}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adjusted_counts</span> <span class="o">=</span> <span class="n">AdjustErrorCounts</span><span class="p">(</span><span class="n">wt</span><span class="p">,</span> <span class="n">counts</span><span class="p">,</span> <span class="n">maxexcess</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="p">[</span><span class="s">&#39;nrpre&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">adjusted_counts</span><span class="p">[</span><span class="s">&#39;nrpre&#39;</span><span class="p">]</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="p">[</span><span class="s">&#39;nrpost&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">adjusted_counts</span><span class="p">[</span><span class="s">&#39;nrpost&#39;</span><span class="p">]</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adjusted_counts</span><span class="p">[</span><span class="s">&#39;nrerr&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="p">{</span><span class="s">&#39;A&#39;</span><span class="p">:</span><span class="mi">1000</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">:</span><span class="mi">21</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">:</span><span class="mi">20</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">:</span><span class="mi">41</span><span class="p">}</span>
<span class="go">True</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;nrstart&#39;</span><span class="p">:{</span><span class="s">&#39;A&#39;</span><span class="p">:</span><span class="mi">500</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">:</span><span class="mi">10</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">:</span><span class="mi">40</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">:</span><span class="mi">20</span><span class="p">},</span> <span class="s">&#39;nrs1&#39;</span><span class="p">:{</span><span class="s">&#39;A&#39;</span><span class="p">:</span><span class="mi">1000</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">:</span><span class="mi">20</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">:</span><span class="mi">80</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">:</span><span class="mi">40</span><span class="p">},</span> <span class="s">&#39;nrs2&#39;</span><span class="p">:{</span><span class="s">&#39;A&#39;</span><span class="p">:</span><span class="mi">500</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">:</span><span class="mi">20</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">:</span><span class="mi">20</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">:</span><span class="mi">20</span><span class="p">},</span> <span class="s">&#39;nrerr&#39;</span><span class="p">:{</span><span class="s">&#39;A&#39;</span><span class="p">:</span><span class="mi">1000</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">:</span><span class="mi">30</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">:</span><span class="mi">20</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">:</span><span class="mi">50</span><span class="p">}}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adjusted_counts</span> <span class="o">=</span> <span class="n">AdjustErrorCounts</span><span class="p">(</span><span class="n">wt</span><span class="p">,</span> <span class="n">counts</span><span class="p">,</span> <span class="n">maxexcess</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="p">[</span><span class="s">&#39;nrstart&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">adjusted_counts</span><span class="p">[</span><span class="s">&#39;nrstart&#39;</span><span class="p">]</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="p">[</span><span class="s">&#39;nrs1&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">adjusted_counts</span><span class="p">[</span><span class="s">&#39;nrs1&#39;</span><span class="p">]</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="p">[</span><span class="s">&#39;nrs2&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">adjusted_counts</span><span class="p">[</span><span class="s">&#39;nrs2&#39;</span><span class="p">]</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adjusted_counts</span><span class="p">[</span><span class="s">&#39;nrerr&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="p">{</span><span class="s">&#39;A&#39;</span><span class="p">:</span><span class="mi">1000</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">:</span><span class="mi">21</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">:</span><span class="mi">20</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">:</span><span class="mi">41</span><span class="p">}</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="utils.AlignSubamplicon">
<tt class="descclassname">utils.</tt><tt class="descname">AlignSubamplicon</tt><big>(</big><em>refseq</em>, <em>r1</em>, <em>r2</em>, <em>refseqstart</em>, <em>refseqend</em>, <em>maxmuts</em>, <em>maxN</em>, <em>chartype</em>, <em>counts</em>, <em>use_cutils=True</em><big>)</big><a class="headerlink" href="#utils.AlignSubamplicon" title="Permalink to this definition">¶</a></dt>
<dd><p>Tries to align subamplicon into reference sequence.</p>
<p>The return value is <em>(True, nmuts)</em> if the subamplicon aligns and <em>False</em> otherwise.
In the case where the subamplicon aligns, <em>nmuts</em> is the number of mutations
of the type indicated by <em>chartype</em>.
If the subamplicon aligns, then the character counts in <em>counts</em> are updated.
A subamplicon is considered to align if there are &lt;= <em>maxmuts</em> mutations
of <em>chartype</em>. Characters that contain ambiguous nucleotides (<tt class="docutils literal"><span class="pre">N</span></tt>)
are not counted as mutations or recorded as counts. If <em>r1</em> and <em>r2</em> 
overlap, identities are counted as ambiguous unless both reads agree.
If <em>r1</em> and <em>r2</em> do not reach far enough to overlap, and uncalled
identities in between are called as ambiguous.</p>
<p>Note that if using the default fast C implementation (which is done if
<em>use_cutils</em> is <em>True</em>), then there is not much error checking on the validity
of the calling variables. Therefore, you could end up with a segmentation
fault with invalid calling variables and <em>use_cutils</em> of <em>False</em>.</p>
<p>Here are the calling variables:</p>
<ul>
<li><p class="first"><em>refseq</em> : string giving the reference sequencing to which we attempt
to align the subamplicon. Assumed uppercase nucleotides.</p>
</li>
<li><p class="first"><em>r1</em> : string giving the forward read, assumed uppercase nucleotides
with <tt class="docutils literal"><span class="pre">N</span></tt> for ambiguous / low-quality sites. Note that if this read
has 5&#8217; portions (such as a barcode) that we aren&#8217;t aligning, they should
have been trimmed prior to calling this function.</p>
</li>
<li><p class="first"><em>r2</em> : like <em>r1</em> but for reverse read.</p>
</li>
<li><dl class="first docutils">
<dt><em>refseqstart</em> <span class="classifier-delimiter">:</span> <span class="classifier">nucleotide in <em>refseq</em> (in 1, 2, ... numbering) where </span></dt>
<dd><p class="first last">first nucleotide in <em>r1</em> aligns.</p>
</dd>
</dl>
</li>
<li><p class="first"><em>refseqend</em> : nucleotide in <em>refseq</em> (in 1, 2, ... numbering) where
first nucleotide in <em>r2</em> aligns (note that <em>r2</em> then reads backwards
towards 5&#8217; end of <em>refseq</em>).</p>
</li>
<li><p class="first"><em>maxmuts</em> : maximum number of mutations of character type <em>chartype</em>
that are allowed; subamplicons with more than this many mutations
are considered <strong>not</strong> to align and <em>False</em> is returned.</p>
</li>
<li><p class="first"><em>maxN</em> : maximum number of ambiguous nucleotides allowed in the 
consensus built up from the two reads (i.e. including the reads
and any overlap).</p>
</li>
<li><p class="first"><em>chartype</em> : type of character for which mutations are counted; currently
the only allowable value is the string &#8220;codon&#8221;.</p>
</li>
<li><p class="first"><em>counts</em> : a dictionary for counting mutations to <em>refseq</em> of character
type <em>chartype</em> in the format used by <em>dms_tools.utils.WriteDMSCounts</em>
and <em>dms_tools.utils.ReadDMSCounts</em>.</p>
</li>
<li><p class="first"><em>use_cutils</em> : do we use the fast C-implemenation of this function
in <em>dms_tools.cutils</em>?</p>
</li>
</ul>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;1&#39;</span><span class="p">:</span><span class="nb">dict</span><span class="p">([(</span><span class="s">&#39;WT&#39;</span><span class="p">,</span> <span class="s">&#39;ATG&#39;</span><span class="p">)]</span> <span class="o">+</span> <span class="p">[(</span><span class="n">codon</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span> <span class="k">for</span> <span class="n">codon</span> <span class="ow">in</span> <span class="n">dms_tools</span><span class="o">.</span><span class="n">codons</span><span class="p">]),</span>
<span class="gp">... </span>          <span class="s">&#39;2&#39;</span><span class="p">:</span><span class="nb">dict</span><span class="p">([(</span><span class="s">&#39;WT&#39;</span><span class="p">,</span> <span class="s">&#39;GGG&#39;</span><span class="p">)]</span> <span class="o">+</span> <span class="p">[(</span><span class="n">codon</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span> <span class="k">for</span> <span class="n">codon</span> <span class="ow">in</span> <span class="n">dms_tools</span><span class="o">.</span><span class="n">codons</span><span class="p">]),</span>
<span class="gp">... </span>          <span class="s">&#39;3&#39;</span><span class="p">:</span><span class="nb">dict</span><span class="p">([(</span><span class="s">&#39;WT&#39;</span><span class="p">,</span> <span class="s">&#39;AAA&#39;</span><span class="p">)]</span> <span class="o">+</span> <span class="p">[(</span><span class="n">codon</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span> <span class="k">for</span> <span class="n">codon</span> <span class="ow">in</span> <span class="n">dms_tools</span><span class="o">.</span><span class="n">codons</span><span class="p">])}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">returntup</span> <span class="o">=</span> <span class="n">AlignSubamplicon</span><span class="p">(</span><span class="s">&#39;ATGGGGAAA&#39;</span><span class="p">,</span> <span class="s">&#39;GGGGAA&#39;</span><span class="p">,</span> <span class="s">&#39;TTTCCC&#39;</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="s">&#39;codon&#39;</span><span class="p">,</span> <span class="n">counts</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">returntup</span> <span class="o">==</span> <span class="p">(</span><span class="bp">True</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">([</span><span class="n">counts</span><span class="p">[</span><span class="s">&#39;1&#39;</span><span class="p">][</span><span class="n">codon</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span> <span class="k">for</span> <span class="n">codon</span> <span class="ow">in</span> <span class="n">dms_tools</span><span class="o">.</span><span class="n">codons</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="p">[</span><span class="s">&#39;2&#39;</span><span class="p">][</span><span class="s">&#39;GGG&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">([</span><span class="n">counts</span><span class="p">[</span><span class="s">&#39;2&#39;</span><span class="p">][</span><span class="n">codon</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span> <span class="k">for</span> <span class="n">codon</span> <span class="ow">in</span> <span class="n">dms_tools</span><span class="o">.</span><span class="n">codons</span> <span class="k">if</span> <span class="n">codon</span> <span class="o">!=</span> <span class="s">&#39;GGG&#39;</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">([</span><span class="n">counts</span><span class="p">[</span><span class="s">&#39;2&#39;</span><span class="p">][</span><span class="n">codon</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span> <span class="k">for</span> <span class="n">codon</span> <span class="ow">in</span> <span class="n">dms_tools</span><span class="o">.</span><span class="n">codons</span> <span class="k">if</span> <span class="n">codon</span> <span class="o">!=</span> <span class="s">&#39;GGG&#39;</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="p">[</span><span class="s">&#39;3&#39;</span><span class="p">][</span><span class="s">&#39;AAA&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">([</span><span class="n">counts</span><span class="p">[</span><span class="s">&#39;3&#39;</span><span class="p">][</span><span class="n">codon</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span> <span class="k">for</span> <span class="n">codon</span> <span class="ow">in</span> <span class="n">dms_tools</span><span class="o">.</span><span class="n">codons</span> <span class="k">if</span> <span class="n">codon</span> <span class="o">!=</span> <span class="s">&#39;AAA&#39;</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">returntup</span> <span class="o">=</span> <span class="n">AlignSubamplicon</span><span class="p">(</span><span class="s">&#39;ATGGGGAAA&#39;</span><span class="p">,</span> <span class="s">&#39;GGGGAA&#39;</span><span class="p">,</span> <span class="s">&#39;TTTCCC&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="s">&#39;codon&#39;</span><span class="p">,</span> <span class="n">counts</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">returntup</span> <span class="o">==</span> <span class="bp">False</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">returntup</span> <span class="o">=</span> <span class="n">AlignSubamplicon</span><span class="p">(</span><span class="s">&#39;ATGGGGAAA&#39;</span><span class="p">,</span> <span class="s">&#39;GGGGAT&#39;</span><span class="p">,</span> <span class="s">&#39;TTTCCC&#39;</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="s">&#39;codon&#39;</span><span class="p">,</span> <span class="n">counts</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">returntup</span> <span class="o">==</span> <span class="bp">False</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">returntup</span> <span class="o">=</span> <span class="n">AlignSubamplicon</span><span class="p">(</span><span class="s">&#39;ATGGGGAAA&#39;</span><span class="p">,</span> <span class="s">&#39;GGGGAT&#39;</span><span class="p">,</span> <span class="s">&#39;TTTCCC&#39;</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="s">&#39;codon&#39;</span><span class="p">,</span> <span class="n">counts</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">returntup</span> <span class="o">==</span> <span class="p">(</span><span class="bp">True</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="p">[</span><span class="s">&#39;2&#39;</span><span class="p">][</span><span class="s">&#39;GGG&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">2</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="p">[</span><span class="s">&#39;3&#39;</span><span class="p">][</span><span class="s">&#39;AAA&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">returntup</span> <span class="o">=</span> <span class="n">AlignSubamplicon</span><span class="p">(</span><span class="s">&#39;ATGGGGAAA&#39;</span><span class="p">,</span> <span class="s">&#39;GAGGAA&#39;</span><span class="p">,</span> <span class="s">&#39;TTTCCT&#39;</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="s">&#39;codon&#39;</span><span class="p">,</span> <span class="n">counts</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">returntup</span> <span class="o">==</span> <span class="bp">False</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">returntup</span> <span class="o">=</span> <span class="n">AlignSubamplicon</span><span class="p">(</span><span class="s">&#39;ATGGGGAAA&#39;</span><span class="p">,</span> <span class="s">&#39;GAGGAA&#39;</span><span class="p">,</span> <span class="s">&#39;TTTCCT&#39;</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="s">&#39;codon&#39;</span><span class="p">,</span> <span class="n">counts</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="p">[</span><span class="s">&#39;2&#39;</span><span class="p">][</span><span class="s">&#39;GGG&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">2</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="p">[</span><span class="s">&#39;2&#39;</span><span class="p">][</span><span class="s">&#39;AGG&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="utils.AvgMutRate">
<tt class="descclassname">utils.</tt><tt class="descname">AvgMutRate</tt><big>(</big><em>counts</em>, <em>chartype</em><big>)</big><a class="headerlink" href="#utils.AvgMutRate" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns average mutation rate.</p>
<p><em>counts</em> is a dictionary of deep mutational scanning counts in the format returned
by <em>file_io.ReadDMSCounts</em>.</p>
<p><em>chartype</em> is the character type of these counts, with the same meaning 
as for <em>file_io.ReadDMSCounts</em>.</p>
<p>This function computes the average mutation rate over all counts for all sites for 
mutations with each possible number of nucleotide changes. Specifically, we define the average
rate of mutations to characters with <span class="math">\(m\)</span> nucleotide changes relative to wildtype as</p>
<div class="math">
\[\overline{\epsilon_m} = \frac{1}{L}\sum\limits_r \frac{1}{N_r} \sum\limits_x n_{r,x} \times \delta_{m,D_{x,\operatorname{wt}\left(r\right)}}\]</div>
<p>where <span class="math">\(r\)</span> ranges over all <span class="math">\(L\)</span> sites for which there are deep mutational scanning counts,
<span class="math">\(N_r\)</span> is the total number of counts at site <span class="math">\(r\)</span>, <span class="math">\(n_{r,x}\)</span> is the number
of counts that report character <span class="math">\(x\)</span> at site <span class="math">\(r\)</span>, <span class="math">\(D_{x,\operatorname{wt}\left(r\right)}\)</span>
is the number of nucleotide differences between character <span class="math">\(x\)</span> and the wildtype character
<span class="math">\(\operatorname{wt}\left(r\right)\)</span> at site <span class="math">\(r\)</span>, and <span class="math">\(\delta_{xy}\)</span> is the Kronecker
delta. Note that this definition implies <span class="math">\(1 = \sum_m \overline{\epsilon_m}\)</span>. For nucleotides,
there are just two values of <span class="math">\(m\)</span>: a value of <span class="math">\(m = 0\)</span> is no differences (the wildtype
character) and a value of <span class="math">\(m = 1\)</span> is the a mutant nucleotide. For codons, <span class="math">\(m\)</span> can
be 1, 2, or 3.</p>
<p>The return value is the dictionary <em>mutrates</em> which is keyed by all values of <span class="math">\(m\)</span>
for the <em>chartype</em> in question. The value of <em>mutrates[m]</em> is the mutation rate for
characters with this number of differences.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;1&#39;</span><span class="p">:{</span><span class="s">&#39;WT&#39;</span><span class="p">:</span><span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="s">&#39;A&#39;</span><span class="p">:</span><span class="mi">7</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">:</span><span class="mi">1</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">:</span><span class="mi">1</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">:</span><span class="mi">1</span><span class="p">},</span>
<span class="gp">... </span>          <span class="s">&#39;2&#39;</span><span class="p">:{</span><span class="s">&#39;WT&#39;</span><span class="p">:</span><span class="s">&#39;C&#39;</span><span class="p">,</span> <span class="s">&#39;A&#39;</span><span class="p">:</span><span class="mi">2</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">:</span><span class="mi">8</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">:</span><span class="mi">0</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">:</span><span class="mi">0</span><span class="p">}}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mutrates</span> <span class="o">=</span> <span class="n">AvgMutRate</span><span class="p">(</span><span class="n">counts</span><span class="p">,</span> <span class="s">&#39;DNA&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="s">&quot;</span><span class="si">%.3f</span><span class="s">&quot;</span> <span class="o">%</span> <span class="n">mutrates</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="go">0.750</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="s">&quot;</span><span class="si">%.3f</span><span class="s">&quot;</span> <span class="o">%</span> <span class="n">mutrates</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="go">0.250</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="utils.BuildReadConsensus">
<tt class="descclassname">utils.</tt><tt class="descname">BuildReadConsensus</tt><big>(</big><em>reads</em>, <em>minreadidentity</em>, <em>minreadconcurrence</em>, <em>maxreadtrim</em>, <em>use_cutils=True</em><big>)</big><a class="headerlink" href="#utils.BuildReadConsensus" title="Permalink to this definition">¶</a></dt>
<dd><p>Builds consensus of reads if they are sufficiently identical.</p>
<p><em>reads</em> : is a list of <em>(r1, r2)</em> read pairs, which are assumed to be upper
case with <tt class="docutils literal"><span class="pre">N</span></tt> inserted for ambiguous or low-quality positions (i.e. the
reads have passed through <em>CheckReadQuality</em>). <em>reads</em> must have at
least two entries.</p>
<p><em>minreadidentity</em> : all reads in <em>reads</em> are required to have &gt;= this fraction
of their nucleotides identical and non-ambiguous (not <tt class="docutils literal"><span class="pre">N</span></tt>) to the
first read in <em>reads</em>. If this is not the case, return <em>False</em>.</p>
<p><em>minreadconcurrence</em> : among reads that pass 
<em>minreadidentity</em>, return the consensus identity at a position
if &gt;= this fraction; otherwise make that position an <tt class="docutils literal"><span class="pre">N</span></tt>.
Must be &gt; 0.5 and &lt;= 1.</p>
<p><em>maxreadtrim</em> : all R1 reads in <em>reads</em> need to be same length. If
they aren&#8217;t initially, trim up to this many nucleotides off the longer
ones. If that doesn&#8217;t make them the same length, then return <em>False</em>.
The same is done for R2 reads.</p>
<p><em>use_cutils</em> specifies that we use the fast C implementation of this
function provided by <em>dms_tools.cutils</em>.</p>
<p>The return variable will be <em>False</em> if reads differ in length
after applying <em>maxreadtrim</em> or
<em>minreadidentity</em> is failed; otherwise it will be a 2-tuple
<em>(r1_consensus, r2_consensus)</em> with each element a string of 
the consensus with <tt class="docutils literal"><span class="pre">N</span></tt> at positions where the read concurrence
does not satisfy <em>minreadconcurrence</em>.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">reads</span> <span class="o">=</span> <span class="p">[(</span><span class="s">&#39;ATGC&#39;</span><span class="p">,</span> <span class="s">&#39;CGAT&#39;</span><span class="p">),</span> <span class="p">(</span><span class="s">&#39;ATGC&#39;</span><span class="p">,</span> <span class="s">&#39;CGAT&#39;</span><span class="p">)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">BuildReadConsensus</span><span class="p">(</span><span class="n">reads</span><span class="p">,</span> <span class="mf">0.75</span><span class="p">,</span> <span class="mf">0.6</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="go">(&#39;ATGC&#39;, &#39;CGAT&#39;)</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">reads</span> <span class="o">=</span> <span class="p">[(</span><span class="s">&#39;TTGC&#39;</span><span class="p">,</span> <span class="s">&#39;CGAT&#39;</span><span class="p">),</span> <span class="p">(</span><span class="s">&#39;ATGC&#39;</span><span class="p">,</span> <span class="s">&#39;CGANA&#39;</span><span class="p">)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">BuildReadConsensus</span><span class="p">(</span><span class="n">reads</span><span class="p">,</span> <span class="mf">0.75</span><span class="p">,</span> <span class="mf">0.6</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="go">(&#39;NTGC&#39;, &#39;CGAN&#39;)</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="utils.CheckReadQuality">
<tt class="descclassname">utils.</tt><tt class="descname">CheckReadQuality</tt><big>(</big><em>r1</em>, <em>r2</em>, <em>q1</em>, <em>q2</em>, <em>minq</em>, <em>maxlowqfrac</em>, <em>barcodelength</em>, <em>use_cutils=True</em><big>)</big><a class="headerlink" href="#utils.CheckReadQuality" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks quality of reads and replaces low Q nucleotides with <tt class="docutils literal"><span class="pre">N</span></tt>.</p>
<p><em>r1</em> and <em>r2</em> are strings representing a pair of sequencing reads.</p>
<p><em>q1</em> and <em>q2</em> are strings giving the Q-scores for the nucleotides in
<em>r1</em> and <em>r2</em>; these Q-scores are assumed to be encoded as the ASCII code
for the Q-score + 33.</p>
<p><em>minq</em> is the minimum allowed Q-score (nucleotides with Q &lt; <em>minq</em>)
are considered low quality.</p>
<p><em>maxlowqfrac</em> is the maximum allowable number of <tt class="docutils literal"><span class="pre">N</span></tt> or low quality
nucleotides allowed in each read.</p>
<p><em>barcodelength</em> is the number of nucleotides at the beginning of each read
that are required to be high-quality (Q &gt; <em>minq</em> and not an <tt class="docutils literal"><span class="pre">N</span></tt>).</p>
<p><em>use_cutils</em> specifies that we use the fast C implementation of this
function provided by <em>dms_tools.cutils</em>.</p>
<p>This function examines the quality of read <em>r1</em> and <em>r2</em>. It returns
<em>False</em> if any of the following are true:</p>
<blockquote>
<div><p>1) Either read has a low-quality nucleotide or an <tt class="docutils literal"><span class="pre">N</span></tt> in the first
<em>barcodelength</em> nucleotides.</p>
<p>2) Either read has &gt; <em>maxlowqfrac</em> of its nucleotides that are either
low-quality or are <tt class="docutils literal"><span class="pre">N</span></tt>.</p>
</div></blockquote>
<p>If neither of those conditions are true, then the return variable
is <em>(r1_new, r2_new)</em>, which are copies of <em>r1</em> and <em>r2</em> where all
nucleotides are upper case and any low quality nucleotides have been
replaced by <tt class="docutils literal"><span class="pre">N</span></tt>.</p>
</dd></dl>

<dl class="function">
<dt id="utils.ClassifyCodonCounts">
<tt class="descclassname">utils.</tt><tt class="descname">ClassifyCodonCounts</tt><big>(</big><em>codon_counts</em><big>)</big><a class="headerlink" href="#utils.ClassifyCodonCounts" title="Permalink to this definition">¶</a></dt>
<dd><p>Classifies codon mutations.</p>
<p>CALLING VARIABLES:</p>
<p><cite>codon_counts</cite> : dictionary as returned by <cite>dms_tools.file_io.ReadDMSCounts</cite>
with <em>chartype</em> of <em>codon</em>.</p>
<p>RESULT OF CALLING THIS FUNCTION:</p>
<p>This function returns a new dictionary that is a copy of <em>codon_counts</em>
but also has the following new keys for each site:</p>
<p>&#8216;COUNTS&#8217; : total number of counts of non-ambiguous codons.</p>
<p>&#8216;N_WT&#8217; : total number of called codons 
that match &#8216;WT&#8217;, the wildtpe codon at this position.</p>
<p>&#8216;N_NS&#8217; : total number of called codons 
that represent non-synonymous mutations from the
wildtype codon. Mutations from stop codons to non-stop codons
are classified as nonsynonymous. Mutations from non-stop codons
to stop codons are not classified as nonsynonymous, instead look
under &#8216;N_STOP&#8217;.</p>
<p>&#8216;N_SYN&#8217; : total number of called codons 
that represent synonymous mutations from the wildtype
codon. Mutations from stop codons to other stop codons are
classified as synonymous.</p>
<p>&#8216;N_1MUT&#8217;, &#8216;N_2MUT&#8217;, &#8216;N_3MUT&#8217; : the total number of called
mutant codons that contain one, two, or
three nucleotide mutations relative to the wildtype codon at the site.</p>
<p>&#8216;N_STOP&#8217; : total number of definitively called codons (all nucleotides
upper case) that encode stop codons.</p>
<p>In addition, the following keys are added directly to <cite>codon_counts</cite>, and
represent totals over all codon positions:</p>
<p>&#8216;TOTAL_COUNTS&#8217; : total number of counts for all codons.</p>
<p>&#8216;TOTAL_NS&#8217; : total number of &#8216;N_NS&#8217; for all codons.</p>
<p>&#8216;TOTAL_SYN&#8217; : total number of &#8216;N_SYN&#8217; for all codons.</p>
<p>&#8216;TOTAL_STOP&#8217; : total number of &#8216;N_STOP&#8217; for all codons.</p>
<p>&#8216;TOTAL_MUT&#8217; : sum of &#8216;N_NS&#8217; + &#8216;N_SYN&#8217; + &#8216;N_STOP&#8217;, total number of mutations.</p>
<p>&#8216;TOTAL_N_1MUT&#8217;, &#8216;TOTAL_N_2MUT&#8217;, &#8216;TOTAL_N_3MUT&#8217; : total number of
&#8216;N_1MUT&#8217;, N_2MUT&#8217;, and &#8216;N_3MUT&#8217; for all codons.</p>
</dd></dl>

<dl class="function">
<dt id="utils.CodonMutsCumulFracs">
<tt class="descclassname">utils.</tt><tt class="descname">CodonMutsCumulFracs</tt><big>(</big><em>codon_counts</em><big>)</big><a class="headerlink" href="#utils.CodonMutsCumulFracs" title="Permalink to this definition">¶</a></dt>
<dd><p>Fraction of codon mutations found &gt;= some number of times.</p>
<p><em>codon_counts</em> : a dictionary of the type returned by calling
<em>dms_tools.file_io.ReadDMSCounts</em> with <em>chartype</em> of <em>codon</em>.</p>
<p>For all sites in <em>codon_counts</em>, this function iterates
through all possible codon mutations and counts how many times
they occur. These counts are then used to generate the following
is the number of codons. It records the number of definitely called
(all uppercase nucleotides) occurrences of each of these mutations. It then
returns the following tuple:
<em>(all_cumulfracs, all_counts, syn_cumulfracs, syn_counts, multi_nt_all_cumulfracs, multi_nt_all_counts, multi_nt_syn_cumulfracs, multi_nt_syn_counts)</em></p>
<blockquote>
<div><ul class="simple">
<li><em>all_cumulfracs[i]</em> is the fraction of all of the possible codon mutations
that are found &gt;= <em>i</em> times.</li>
<li><em>all_counts</em> is the total number of unique codon mutations.</li>
<li><em>syn_cumulfracs[i]</em> is the fraction of all of synonymous codon mutations
that are found &gt;= <em>i</em> times.</li>
<li><em>syn_counts</em> is the total number of unique synonymous codon mutations.</li>
<li>The next four elements in the tuple (prefixed with <em>multi_nt_</em>)
are like the four above but <strong>only</strong> for multi-nucleotide 
codon mutations.</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="utils.NaturalSort">
<tt class="descclassname">utils.</tt><tt class="descname">NaturalSort</tt><big>(</big><em>xlist</em><big>)</big><a class="headerlink" href="#utils.NaturalSort" title="Permalink to this definition">¶</a></dt>
<dd><p>Performs a natural sort (alphanumeric sort) in a list of strings.</p>
<p><em>xlist</em> is a list of strings. Sorts them, reading the integer portion
as numbers. Currently works for integers, but not floats (anything 
including or after a decimal is handled as a string, not number).</p>
<p>Any entries that are integers or floats are converted to strings.</p>
<p>This function simply returns <em>None</em>, but <em>xlist</em> is sorted in place,
with the integer/float to string conversion if relevant.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">xlist</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;1&#39;</span><span class="p">,</span> <span class="s">&#39;20&#39;</span><span class="p">,</span> <span class="s">&#39;100&#39;</span><span class="p">,</span> <span class="s">&#39;5A&#39;</span><span class="p">,</span> <span class="s">&#39;51A&#39;</span><span class="p">,</span> <span class="s">&#39;5&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">NaturalSort</span><span class="p">(</span><span class="n">xlist</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">xlist</span>
<span class="go">[&#39;1&#39;, &#39;5&#39;, &#39;5A&#39;, &#39;20&#39;, &#39;51A&#39;, &#39;100&#39;]</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">xlist</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="s">&#39;20&#39;</span><span class="p">,</span> <span class="mi">100</span><span class="p">,</span> <span class="s">&#39;5A&#39;</span><span class="p">,</span> <span class="s">&#39;51A&#39;</span><span class="p">,</span> <span class="s">&#39;5&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">NaturalSort</span><span class="p">(</span><span class="n">xlist</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">xlist</span>
<span class="go">[&#39;1&#39;, &#39;5&#39;, &#39;5A&#39;, &#39;20&#39;, &#39;51A&#39;, &#39;100&#39;]</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="utils.Pref_or_DiffPref">
<tt class="descclassname">utils.</tt><tt class="descname">Pref_or_DiffPref</tt><big>(</big><em>data</em>, <em>tol=0.001</em><big>)</big><a class="headerlink" href="#utils.Pref_or_DiffPref" title="Permalink to this definition">¶</a></dt>
<dd><p>Determines if data represent preferences or differential preferences.</p>
<p><em>data</em> is a dictionary keyed by site, with each entry a dictionary
keyed by character giving the preference / differential preference
for that character.</p>
<p>If values for each site are &gt;= 0 and &lt;= 1 and sum to one (within <em>tol</em>), then
returns &#8216;preferences&#8217;.</p>
<p>If values are &lt;= 1 and sum to zero (within <em>tol</em>), then returns
&#8216;diffprefs&#8217;.</p>
<p>Otherwise returns &#8216;neither preferences nor diffprefs&#8217;</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">data</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;1&#39;</span><span class="p">:{</span><span class="s">&#39;A&#39;</span><span class="p">:</span><span class="mf">0.5</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">:</span><span class="mf">0.1</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">:</span><span class="mf">0.2</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">:</span><span class="mf">0.2</span><span class="p">}}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">Pref_or_DiffPref</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
<span class="go">preferences</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">data</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;1&#39;</span><span class="p">:{</span><span class="s">&#39;A&#39;</span><span class="p">:</span><span class="o">-</span><span class="mf">0.5</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">:</span><span class="mf">0.1</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">:</span><span class="mf">0.2</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">:</span><span class="mf">0.2</span><span class="p">}}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">Pref_or_DiffPref</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
<span class="go">diffprefs</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">data</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;1&#39;</span><span class="p">:{</span><span class="s">&#39;A&#39;</span><span class="p">:</span><span class="o">-</span><span class="mf">0.5</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">:</span><span class="mf">0.1</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">:</span><span class="mf">0.2</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">:</span><span class="mf">0.1</span><span class="p">}}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">Pref_or_DiffPref</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
<span class="go">neither preferences nor diffprefs</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="utils.PrefsToEnrichments">
<tt class="descclassname">utils.</tt><tt class="descname">PrefsToEnrichments</tt><big>(</big><em>wts</em>, <em>pi_means</em>, <em>include_wt</em><big>)</big><a class="headerlink" href="#utils.PrefsToEnrichments" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts site-specific preferences to enrichment ratios.</p>
<p><em>wts</em> and <em>pi_means</em> have the same meaning as the arguments by the
same name returned by <em>file_io.ReadPreferences</em>.</p>
<p>If <em>include_wt</em> is <em>True</em>, then the returned dictionary includes
ratios for the wildtype characters (these are always one). If it is
<em>False</em>, then ratios for the wildtype characters are not included.</p>
<p>The return value is a dictionary <em>enrichments</em>, where <em>enrichments[r][x]</em>
returns the enrichment ratio for character <em>x</em> at site <em>r</em> for all 
sites and characters in <em>pi_means</em>.</p>
<p>The enrichment ratio <span class="math">\(\phi_{r,x}\)</span> of site <span class="math">\(r\)</span> for 
character <span class="math">\(x\)</span> is defined in terms of the preference
<span class="math">\(\pi_{r,x}\)</span> as</p>
<div class="math">
\[\phi_{r,x} = \frac{\pi_{r,x}}{\pi_{r,\operatorname{wt}\left(r\right)}}\]</div>
<p>where <span class="math">\(\operatorname{wt}\left(r\right)\)</span> is the wildtype residue
at site <span class="math">\(r\)</span>.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">wts</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;1&#39;</span><span class="p">:</span><span class="s">&#39;A&#39;</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pi_means</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;1&#39;</span><span class="p">:{</span><span class="s">&#39;A&#39;</span><span class="p">:</span><span class="mf">0.4</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">:</span><span class="mf">0.45</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">:</span><span class="mf">0.1</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">:</span><span class="mf">0.05</span><span class="p">}}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">enrichments</span> <span class="o">=</span> <span class="n">PrefsToEnrichments</span><span class="p">(</span><span class="n">wts</span><span class="p">,</span> <span class="n">pi_means</span><span class="p">,</span> <span class="n">include_wt</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="s">&quot;</span><span class="si">%.3f</span><span class="s">&quot;</span> <span class="o">%</span> <span class="n">enrichments</span><span class="p">[</span><span class="s">&#39;1&#39;</span><span class="p">][</span><span class="s">&#39;A&#39;</span><span class="p">]</span>
<span class="go">1.000</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="s">&quot;</span><span class="si">%.3f</span><span class="s">&quot;</span> <span class="o">%</span> <span class="n">enrichments</span><span class="p">[</span><span class="s">&#39;1&#39;</span><span class="p">][</span><span class="s">&#39;C&#39;</span><span class="p">]</span>
<span class="go">1.125</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="s">&quot;</span><span class="si">%.3f</span><span class="s">&quot;</span> <span class="o">%</span> <span class="n">enrichments</span><span class="p">[</span><span class="s">&#39;1&#39;</span><span class="p">][</span><span class="s">&#39;G&#39;</span><span class="p">]</span>
<span class="go">0.250</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="s">&quot;</span><span class="si">%.3f</span><span class="s">&quot;</span> <span class="o">%</span> <span class="n">enrichments</span><span class="p">[</span><span class="s">&#39;1&#39;</span><span class="p">][</span><span class="s">&#39;T&#39;</span><span class="p">]</span>
<span class="go">0.125</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">enrichments</span> <span class="o">=</span> <span class="n">PrefsToEnrichments</span><span class="p">(</span><span class="n">wts</span><span class="p">,</span> <span class="n">pi_means</span><span class="p">,</span> <span class="n">include_wt</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="s">&#39;A&#39;</span> <span class="ow">in</span> <span class="n">enrichments</span>
<span class="go">False</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="utils.RMS">
<tt class="descclassname">utils.</tt><tt class="descname">RMS</tt><big>(</big><em>xlist</em><big>)</big><a class="headerlink" href="#utils.RMS" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes root-mean-square value of entries in <em>xlist</em>.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">xlist</span> <span class="o">=</span> <span class="p">[</span><span class="o">-</span><span class="mf">0.2</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">,</span> <span class="mf">0.03</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="s">&quot;</span><span class="si">%.5f</span><span class="s">&quot;</span> <span class="o">%</span> <span class="n">RMS</span><span class="p">(</span><span class="n">xlist</span><span class="p">)</span>
<span class="go">0.27427</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="utils.RemoveStopFromDiffPrefs">
<tt class="descclassname">utils.</tt><tt class="descname">RemoveStopFromDiffPrefs</tt><big>(</big><em>deltapi_means</em><big>)</big><a class="headerlink" href="#utils.RemoveStopFromDiffPrefs" title="Permalink to this definition">¶</a></dt>
<dd><p>Removes stop codon and rescales differential preferences to sum to zero.</p>
<p>Stop codons are denoted by <tt class="docutils literal"><span class="pre">*</span></tt>.</p>
<p>Essentially, the preference assigned to the stop codon is removed
and all other preferences are scaled by an additive factor
so they sum to zero.</p>
<p>If there is not a stop codon present, then nothing is done.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">aas</span> <span class="o">=</span> <span class="n">dms_tools</span><span class="o">.</span><span class="n">aminoacids_withstop</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">deltapi_means</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;1&#39;</span><span class="p">:</span><span class="nb">dict</span><span class="p">([(</span><span class="n">aa</span><span class="p">,</span> <span class="mf">1.0</span> <span class="o">/</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">aas</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">))</span> <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">aas</span> <span class="k">if</span> <span class="n">aa</span> <span class="o">!=</span> <span class="s">&#39;*&#39;</span><span class="p">])}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">deltapi_means</span><span class="p">[</span><span class="s">&#39;1&#39;</span><span class="p">][</span><span class="s">&#39;*&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="mf">1.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">nostop_deltapi_means</span> <span class="o">=</span> <span class="n">RemoveStopFromDiffPrefs</span><span class="p">(</span><span class="n">deltapi_means</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">deltapi_means</span> <span class="o">==</span> <span class="n">nostop_deltapi_means</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">([</span><span class="s">&quot;</span><span class="si">%.3f</span><span class="s">&quot;</span> <span class="o">%</span> <span class="n">nostop_deltapi_means</span><span class="p">[</span><span class="s">&#39;1&#39;</span><span class="p">][</span><span class="n">aa</span><span class="p">]</span> <span class="o">==</span> <span class="s">&#39;0.000&#39;</span> <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">aas</span> <span class="k">if</span> <span class="n">aa</span> <span class="o">!=</span> <span class="s">&#39;*&#39;</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="s">&#39;*&#39;</span> <span class="ow">in</span> <span class="n">nostop_deltapi_means</span><span class="p">[</span><span class="s">&#39;1&#39;</span><span class="p">]</span>
<span class="go">False</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">aas</span> <span class="o">=</span> <span class="n">dms_tools</span><span class="o">.</span><span class="n">aminoacids_withstop</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">deltapi_means</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;1&#39;</span><span class="p">:</span><span class="nb">dict</span><span class="p">([(</span><span class="n">aa</span><span class="p">,</span> <span class="mf">1.0</span> <span class="o">/</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">aas</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">))</span> <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">aas</span><span class="p">[</span><span class="mi">1</span> <span class="p">:</span> <span class="p">]])}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">deltapi_means</span><span class="p">[</span><span class="s">&#39;1&#39;</span><span class="p">][</span><span class="n">aas</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">=</span> <span class="o">-</span><span class="mf">1.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">nostop_deltapi_means</span> <span class="o">=</span> <span class="n">RemoveStopFromDiffPrefs</span><span class="p">(</span><span class="n">deltapi_means</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">deltapi_means</span> <span class="o">==</span> <span class="n">nostop_deltapi_means</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="s">&#39;*&#39;</span> <span class="ow">in</span> <span class="n">nostop_deltapi_means</span><span class="p">[</span><span class="s">&#39;1&#39;</span><span class="p">]</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="s">&#39;</span><span class="si">%.5f</span><span class="s">&#39;</span> <span class="o">%</span> <span class="n">nostop_deltapi_means</span><span class="p">[</span><span class="s">&#39;1&#39;</span><span class="p">][</span><span class="n">aas</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span>
<span class="go">-0.99750</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="s">&#39;</span><span class="si">%.5f</span><span class="s">&#39;</span> <span class="o">%</span> <span class="n">nostop_deltapi_means</span><span class="p">[</span><span class="s">&#39;1&#39;</span><span class="p">][</span><span class="n">aas</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span>
<span class="go">0.05250</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">aas</span> <span class="o">=</span> <span class="n">dms_tools</span><span class="o">.</span><span class="n">aminoacids_nostop</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">deltapi_means</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;1&#39;</span><span class="p">:</span><span class="nb">dict</span><span class="p">([(</span><span class="n">aa</span><span class="p">,</span> <span class="mf">1.0</span> <span class="o">/</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">aas</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">))</span> <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">aas</span><span class="p">[</span><span class="mi">1</span> <span class="p">:</span> <span class="p">]])}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">deltapi_means</span><span class="p">[</span><span class="s">&#39;1&#39;</span><span class="p">][</span><span class="n">aas</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">=</span> <span class="o">-</span><span class="mf">1.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">nostop_deltapi_means</span> <span class="o">=</span> <span class="n">RemoveStopFromDiffPrefs</span><span class="p">(</span><span class="n">deltapi_means</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">deltapi_means</span> <span class="o">==</span> <span class="n">nostop_deltapi_means</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="utils.RemoveStopFromPreferences">
<tt class="descclassname">utils.</tt><tt class="descname">RemoveStopFromPreferences</tt><big>(</big><em>pi_means</em><big>)</big><a class="headerlink" href="#utils.RemoveStopFromPreferences" title="Permalink to this definition">¶</a></dt>
<dd><p>Removes stop codon and renormalizes preference to sum to one.</p>
<p>Stop codons are denoted by <tt class="docutils literal"><span class="pre">*</span></tt>.</p>
<p>Essentially, the preference assigned to the stop codon is removed
and all other preferences are scaled by a multiplicative factor
<span class="math">\(\ge 1\)</span> until they sum to one.</p>
<p>If there is not a stop codon present, then nothing is done.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">aas</span> <span class="o">=</span> <span class="n">dms_tools</span><span class="o">.</span><span class="n">aminoacids_withstop</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pi_means</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;1&#39;</span><span class="p">:</span><span class="nb">dict</span><span class="p">([(</span><span class="n">aa</span><span class="p">,</span> <span class="mf">1.0</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">aas</span><span class="p">))</span> <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">aas</span><span class="p">]),</span> <span class="s">&#39;2&#39;</span><span class="p">:</span><span class="nb">dict</span><span class="p">([(</span><span class="n">aas</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="mf">0.9</span><span class="p">)]</span> <span class="o">+</span> <span class="p">[(</span><span class="n">aa</span><span class="p">,</span> <span class="mf">0.1</span> <span class="o">/</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">aas</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">))</span> <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">aas</span><span class="p">[</span><span class="mi">1</span> <span class="p">:</span> <span class="p">]])}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">nostop_pi_means</span> <span class="o">=</span> <span class="n">RemoveStopFromPreferences</span><span class="p">(</span><span class="n">pi_means</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pi_means</span> <span class="o">==</span> <span class="n">nostop_pi_means</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">([</span><span class="s">&quot;</span><span class="si">%.3f</span><span class="s">&quot;</span> <span class="o">%</span> <span class="n">nostop_pi_means</span><span class="p">[</span><span class="s">&#39;1&#39;</span><span class="p">][</span><span class="n">aa</span><span class="p">]</span> <span class="o">==</span> <span class="s">&#39;0.050&#39;</span> <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">aas</span> <span class="k">if</span> <span class="n">aa</span> <span class="o">!=</span> <span class="s">&#39;*&#39;</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="s">&#39;*&#39;</span> <span class="ow">in</span> <span class="n">nostop_pi_means</span><span class="p">[</span><span class="s">&#39;1&#39;</span><span class="p">]</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="s">&#39;*&#39;</span> <span class="ow">in</span> <span class="n">nostop_pi_means</span><span class="p">[</span><span class="s">&#39;2&#39;</span><span class="p">]</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="s">&#39;</span><span class="si">%.4f</span><span class="s">&#39;</span> <span class="o">%</span> <span class="n">nostop_pi_means</span><span class="p">[</span><span class="s">&#39;2&#39;</span><span class="p">][</span><span class="n">aas</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span>
<span class="go">0.9045</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="s">&#39;</span><span class="si">%.6f</span><span class="s">&#39;</span> <span class="o">%</span> <span class="n">nostop_pi_means</span><span class="p">[</span><span class="s">&#39;2&#39;</span><span class="p">][</span><span class="n">aas</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span>
<span class="go">0.005025</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">aas</span> <span class="o">=</span> <span class="n">dms_tools</span><span class="o">.</span><span class="n">aminoacids_nostop</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pi_means</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;1&#39;</span><span class="p">:</span><span class="nb">dict</span><span class="p">([(</span><span class="n">aa</span><span class="p">,</span> <span class="mf">1.0</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">aas</span><span class="p">))</span> <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">aas</span><span class="p">]),</span> <span class="s">&#39;2&#39;</span><span class="p">:</span><span class="nb">dict</span><span class="p">([(</span><span class="n">aas</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="mf">0.9</span><span class="p">)]</span> <span class="o">+</span> <span class="p">[(</span><span class="n">aa</span><span class="p">,</span> <span class="mf">0.1</span> <span class="o">/</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">aas</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">))</span> <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">aas</span><span class="p">[</span><span class="mi">1</span> <span class="p">:</span> <span class="p">]])}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">nostop_pi_means</span> <span class="o">=</span> <span class="n">RemoveStopFromPreferences</span><span class="p">(</span><span class="n">pi_means</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pi_means</span> <span class="o">==</span> <span class="n">nostop_pi_means</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="utils.SiteEntropy">
<tt class="descclassname">utils.</tt><tt class="descname">SiteEntropy</tt><big>(</big><em>xlist</em>, <em>base=2.0</em><big>)</big><a class="headerlink" href="#utils.SiteEntropy" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes site entropy to log base <em>base</em>.</p>
<p><em>xlist</em> should be a list of numbers <span class="math">\(\ge 0\)</span> that
sum to one.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="s">&quot;</span><span class="si">%.3f</span><span class="s">&quot;</span> <span class="o">%</span> <span class="n">SiteEntropy</span><span class="p">([</span><span class="mf">1.0</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">])</span>
<span class="go">0.000</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="s">&quot;</span><span class="si">%.3f</span><span class="s">&quot;</span> <span class="o">%</span> <span class="n">SiteEntropy</span><span class="p">([</span><span class="mf">0.5</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">])</span>
<span class="go">1.000</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="utils.SumCodonToAA">
<tt class="descclassname">utils.</tt><tt class="descname">SumCodonToAA</tt><big>(</big><em>codondict</em>, <em>includestop=True</em><big>)</big><a class="headerlink" href="#utils.SumCodonToAA" title="Permalink to this definition">¶</a></dt>
<dd><p>Sums all codon entries for each amino acid.</p>
<p><em>codondict</em> is a dictionary keyed by all codons (<em>dms_tools.codons</em>).
The values should be numbers.</p>
<p><em>includestop</em> specifies whether we also sum values for stop codons to
give an entry for <tt class="docutils literal"><span class="pre">*</span></tt> (the stop codon). Do this only if has default
value of <em>True</em>.</p>
<p>The return value is <em>aadict</em> which is keyed by all amino acids and
has values equal to the sum of the entries for the encoding
codons in <em>codondict</em>.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">codondict</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">([(</span><span class="n">codon</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span> <span class="k">for</span> <span class="n">codon</span> <span class="ow">in</span> <span class="n">dms_tools</span><span class="o">.</span><span class="n">codons</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">codondict</span><span class="p">[</span><span class="s">&#39;GGG&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mf">0.1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">codondict</span><span class="p">[</span><span class="s">&#39;GGA&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mf">0.2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">codondict</span><span class="p">[</span><span class="s">&#39;TGA&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mf">0.3</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">codondict</span><span class="p">[</span><span class="s">&#39;CCC&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mf">0.4</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aadict</span> <span class="o">=</span> <span class="n">SumCodonToAA</span><span class="p">(</span><span class="n">codondict</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="s">&quot;</span><span class="si">%.2f</span><span class="s">&quot;</span> <span class="o">%</span> <span class="n">aadict</span><span class="p">[</span><span class="s">&#39;G&#39;</span><span class="p">]</span>
<span class="go">0.30</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="s">&quot;</span><span class="si">%.2f</span><span class="s">&quot;</span> <span class="o">%</span> <span class="n">aadict</span><span class="p">[</span><span class="s">&#39;P&#39;</span><span class="p">]</span>
<span class="go">0.40</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="s">&quot;</span><span class="si">%.2f</span><span class="s">&quot;</span> <span class="o">%</span> <span class="n">aadict</span><span class="p">[</span><span class="s">&#39;*&#39;</span><span class="p">]</span>
<span class="go">0.30</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="s">&quot;</span><span class="si">%.2f</span><span class="s">&quot;</span> <span class="o">%</span> <span class="n">aadict</span><span class="p">[</span><span class="s">&#39;Y&#39;</span><span class="p">]</span>
<span class="go">0.00</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aadict</span> <span class="o">=</span> <span class="n">SumCodonToAA</span><span class="p">(</span><span class="n">codondict</span><span class="p">,</span> <span class="n">includestop</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="s">&quot;</span><span class="si">%.2f</span><span class="s">&quot;</span> <span class="o">%</span> <span class="n">aadict</span><span class="p">[</span><span class="s">&#39;G&#39;</span><span class="p">]</span>
<span class="go">0.30</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="s">&quot;</span><span class="si">%.2f</span><span class="s">&quot;</span> <span class="o">%</span> <span class="n">aadict</span><span class="p">[</span><span class="s">&#39;P&#39;</span><span class="p">]</span>
<span class="go">0.40</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="s">&quot;</span><span class="si">%.2f</span><span class="s">&quot;</span> <span class="o">%</span> <span class="n">aadict</span><span class="p">[</span><span class="s">&#39;*&#39;</span><span class="p">]</span>
<span class="gt">Traceback (most recent call last):</span>
    <span class="o">...</span>
<span class="gr">KeyError</span>: <span class="n">&#39;*&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="s">&quot;</span><span class="si">%.2f</span><span class="s">&quot;</span> <span class="o">%</span> <span class="n">aadict</span><span class="p">[</span><span class="s">&#39;Y&#39;</span><span class="p">]</span>
<span class="go">0.00</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="utils.SumPrefsDiffPrefs">
<tt class="descclassname">utils.</tt><tt class="descname">SumPrefsDiffPrefs</tt><big>(</big><em>datalist</em>, <em>minus=</em><span class="optional">[</span><span class="optional">]</span><big>)</big><a class="headerlink" href="#utils.SumPrefsDiffPrefs" title="Permalink to this definition">¶</a></dt>
<dd><p>Sums preferences and differential preferences.</p>
<p><em>datalist</em> is a list of preferences or differential preferences, which must be keyed
by the same sets of sites and characters.</p>
<p><em>minus</em> is an optional list that is <strong>subtracted</strong> rather than added in the sum.</p>
<p>Returns a dictionary giving the sums.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">prefs</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;1&#39;</span><span class="p">:{</span><span class="s">&#39;A&#39;</span><span class="p">:</span><span class="mf">0.2</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">:</span><span class="mf">0.1</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">:</span><span class="mf">0.5</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">:</span><span class="mf">0.2</span><span class="p">}}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">diffprefs</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;1&#39;</span><span class="p">:{</span><span class="s">&#39;A&#39;</span><span class="p">:</span><span class="o">-</span><span class="mf">0.1</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">:</span><span class="mf">0.3</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">:</span><span class="o">-</span><span class="mf">0.2</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">:</span><span class="mf">0.0</span><span class="p">}}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">datasum</span> <span class="o">=</span> <span class="n">SumPrefsDiffPrefs</span><span class="p">([</span><span class="n">prefs</span><span class="p">,</span> <span class="n">diffprefs</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="s">&#39;</span><span class="si">%.2f</span><span class="s">&#39;</span> <span class="o">%</span> <span class="n">datasum</span><span class="p">[</span><span class="s">&#39;1&#39;</span><span class="p">][</span><span class="s">&#39;A&#39;</span><span class="p">]</span>
<span class="go">0.10</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="s">&#39;</span><span class="si">%.2f</span><span class="s">&#39;</span> <span class="o">%</span> <span class="n">datasum</span><span class="p">[</span><span class="s">&#39;1&#39;</span><span class="p">][</span><span class="s">&#39;T&#39;</span><span class="p">]</span>
<span class="go">0.40</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="s">&#39;</span><span class="si">%.2f</span><span class="s">&#39;</span> <span class="o">%</span> <span class="n">datasum</span><span class="p">[</span><span class="s">&#39;1&#39;</span><span class="p">][</span><span class="s">&#39;C&#39;</span><span class="p">]</span>
<span class="go">0.30</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="s">&#39;</span><span class="si">%.2f</span><span class="s">&#39;</span> <span class="o">%</span> <span class="n">datasum</span><span class="p">[</span><span class="s">&#39;1&#39;</span><span class="p">][</span><span class="s">&#39;G&#39;</span><span class="p">]</span>
<span class="go">0.20</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">minusprefs</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;1&#39;</span><span class="p">:{</span><span class="s">&#39;A&#39;</span><span class="p">:</span><span class="mf">0.1</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">:</span><span class="mf">0.1</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">:</span><span class="o">-</span><span class="mf">0.1</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">:</span><span class="o">-</span><span class="mf">0.1</span><span class="p">}}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">datasum</span> <span class="o">=</span> <span class="n">SumPrefsDiffPrefs</span><span class="p">([</span><span class="n">prefs</span><span class="p">,</span> <span class="n">diffprefs</span><span class="p">],</span> <span class="n">minus</span><span class="o">=</span><span class="p">[</span><span class="n">minusprefs</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="s">&#39;</span><span class="si">%.2f</span><span class="s">&#39;</span> <span class="o">%</span> <span class="n">datasum</span><span class="p">[</span><span class="s">&#39;1&#39;</span><span class="p">][</span><span class="s">&#39;A&#39;</span><span class="p">]</span>
<span class="go">0.00</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="s">&#39;</span><span class="si">%.2f</span><span class="s">&#39;</span> <span class="o">%</span> <span class="n">datasum</span><span class="p">[</span><span class="s">&#39;1&#39;</span><span class="p">][</span><span class="s">&#39;T&#39;</span><span class="p">]</span>
<span class="go">0.30</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="s">&#39;</span><span class="si">%.2f</span><span class="s">&#39;</span> <span class="o">%</span> <span class="n">datasum</span><span class="p">[</span><span class="s">&#39;1&#39;</span><span class="p">][</span><span class="s">&#39;C&#39;</span><span class="p">]</span>
<span class="go">0.40</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="s">&#39;</span><span class="si">%.2f</span><span class="s">&#39;</span> <span class="o">%</span> <span class="n">datasum</span><span class="p">[</span><span class="s">&#39;1&#39;</span><span class="p">][</span><span class="s">&#39;G&#39;</span><span class="p">]</span>
<span class="go">0.30</span>
</pre></div>
</div>
</dd></dl>

</div>
</div>
<span class="target" id="module-parsearguments"></span><div class="section" id="parsearguments-module">
<h2><tt class="docutils literal"><span class="pre">parsearguments</span></tt> module<a class="headerlink" href="#parsearguments-module" title="Permalink to this headline">¶</a></h2>
<p>Module that parses command-line arguments for the scripts in <tt class="docutils literal"><span class="pre">dms_tools</span></tt>.</p>
<p>Written by Jesse Bloom.</p>
<div class="section" id="defined-in-this-module">
<h3>Defined in this module<a class="headerlink" href="#defined-in-this-module" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><em>ArgumentParserNoArgHelp</em> : argument parser that prints help when no args</li>
<li><em>FloatGreaterThanZero</em> : parses whether string is float greater than zero.</li>
<li><em>FloatBetweenZeroAndOne</em> : parses whether is float between 0 and 1.</li>
<li><em>FloatBetweenHalfAndOne</em> : parses whether is float &gt; 0.5 and &lt;= 1.</li>
<li><em>NonNegativeInt</em> : parses whether string is non-negative integer.</li>
<li><em>IntGreaterEqual2</em> : parses whether string is integer &gt;= 2.</li>
<li><em>ExistingFile</em> : parses whether string gives an existing file.</li>
<li><em>PDF_File</em> : parses whether string gives a PDF file name.</li>
<li><em>CommaSeparatedFASTQFiles</em> : parses list of FASTQ files.</li>
<li><em>AlignSpecs</em> : parses alignment specs for <em>BarcodedSubampliconsParser</em></li>
<li><em>AlignprefixName</em> : parses alignment prefixes/names for <em>SummarizeAlignmentsParser</em></li>
<li><em>InferPrefsParser</em> : parser for <tt class="docutils literal"><span class="pre">dms_inferprefs</span></tt></li>
<li><em>InferDiffPrefsParser</em> : parser for <tt class="docutils literal"><span class="pre">dms_inferdiffprefs</span></tt></li>
<li><em>MergeParser</em> : parser for <tt class="docutils literal"><span class="pre">dms_merge</span></tt></li>
<li><em>EditSitesParser</em> : parsers for <tt class="docutils literal"><span class="pre">dms_editsites</span></tt></li>
<li><em>CorrelateParser</em> : parser for <tt class="docutils literal"><span class="pre">dms_correlate</span></tt></li>
<li><em>LogoPlotParser</em> : parser for <tt class="docutils literal"><span class="pre">dms_logoplot</span></tt></li>
<li><em>BarcodedSubampliconsParser</em> : parser for <tt class="docutils literal"><span class="pre">dms_barcodedsubamplicons</span></tt></li>
<li><em>SummarizeAlignmentsParser</em> : parser for <tt class="docutils literal"><span class="pre">dms_summarizealignments</span></tt></li>
<li><em>SmartHelpFormatter</em> : new class for formatting argument helps.</li>
</ul>
</div>
<div class="section" id="detailed-documentation">
<h3>Detailed documentation<a class="headerlink" href="#detailed-documentation" title="Permalink to this headline">¶</a></h3>
<dl class="function">
<dt id="parsearguments.AlignSpecs">
<tt class="descclassname">parsearguments.</tt><tt class="descname">AlignSpecs</tt><big>(</big><em>alignspecs</em><big>)</big><a class="headerlink" href="#parsearguments.AlignSpecs" title="Permalink to this definition">¶</a></dt>
<dd><p>Parses alignment specs for <em>BarcodedSubampliconsParser</em>.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">AlignSpecs</span><span class="p">(</span><span class="s">&#39;1,300,10,12&#39;</span><span class="p">)</span>
<span class="go">(1, 300, 10, 12)</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="parsearguments.AlignprefixName">
<tt class="descclassname">parsearguments.</tt><tt class="descname">AlignprefixName</tt><big>(</big><em>alignprefixname</em><big>)</big><a class="headerlink" href="#parsearguments.AlignprefixName" title="Permalink to this definition">¶</a></dt>
<dd><p>Parses prefix / name for <em>SummarizeAlignmentsParser</em>.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">AlignprefixName</span><span class="p">(</span><span class="s">&#39;replicate_1/replicate_1_mutDNA_,mutDNA&#39;</span><span class="p">)</span>
<span class="go">(&#39;replicate_1/replicate_1_mutDNA_&#39;, &#39;mutDNA&#39;)</span>
</pre></div>
</div>
</dd></dl>

<dl class="class">
<dt id="parsearguments.ArgumentParserNoArgHelp">
<em class="property">class </em><tt class="descclassname">parsearguments.</tt><tt class="descname">ArgumentParserNoArgHelp</tt><big>(</big><em>prog=None</em>, <em>usage=None</em>, <em>description=None</em>, <em>epilog=None</em>, <em>version=None</em>, <em>parents=</em>, <span class="optional">[</span><span class="optional">]</span><em>formatter_class=&lt;class 'argparse.HelpFormatter'&gt;</em>, <em>prefix_chars='-'</em>, <em>fromfile_prefix_chars=None</em>, <em>argument_default=None</em>, <em>conflict_handler='error'</em>, <em>add_help=True</em><big>)</big><a class="headerlink" href="#parsearguments.ArgumentParserNoArgHelp" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <tt class="xref py py-class docutils literal"><span class="pre">argparse.ArgumentParser</span></tt></p>
<p>Like <em>argparse.ArgumentParser</em>, but prints help when no arguments.</p>
<dl class="method">
<dt id="parsearguments.ArgumentParserNoArgHelp.error">
<tt class="descname">error</tt><big>(</big><em>message</em><big>)</big><a class="headerlink" href="#parsearguments.ArgumentParserNoArgHelp.error" title="Permalink to this definition">¶</a></dt>
<dd><p>Prints error message, then help.</p>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="parsearguments.BarcodedSubampliconsParser">
<tt class="descclassname">parsearguments.</tt><tt class="descname">BarcodedSubampliconsParser</tt><big>(</big><big>)</big><a class="headerlink" href="#parsearguments.BarcodedSubampliconsParser" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns <em>argparse.ArgumentParser</em> for <tt class="docutils literal"><span class="pre">dms_barcodedsubamplicons</span></tt> script.</p>
</dd></dl>

<dl class="function">
<dt id="parsearguments.CommaSeparatedFASTQFiles">
<tt class="descclassname">parsearguments.</tt><tt class="descname">CommaSeparatedFASTQFiles</tt><big>(</big><em>files</em><big>)</big><a class="headerlink" href="#parsearguments.CommaSeparatedFASTQFiles" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns name of comma-separated FASTQ files.</p>
<p>Checks that all files exist. Returns error if files
cannot be found or if they are not .fastq or .fastq.gz
Otherwise returns list of files.</p>
</dd></dl>

<dl class="function">
<dt id="parsearguments.CorrelateParser">
<tt class="descclassname">parsearguments.</tt><tt class="descname">CorrelateParser</tt><big>(</big><big>)</big><a class="headerlink" href="#parsearguments.CorrelateParser" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns <em>argparse.ArgumentParser</em> for <tt class="docutils literal"><span class="pre">dms_correlate</span></tt> script.</p>
</dd></dl>

<dl class="function">
<dt id="parsearguments.EditSitesParser">
<tt class="descclassname">parsearguments.</tt><tt class="descname">EditSitesParser</tt><big>(</big><big>)</big><a class="headerlink" href="#parsearguments.EditSitesParser" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="parsearguments.ExistingFile">
<tt class="descclassname">parsearguments.</tt><tt class="descname">ExistingFile</tt><big>(</big><em>fname</em><big>)</big><a class="headerlink" href="#parsearguments.ExistingFile" title="Permalink to this definition">¶</a></dt>
<dd><p>If <em>fname</em> is name of an existing file return it, otherwise an error.</p>
<p>It is also acceptable for <em>fname</em> to be the string &#8220;none&#8221;.</p>
</dd></dl>

<dl class="function">
<dt id="parsearguments.FloatBetweenHalfAndOne">
<tt class="descclassname">parsearguments.</tt><tt class="descname">FloatBetweenHalfAndOne</tt><big>(</big><em>x</em><big>)</big><a class="headerlink" href="#parsearguments.FloatBetweenHalfAndOne" title="Permalink to this definition">¶</a></dt>
<dd><p>If <em>x</em> is float &gt; 0.5 or &lt;= 1, returns it, otherwise error.</p>
</dd></dl>

<dl class="function">
<dt id="parsearguments.FloatBetweenZeroAndOne">
<tt class="descclassname">parsearguments.</tt><tt class="descname">FloatBetweenZeroAndOne</tt><big>(</big><em>x</em><big>)</big><a class="headerlink" href="#parsearguments.FloatBetweenZeroAndOne" title="Permalink to this definition">¶</a></dt>
<dd><p>If <em>x</em> is float &gt;= 0 or &lt;= 1, returns it, otherwise error.</p>
</dd></dl>

<dl class="function">
<dt id="parsearguments.FloatGreaterThanZero">
<tt class="descclassname">parsearguments.</tt><tt class="descname">FloatGreaterThanZero</tt><big>(</big><em>x</em><big>)</big><a class="headerlink" href="#parsearguments.FloatGreaterThanZero" title="Permalink to this definition">¶</a></dt>
<dd><p>If <em>x</em> is string for float &gt; 0, returns it, otherwise an error.</p>
<p>Designed based on this: <a class="reference external" href="http://stackoverflow.com/questions/12116685/how-can-i-require-my-python-scripts-argument-to-be-a-float-between-0-0-1-0-usin">http://stackoverflow.com/questions/12116685/how-can-i-require-my-python-scripts-argument-to-be-a-float-between-0-0-1-0-usin</a></p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="s">&quot;</span><span class="si">%.3f</span><span class="s">&quot;</span> <span class="o">%</span> <span class="n">FloatGreaterThanZero</span><span class="p">(</span><span class="s">&#39;0.1&#39;</span><span class="p">)</span>
<span class="go">0.100</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">FloatGreaterThanZero</span><span class="p">(</span><span class="s">&#39;0.0&#39;</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
    <span class="o">...</span>
<span class="gr">ArgumentTypeError</span>: <span class="n">0.0 not a float greater than zero</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">FloatGreaterThanZero</span><span class="p">(</span><span class="s">&#39;hi&#39;</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
    <span class="o">...</span>
<span class="gr">ValueError</span>: <span class="n">could not convert string to float: hi</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="parsearguments.InferDiffPrefsParser">
<tt class="descclassname">parsearguments.</tt><tt class="descname">InferDiffPrefsParser</tt><big>(</big><big>)</big><a class="headerlink" href="#parsearguments.InferDiffPrefsParser" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns <em>argparse.ArgumentParser</em> for <tt class="docutils literal"><span class="pre">dms_inferdiffprefs</span></tt> script.</p>
</dd></dl>

<dl class="function">
<dt id="parsearguments.InferPrefsParser">
<tt class="descclassname">parsearguments.</tt><tt class="descname">InferPrefsParser</tt><big>(</big><big>)</big><a class="headerlink" href="#parsearguments.InferPrefsParser" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns <em>argparse.ArgumentParser</em> for <tt class="docutils literal"><span class="pre">dms_inferprefs</span></tt> script.</p>
</dd></dl>

<dl class="function">
<dt id="parsearguments.IntGreaterEqual2">
<tt class="descclassname">parsearguments.</tt><tt class="descname">IntGreaterEqual2</tt><big>(</big><em>n</em><big>)</big><a class="headerlink" href="#parsearguments.IntGreaterEqual2" title="Permalink to this definition">¶</a></dt>
<dd><p>If <em>n</em> is integer &gt;= 2 returns it, otherwise an error.</p>
</dd></dl>

<dl class="function">
<dt id="parsearguments.LogoPlotParser">
<tt class="descclassname">parsearguments.</tt><tt class="descname">LogoPlotParser</tt><big>(</big><big>)</big><a class="headerlink" href="#parsearguments.LogoPlotParser" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns <em>argparse.ArgumentParser</em> for <tt class="docutils literal"><span class="pre">dms_logoplot</span></tt> script.</p>
</dd></dl>

<dl class="function">
<dt id="parsearguments.MergeParser">
<tt class="descclassname">parsearguments.</tt><tt class="descname">MergeParser</tt><big>(</big><big>)</big><a class="headerlink" href="#parsearguments.MergeParser" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns <em>argparse.ArgumentParser</em> for <tt class="docutils literal"><span class="pre">dms_merge</span></tt> script.</p>
</dd></dl>

<dl class="function">
<dt id="parsearguments.NonNegativeInt">
<tt class="descclassname">parsearguments.</tt><tt class="descname">NonNegativeInt</tt><big>(</big><em>n</em><big>)</big><a class="headerlink" href="#parsearguments.NonNegativeInt" title="Permalink to this definition">¶</a></dt>
<dd><p>If <em>n</em> is non-negative integer returns it, otherwise an error.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="s">&quot;</span><span class="si">%d</span><span class="s">&quot;</span> <span class="o">%</span> <span class="n">NonNegativeInt</span><span class="p">(</span><span class="s">&#39;8&#39;</span><span class="p">)</span>
<span class="go">8</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">NonNegativeInt</span><span class="p">(</span><span class="s">&#39;8.1&#39;</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
   <span class="c">...</span>
<span class="gr">ArgumentTypeError</span>: <span class="n">8.1 is not an integer</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="s">&quot;</span><span class="si">%d</span><span class="s">&quot;</span> <span class="o">%</span> <span class="n">NonNegativeInt</span><span class="p">(</span><span class="s">&#39;0&#39;</span><span class="p">)</span>
<span class="go">0</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">NonNegativeInt</span><span class="p">(</span><span class="s">&#39;-1&#39;</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
   <span class="c">...</span>
<span class="gr">ArgumentTypeError</span>: <span class="n">-1 is not non-negative</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="parsearguments.PDF_File">
<tt class="descclassname">parsearguments.</tt><tt class="descname">PDF_File</tt><big>(</big><em>fname</em><big>)</big><a class="headerlink" href="#parsearguments.PDF_File" title="Permalink to this definition">¶</a></dt>
<dd><p>If <em>fname</em> ends in PDF extension, returns it. Otherwise an error.</p>
</dd></dl>

<dl class="class">
<dt id="parsearguments.SmartHelpFormatter">
<em class="property">class </em><tt class="descclassname">parsearguments.</tt><tt class="descname">SmartHelpFormatter</tt><big>(</big><em>prog</em>, <em>indent_increment=2</em>, <em>max_help_position=24</em>, <em>width=None</em><big>)</big><a class="headerlink" href="#parsearguments.SmartHelpFormatter" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <tt class="xref py py-class docutils literal"><span class="pre">argparse.ArgumentDefaultsHelpFormatter</span></tt></p>
<p>Raw text for help beginning with <tt class="docutils literal"><span class="pre">R|</span></tt>, <em>ArgumentDefaultsHelpFormatter</em> otherwise.</p>
<p>Based on <em>SmartFormatter</em> described at <a class="reference external" href="http://stackoverflow.com/questions/3853722/python-argparse-how-to-insert-newline-in-the-help-text">http://stackoverflow.com/questions/3853722/python-argparse-how-to-insert-newline-in-the-help-text</a>.</p>
<p>This option allows you to specify exact help format by preceding <tt class="docutils literal"><span class="pre">R|</span></tt>.</p>
</dd></dl>

<dl class="function">
<dt id="parsearguments.SummarizeAlignmentsParser">
<tt class="descclassname">parsearguments.</tt><tt class="descname">SummarizeAlignmentsParser</tt><big>(</big><big>)</big><a class="headerlink" href="#parsearguments.SummarizeAlignmentsParser" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns <em>argparse.ArgumentParser</em> for <tt class="docutils literal"><span class="pre">dms_summarizealignments</span></tt>.</p>
</dd></dl>

</div>
</div>
<span class="target" id="module-plot"></span><div class="section" id="plot-module">
<h2><tt class="docutils literal"><span class="pre">plot</span></tt> module<a class="headerlink" href="#plot-module" title="Permalink to this headline">¶</a></h2>
<p>This module uses <tt class="docutils literal"><span class="pre">pylab</span></tt> and <tt class="docutils literal"><span class="pre">matplotlib</span></tt> to make plots. 
The <tt class="docutils literal"><span class="pre">pdf</span></tt> backend is used for <tt class="docutils literal"><span class="pre">matplotlib</span></tt> / <tt class="docutils literal"><span class="pre">pylab</span></tt>.</p>
<p>Written by Jesse Bloom.</p>
<div class="section" id="id5">
<h3>Functions in this module<a class="headerlink" href="#id5" title="Permalink to this headline">¶</a></h3>
<p><cite>PlotCorrelation</cite> : plots correlation between two variables.</p>
<p><cite>Base10Formatter</cite> : formats numbers in base 10</p>
<p><cite>PlotPairedMutFracs</cite> : bar graph of mutation frequencies per sample</p>
<p><cite>PlotMutCountFracs</cite> : plots fraction of mutations that occur &gt;= a given number of times.</p>
<p><cite>PlotDepth</cite> : plots per-site depth along primary sequence</p>
<p><cite>PlotSampleBargraphs</cite> : plots bargraphs of categories for several samples.</p>
</div>
<div class="section" id="id6">
<h3>Function documentation<a class="headerlink" href="#id6" title="Permalink to this headline">¶</a></h3>
<dl class="function">
<dt id="plot.Base10Formatter">
<tt class="descclassname">plot.</tt><tt class="descname">Base10Formatter</tt><big>(</big><em>number</em>, <em>exp_cutoff</em>, <em>exp_decimal_digits</em>, <em>decimal_digits</em><big>)</big><a class="headerlink" href="#plot.Base10Formatter" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts a number into Latex formatting with scientific notation.</p>
<p>Takes a number and converts it to a string that can be shown
in LaTex using math mode. It is converted to scientific notation
if the criteria specified by <cite>exp_cutoff</cite> is met.</p>
<p><cite>number</cite> : the number to be formatted, should be a float or integer.
Currently only works for <cite>number &gt;= 0</cite>.</p>
<p><cite>exp_cutoff</cite> : convert to scientific notation if <tt class="docutils literal"><span class="pre">abs(math.log10(number))</span> 
<span class="pre">&gt;=</span> <span class="pre">exp_cutoff</span></tt>.</p>
<p><cite>exp_decimal_digits</cite> : show this many digits after the decimal if number
is converted to scientific notation.</p>
<p><cite>decimal_digits</cite> : show this many digits after the decimal if number
is NOT converted to scientific notation.</p>
<p>The returned value is a LaTex formatted string. If the number is zero, the
returned string is simply &#8216;0&#8217;.</p>
<p>EXAMPLES:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">Base10Formatter</span><span class="p">(</span><span class="mi">103</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="go">&#39;103.0&#39;</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">Base10Formatter</span><span class="p">(</span><span class="mf">103.0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="go">&#39;1.0 \\times 10^{2}&#39;</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">Base10Formatter</span><span class="p">(</span><span class="mf">103.0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="go">&#39;1.03 \\times 10^{2}&#39;</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">Base10Formatter</span><span class="p">(</span><span class="mf">2892.3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span> 
<span class="go">&#39;2.9 \\times 10^{3}&#39;</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">Base10Formatter</span><span class="p">(</span><span class="mf">0.0</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span> 
<span class="go">&#39;0&#39;</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">Base10Formatter</span><span class="p">(</span><span class="mf">0.012</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="go">&#39;1.2 \\times 10^{-2}&#39;</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">Base10Formatter</span><span class="p">(</span><span class="o">-</span><span class="mf">0.1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
    <span class="o">...</span>
<span class="gr">ValueError</span>: <span class="n">number must be &gt;= 0</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="plot.PlotCorrelation">
<tt class="descclassname">plot.</tt><tt class="descname">PlotCorrelation</tt><big>(</big><em>xs</em>, <em>ys</em>, <em>plotfile</em>, <em>xlabel</em>, <em>ylabel</em>, <em>logx=False</em>, <em>logy=False</em>, <em>corr=None</em>, <em>title=False</em>, <em>alpha=1.0</em>, <em>symmetrize=False</em>, <em>fixaxes=False</em>, <em>additionalxy=</em>, <span class="optional">[</span><span class="optional">]</span><em>bigmargin=0.29</em>, <em>xsize=1.8</em>, <em>r2=False</em><big>)</big><a class="headerlink" href="#plot.PlotCorrelation" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots the correlation between two variables as a scatter plot.</p>
<p>The data is plotted as a scatter plot.</p>
<p>This function uses <tt class="docutils literal"><span class="pre">pylab</span></tt> / <tt class="docutils literal"><span class="pre">matplotlib</span></tt>.</p>
<p>The calling variables use LaTex format for strings. So for
example, &#8216;$10^5$&#8217; will print the LaTex equivalent of this 
string. Similarly, certain raw text strings (such as those
including underscores) will cause problems if you do not
escape the LaTex format meaning. For instance, &#8216;x_label&#8217;
will cause a problem since underscore is not valid
outside of math mode in LaTex, so you would need to use
&#8216;x_label&#8217; to escape the underscore.</p>
<p>CALLING VARIABLES:</p>
<ul class="simple">
<li><em>xs</em> and <em>ys</em> are lists of numbers, with the lists
being of the same length. Entry <em>xs[i]</em> is plotted
on the x-axis against entries <em>ys[i]</em> on the y-axis.</li>
<li><em>plotfile</em> is a string giving the name of the plot PDF file
that we create. It should end in the extension <tt class="docutils literal"><span class="pre">.pdf</span></tt>.
If this plot already exists, it is overwritten.</li>
<li><em>xlabel</em> is a string giving the label placed on the x-axis.</li>
<li><em>ylabel</em> is a string giving the label placed on the y-axis.</li>
<li><em>logx</em> specifies that we log transform the data in <em>xs</em>.
This is <em>False</em> by default; set to <em>True</em> if you want
to log transform (base 10 logarithms) the data.</li>
<li><em>logy</em> is like <em>logx</em>, but for the data in <em>ys</em>.</li>
<li><em>corr</em> specifies that we include a correlation coefficient on the
plot. If it has its default value of <em>None</em>, then no correlation
coefficient is included. Otherwise, <em>corr = (r, p)</em> where <em>r</em> is
the correlation coefficient (displayed with the label <tt class="docutils literal"><span class="pre">R</span></tt>) and <em>p</em>
is the P-value (displayed with the label <tt class="docutils literal"><span class="pre">P</span></tt>).</li>
<li><em>r2</em> : if <em>True</em> and using <em>corr</em>, then display <span class="math">\(R^2\)</span> rather
than <span class="math">\(R\)</span>.</li>
<li><em>title</em> is a string giving the title placed above the plot. 
It can be <em>False</em> if no title is to be used. Otherwise, it should
be the title string (using LaTex formatting, spaces are allowed).
Is <em>False</em> by default.</li>
<li><em>alpha</em> is the transparency of the plotted points. By default
it is one, which means that there is no transparency. If you make
the value closer to zero, then the points become partially transparent.
The rationale is that if you have many overlapping points, making
them partially transparent helps you better see the density.
At any position on the plot, the intensity will be saturated
when there are 1.0 / alpha points plotted. So a reasonable value
of <em>alpha</em> might be something like 0.1.</li>
<li><em>symmetrize</em> is an optional argument that is <em>False</em> by default.
If <em>True</em>, we make the X and Y limits on the plot the same.</li>
<li><em>fixaxes</em> is an optional argument that is <em>False</em> by default.
If <em>True</em>, we fix both the X and Y axes to go from 0 to 1, 
with ticks at 0, 0.5, and 1. If you set this option to <em>True</em>,
then you must set <em>logx</em> and <em>logy</em> to <em>False</em>.</li>
<li><em>additionalxy</em> is an optional list. By default, it is an empty list.
However, you can use it to plot additional data points with a different
color. The main data (specified by <em>xs</em> and <em>ys</em>) is plotted with
blue circles. For each addition set of data that you want to plot,
include a 2-tuple of lists of the form <em>(x2s, y2s)</em> in 
<em>additionalxy</em>. Currently, only one such 2-tuple is allowed
(so only one additional data set). These are plotted as red
triangles, whereas the main data is plotted as blue
circles. The same <em>alpha</em> value set by the <em>alpha</em> parameter applies
to these points as well.</li>
<li><em>bigmargin</em> is an optional argument that is 0.26 by default. It is
the fraction of the plot width taken up by the larger margin, which
is the bottom and left. Make larger if you need more space for axis
labels.</li>
<li><em>xsize</em> is an optional argument that is the width of the plot in inches.
It is 1.8 by default.</li>
</ul>
</dd></dl>

<dl class="function">
<dt id="plot.PlotDepth">
<tt class="descclassname">plot.</tt><tt class="descname">PlotDepth</tt><big>(</big><em>codon_counts</em>, <em>names</em>, <em>plotfile</em>, <em>mutdepth=False</em><big>)</big><a class="headerlink" href="#plot.PlotDepth" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots per-site depth along primary sequence.</p>
<p><cite>codon_counts</cite> : a list of dictionaries giving the codon counts for
each sample as read by <cite>dms_tools.file_io.ReadDMSCounts</cite>.</p>
<p><cite>names</cite> : a list of strings giving the names of the samples
corresponding to each entry in <cite>codon_counts</cite>.</p>
<p><cite>plotfile</cite> : name of the output plot file created by this method
(such as &#8216;plot.pdf&#8217;). The extension must be <tt class="docutils literal"><span class="pre">.pdf</span></tt>.</p>
<p><cite>mutdepth</cite> : Boolean switch, if <em>True</em> then rather than plotting
sequencing depth, we plot per-site mutation rate.</p>
</dd></dl>

<dl class="function">
<dt id="plot.PlotMutCountFracs">
<tt class="descclassname">plot.</tt><tt class="descname">PlotMutCountFracs</tt><big>(</big><em>plotfile</em>, <em>title</em>, <em>names</em>, <em>all_cumulfracs</em>, <em>syn_cumulfracs</em>, <em>all_counts</em>, <em>syn_counts</em>, <em>legendloc</em>, <em>writecounts=True</em><big>)</big><a class="headerlink" href="#plot.PlotMutCountFracs" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots fraction of mutations with &gt;= a given number of counts.</p>
<p>Does this for both all mutations and synonymous mutations. The plots
are placed side by side.</p>
<p>The plots show the fractions of mutations
that are found &gt;= <em>n</em> times for a range of values of <em>n</em>.</p>
<p>CALLING VARIABLES:</p>
<ul>
<li><p class="first"><em>plotfile</em> : string giving name of the created plot file. Must end in
the extension <tt class="docutils literal"><span class="pre">.pdf</span></tt>.</p>
</li>
<li><p class="first"><em>title</em> : string giving the title placed about the plot.</p>
</li>
<li><p class="first"><em>names</em> : a list of strings giving the names of the samples to
plot.</p>
</li>
<li><p class="first"><em>all_cumulfracs</em> : a list of the same length as <em>names</em> giving
the cumulative fractions for all mutations. Each entry should be a list, 
and all of these lists should be the same length. <em>cumulfracs[i][n]</em>
gives the fraction of mutations to sample <em>names[i]</em> that
are found &gt;= <em>n</em> times. The x-axis of the created plot
will go from 0 to <em>len(all_cumulfracs[0]) - 1</em>.</p>
</li>
<li><p class="first"><em>syn_cumulfracs</em> : a list like <em>all_cumulfracs</em> except for synonymous
mutations.</p>
</li>
<li><p class="first"><em>all_counts</em> : integer counts of all mutations (the total number of mutations
used for <em>all_cumulfracs</em>), used to create plot title.</p>
</li>
<li><p class="first"><em>syn_counts</em> : like <em>all_counts</em> but for synonymous mutations.</p>
</li>
<li><p class="first"><em>legendloc</em> : specifies the location of the legend. Should be a string.
Valid values are:</p>
<blockquote>
<div><ul class="simple">
<li><em>bottom</em> : put legend at the bottom of the plot.</li>
<li><em>right</em> : put legend at the right of the plot.</li>
</ul>
</div></blockquote>
</li>
<li><p class="first"><em>writecounts</em> is a Boolean switch specifying whether we include the counts of all
mutations (specified by <em>all_counts</em> and <em>syn_counts</em>) in the plot title. We do
this if <em>writecounts</em> is <em>True</em>, and do not if it is <em>False</em>.</p>
</li>
</ul>
</dd></dl>

<dl class="function">
<dt id="plot.PlotPairedMutFracs">
<tt class="descclassname">plot.</tt><tt class="descname">PlotPairedMutFracs</tt><big>(</big><em>codon_counts</em>, <em>names</em>, <em>plotfile</em><big>)</big><a class="headerlink" href="#plot.PlotPairedMutFracs" title="Permalink to this definition">¶</a></dt>
<dd><p>Makes paired bar graph of mutation fractions per codon.</p>
<p>For each sample, calculates the fraction of codon counts
that represent synonymous / nonsynoymous /
stop codons, and one- / two- / three-nucleotide mutations. Makes
paired stacked bar graphs showing these fractions. The fractions
for all sites in the gene are aggregated together.</p>
<p><cite>codon_counts</cite> : a list of dictionaries giving the codon counts for
each sample as read by <cite>dms_tools.file_io.ReadDMSCounts</cite>.</p>
<p><cite>names</cite> : a list of strings giving the names of the samples
corresponding to each entry in <cite>codon_counts</cite>.</p>
<p><cite>plotfile</cite> : name of the output plot file created by this method
(such as &#8216;plot.pdf&#8217;). The extension must be <tt class="docutils literal"><span class="pre">.pdf</span></tt>.</p>
</dd></dl>

<dl class="function">
<dt id="plot.PlotSampleBargraphs">
<tt class="descclassname">plot.</tt><tt class="descname">PlotSampleBargraphs</tt><big>(</big><em>names</em>, <em>categories</em>, <em>data</em>, <em>plotfile</em>, <em>ylabel</em>, <em>groupbyfirstword=True</em><big>)</big><a class="headerlink" href="#plot.PlotSampleBargraphs" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots bargraph of counts for different samples.</p>
<p><em>names</em> is a list of the names of the different samples.</p>
<p><em>categories</em> is a list of the different categories for each sample.</p>
<p><em>data</em> is a dictionary keyed by each name in <em>names</em>, and that value
is in turn a dictionary keyed by each category in <em>categories</em>.</p>
<p><em>plotfile</em> is the name of the PDF file that we are creating.</p>
<p><em>ylabel</em> is the label placed on the y-axis.</p>
<p><em>groupbyfirstword</em> is a Boolean switch that specifies whether we group 
consecutive categories with the same first word in the string to have
the same color.</p>
</dd></dl>

</div>
</div>
<span class="target" id="module-weblogo"></span><div class="section" id="weblogo-module">
<h2><tt class="docutils literal"><span class="pre">weblogo</span></tt> module<a class="headerlink" href="#weblogo-module" title="Permalink to this headline">¶</a></h2>
<p>Module for making sequence logos with the <em>weblogolib</em> package distributed with
<tt class="docutils literal"><span class="pre">weblogo</span></tt> (<a class="reference external" href="http://weblogo.threeplusone.com/">http://weblogo.threeplusone.com/</a>). This module interfaces with the 
<em>weblogolib</em> API, and so is only known to work with <em>weblogolib</em> version 3.4.</p>
<p>Written by Jesse Bloom.</p>
<div class="section" id="id7">
<h3>Functions in this module<a class="headerlink" href="#id7" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><em>LogoPlot</em> : plots a sequence logo of site-specific preferences or differential preferences.</li>
<li><em>LogoOverlay</em> : makes logo plot overlays, called by <em>LogoPlot</em>.</li>
<li><em>KyteDoolittleColorMapping</em> : maps amino-acid hydrophobicities to colors.</li>
</ul>
</div>
<div class="section" id="id8">
<h3>Function documentation<a class="headerlink" href="#id8" title="Permalink to this headline">¶</a></h3>
<dl class="function">
<dt id="weblogo.KyteDoolittleColorMapping">
<tt class="descclassname">weblogo.</tt><tt class="descname">KyteDoolittleColorMapping</tt><big>(</big><em>maptype='jet'</em>, <em>reverse=True</em><big>)</big><a class="headerlink" href="#weblogo.KyteDoolittleColorMapping" title="Permalink to this definition">¶</a></dt>
<dd><p>Maps amino-acid hydrophobicities to colors.</p>
<p>Uses the Kyte-Doolittle hydrophobicity scale defined by:</p>
<div class="highlight-python"><div class="highlight"><pre>J. Kyte &amp; R. F. Doolittle: 
&quot;A simple method for displaying the hydropathic character of a protein.&quot; 
J Mol Biol, 157, 105-132
</pre></div>
</div>
<p>More positive values indicate higher hydrophobicity, while more negative values
indicate lower hydrophobicity.</p>
<p>The returned variable is the 3-tuple <em>(cmap, mapping_d, mapper)</em>:</p>
<blockquote>
<div><ul class="simple">
<li><em>cmap</em> is a <tt class="docutils literal"><span class="pre">pylab</span></tt> <em>LinearSegmentedColorMap</em> object.</li>
<li><em>mapping_d</em> is a dictionary keyed by the one-letter amino-acid
codes. The values are the colors in CSS2 format (e.g. #FF0000
for red) for that amino acid. The value for a stop codon 
(denoted by a <tt class="docutils literal"><span class="pre">*</span></tt> character) is black (#000000).</li>
<li><em>mapper</em> is the actual <em>pylab.cm.ScalarMappable</em> object.</li>
</ul>
</div></blockquote>
<p>The optional calling argument <em>maptype</em> should specify a valid <tt class="docutils literal"><span class="pre">pylab</span></tt> color map.</p>
<p>The optional calling argument <em>reverse</em> specifies that we set up the color
map so that the most hydrophobic residue comes first (in the Kyte-Doolittle
scale the most hydrophobic comes last as it has the largest value). This option
is <em>True</em> by default as it seems more intuitive to have charged residues red
and hydrophobic ones blue.</p>
</dd></dl>

<dl class="function">
<dt id="weblogo.LogoOverlay">
<tt class="descclassname">weblogo.</tt><tt class="descname">LogoOverlay</tt><big>(</big><em>sites</em>, <em>overlayfile</em>, <em>overlay</em>, <em>nperline</em>, <em>sitewidth</em>, <em>rmargin</em>, <em>logoheight</em>, <em>barheight</em>, <em>barspacing</em>, <em>fix_limits={}</em>, <em>fixlongname=False</em>, <em>overlay_cmap=None</em><big>)</big><a class="headerlink" href="#weblogo.LogoOverlay" title="Permalink to this definition">¶</a></dt>
<dd><p>Makes overlay for <em>LogoPlot</em>.</p>
<p>This function creates colored bars overlay bars showing up to two
properties.
The trick of this function is to create the bars the right
size so they align when they overlay the logo plot.</p>
<p>CALLING VARIABLES:</p>
<ul class="simple">
<li><em>sites</em> : same as the variable of this name used by <em>LogoPlot</em>.</li>
<li><em>overlayfile</em> is a string giving the name of created PDF file containing
the overlay. It must end in the extension <tt class="docutils literal"><span class="pre">.pdf</span></tt>.</li>
<li><em>overlay</em> : same as the variable of this name used by <em>LogoPlot</em>.</li>
<li><em>nperline</em> : same as the variable of this name used by <em>LogoPlot</em>.</li>
<li><em>sitewidth</em> is the width of each site in points.</li>
<li><em>rmargin</em> is the right margin in points.</li>
<li><em>logoheight</em> is the total height of each logo row in points.</li>
<li><em>barheight</em> is the total height of each bar in points.</li>
<li><em>barspacing</em> is the vertical spacing between bars in points.</li>
<li><em>fix_limits</em> has the same meaning of the variable of this name used by <em>LogoPlot</em>.</li>
<li><em>fixlongname</em> has the same meaning of the variable of this name used by <em>LogoPlot</em>.</li>
<li><em>overlay_cmap</em> has the same meaning of the variable of this name used by <em>LogoPlot</em>.</li>
</ul>
</dd></dl>

<dl class="function">
<dt id="weblogo.LogoPlot">
<tt class="descclassname">weblogo.</tt><tt class="descname">LogoPlot</tt><big>(</big><em>sites</em>, <em>datatype</em>, <em>data</em>, <em>plotfile</em>, <em>nperline</em>, <em>numberevery=10</em>, <em>allowunsorted=False</em>, <em>ydatamax=1.01</em>, <em>overlay=None</em>, <em>fix_limits={}</em>, <em>fixlongname=False</em>, <em>overlay_cmap=None</em><big>)</big><a class="headerlink" href="#weblogo.LogoPlot" title="Permalink to this definition">¶</a></dt>
<dd><p>Constructs a sequence logo showing amino-acid or nucleotide preferences.</p>
<p>The heights of each letter is equal to the preference of
that site for that amino acid or nucleotide.</p>
<p>Note that stop codons may or may not be included in the logo
depending on whether they are present in <em>pi_d</em>.</p>
<p>CALLING VARIABLES:</p>
<ul>
<li><p class="first"><em>sites</em> is a list of all of the sites that are being included
in the logo, as strings. They must be in natural sort order (as
is done by <em>dms_tools.utils.NaturalSort</em>) or an error will
be raised <strong>unless</strong> <em>allowunsorted</em> is <em>True</em>. The sites
in the plot are ordered in the same arrangement
listed in <em>sites</em>. These should be <strong>strings</strong>, not integers.</p>
</li>
<li><p class="first"><em>datatype</em> should be one of the two following strings depending on whether
we are making a plot of preferences or differential preferences: 
&#8216;prefs&#8217; or &#8216;diffprefs&#8217;</p>
</li>
<li><p class="first"><em>data</em> is a dictionary that has a key for every entry in
<em>sites</em>. For every site <em>r</em> in <em>sites</em>, <em>sites[r][x]</em>
is the preference or differential preference for character <em>x</em>. 
Preferences must sum to one; differential preferences to zero.
All sites must have the same set of characters. The characters
must be the set of nucleotides (<em>dms_tools.nts</em>)
or the set of amino acids with or without stop codons
(<em>dms_tools.aminoacids_nostop</em> or <em>dms_tools.aminoacids_withstop</em>).</p>
</li>
<li><p class="first"><em>plotfile</em> is a string giving the name of the created PDF file 
of the logo plot.
It must end in the extension <tt class="docutils literal"><span class="pre">.pdf</span></tt>.</p>
</li>
<li><p class="first"><em>nperline</em> is the number of sites per line. Often 40 to 80 are good values.</p>
</li>
<li><p class="first"><em>numberevery</em> is specifies how frequently we put labels for the sites on
x-axis.</p>
</li>
<li><p class="first"><em>allowunsorted</em> : if <em>True</em> then we allow the entries in <em>sites</em> to 
<strong>not</strong> be sorted. This means that the logo plot will <strong>not</strong> have
sites in sorted order.</p>
</li>
<li><p class="first"><em>ydatamax</em> : meaningful only if <em>datatype</em> is &#8216;diffprefs&#8217;. In this case, it gives
the maximum that the logo stacks extend in the positive and negative directions.
Cannot be smaller than the maximum extent of the differential preferences.</p>
</li>
<li><p class="first"><em>overlay</em> : this argument allows you to make overlay bars that indicated
other properties for the sites. By default, this option is <em>None</em>, meaning that
no overlay is created. If you set it to something else, it must be a list
giving either one or two properties. Each property is a tuple:
<em>(prop_d, shortname, longname)</em> where:</p>
<blockquote>
<div><ul>
<li><p class="first"><em>prop_d</em> is a dictionary keyed by site numbers that are in <em>sites</em>.
For each <em>r</em> in <em>sites</em>, <em>prop_d[r]</em> gives the value of the property,
or if there is no entry in <em>prop_d</em> for <em>r</em>, then the property
is undefined and is colored white. Properties can either be:</p>
<blockquote>
<div><ul class="simple">
<li>continuous: in this case, all of the values should be numbers.</li>
<li>discrete : in this case, all of the values should be strings.
While in practice, if you have more than a few discrete
categories (different strings), the plot will be a mess.</li>
</ul>
</div></blockquote>
</li>
<li><p class="first"><em>shortname</em> : short name for the property; will not format well
if more than 4 or 5 characters.</p>
</li>
<li><p class="first"><em>longname</em> : longer name for property used on axes label. Can be the
same as <em>shortname</em> if you don&#8217;t need a different long name.</p>
</li>
</ul>
</div></blockquote>
</li>
<li><p class="first"><em>fix_limits</em> is only meaningful if <em>overlay</em> is being used. In this case, for any
<em>shortname</em> in <em>overlay</em> that also keys and entry in <em>fix_limits</em>, we use
<em>fix_limits[shortname]</em> to set the limits for <em>shortname</em>. Specifically,
<em>fix_limits[shortname]</em> should be the 2-tuple <em>(ticks, ticknames)</em>. <em>ticks</em>
should be a list of tick locations (numbers) and <em>ticknames</em> should be a list of
the corresponding tick label for that tick.</p>
</li>
<li><p class="first">If <em>fixlongname</em> is <em>True</em>, then we use the <em>longname</em> in <em>overlay</em> exactly as written;
otherwise we add a parenthesis indicating the <em>shortname</em> for which this <em>longname</em>
stands.</p>
</li>
<li><p class="first"><em>overlay_cmap</em> can be the name of a valid <em>matplotlib.colors.Colormap</em>, such as the
string <em>jet</em> or <em>bwr</em>. Otherwise, it can be <em>None</em> and a (hopefully) good choice will 
be made for you.</p>
</li>
</ul>
</dd></dl>

</div>
</div>
<span class="target" id="module-simulate"></span><div class="section" id="simulate-module">
<h2><tt class="docutils literal"><span class="pre">simulate</span></tt> module<a class="headerlink" href="#simulate-module" title="Permalink to this headline">¶</a></h2>
<p>Module for simulating test data for analysis with <tt class="docutils literal"><span class="pre">dms_tools</span></tt>.</p>
<p>Written by Jesse Bloom.</p>
<div class="section" id="id9">
<h3>Functions in this module<a class="headerlink" href="#id9" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><em>SimulateLibraryCounts</em> : simulates counts for testing inference of site-specific preferences.</li>
<li><em>SimulatedDiffPrefCounts</em> : simulates counts for testing inference of differential preferences.</li>
</ul>
</div>
<div class="section" id="id10">
<h3>Function documentation<a class="headerlink" href="#id10" title="Permalink to this headline">¶</a></h3>
<dl class="function">
<dt id="simulate.SimulateDiffPrefCounts">
<tt class="descclassname">simulate.</tt><tt class="descname">SimulateDiffPrefCounts</tt><big>(</big><em>depths_outfiles</em>, <em>start_counts</em>, <em>diffprefs_file</em>, <em>avgepsilon</em>, <em>control_concentration=50.0</em>, <em>fracdiffer=0.25</em>, <em>diff_concentration=0.5</em>, <em>epsilon_concentration=300.0</em>, <em>seed=1</em><big>)</big><a class="headerlink" href="#simulate.SimulateDiffPrefCounts" title="Permalink to this definition">¶</a></dt>
<dd><p>Simulates counts data for testing differential preference inference.</p>
<p>Use this function if you want to simulate creation of a mutant library with known differential
preferences and then test how accurately you can infer them.</p>
<p>This function assumes mutations made at codon level, and that the site-specific preferences
are at the amino-acid level.</p>
<p>CALLING VARIABLES:</p>
<ul>
<li><p class="first"><em>depths_outfiles</em> : list of 2-tuples <em>(depth, outfile)</em>.</p>
<blockquote>
<div><ul>
<li><p class="first"><em>depth</em> is an integer giving the per-site sequencing depth.</p>
</li>
<li><p class="first"><em>outfile</em> is a string giving the prefix for the created
deep mutational scanning counts file. These files are deleted
if they already exist, and then re-created. The suffixes are:</p>
<blockquote>
<div><ul class="simple">
<li><tt class="docutils literal"><span class="pre">_start.txt</span></tt> starting library counts</li>
<li><tt class="docutils literal"><span class="pre">_s1.txt</span></tt> control selection counts</li>
<li><tt class="docutils literal"><span class="pre">_s2.txt</span></tt> alternate selection counts.</li>
<li><tt class="docutils literal"><span class="pre">_err.txt</span></tt> error control counts, only created if <em>avgepsilon</em> is not zero.</li>
</ul>
</div></blockquote>
</li>
</ul>
</div></blockquote>
</li>
<li><p class="first"><em>start_counts</em> : an existing deep mutational scanning counts file; the starting counts are proportional to these.</p>
</li>
<li><p class="first"><em>diffprefs_file</em> : created file giving differential preferences used in simulation.</p>
</li>
<li><p class="first"><em>avgepsilon</em> : average error rate in sequencing. This is the average
rate at which each nucleotide is mis-read as some other nucleotide. Set to zero
if there are no errors.</p>
</li>
<li><p class="first"><em>control_concentration</em> : concentration for &#8220;control&#8221; preferences.</p>
</li>
<li><p class="first"><em>fracdiffer</em> fraction of sites for which differential preferences are non-zero.</p>
</li>
<li><p class="first"><em>diff_concentration</em> : concentration parameter for differential preferences.</p>
</li>
<li><p class="first"><em>epsilon_concentration</em> : concentration parameter for errors.</p>
</li>
<li><p class="first"><em>seed</em> : random number seed.</p>
</li>
</ul>
<p>RESULT OF THIS FUNCTION:</p>
<p>The result of this function is creation of the files specified in <em>depths_outfiles</em>.</p>
</dd></dl>

<dl class="function">
<dt id="simulate.SimulateLibraryCounts">
<tt class="descclassname">simulate.</tt><tt class="descname">SimulateLibraryCounts</tt><big>(</big><em>depths_outfiles</em>, <em>prefs_file</em>, <em>avgmu</em>, <em>avgepsilon</em>, <em>avgrho</em>, <em>mu_concentration=300.0</em>, <em>epsilon_concentration=300.0</em>, <em>rho_concentration=300.0</em>, <em>seed=1</em><big>)</big><a class="headerlink" href="#simulate.SimulateLibraryCounts" title="Permalink to this definition">¶</a></dt>
<dd><p>Simulates counts data for deep mutational scanning experiment.</p>
<p>This function can be used if you want to simulate the creation of a mutant library
with a set of known site-specific preferences to then test how accurately you can
infer these preferences.</p>
<p>This function assumes mutations made at codon level, and that the site-specific preferences
are at the amino-acid level.</p>
<p>All random variables are drawn from Dirichlet distributions with the indicated
concentration parameters.</p>
<p>CALLING VARIABLES:</p>
<ul>
<li><p class="first"><em>depths_outfiles</em> : list of 2-tuples <em>(depth, outfile)</em>.</p>
<blockquote>
<div><ul>
<li><p class="first"><em>depth</em> is an integer giving the per-site sequencing depth.</p>
</li>
<li><p class="first"><em>outfile</em> is a string giving the prefix for the created
deep mutational scanning counts file. These files are deleted
if they already exist, and then re-created. The suffixes are:</p>
<blockquote>
<div><ul class="simple">
<li><tt class="docutils literal"><span class="pre">_pre.txt</span></tt> for pre-selection counts</li>
<li><tt class="docutils literal"><span class="pre">_post.txt</span></tt> for post-selection counts</li>
<li><tt class="docutils literal"><span class="pre">_errpre.txt</span></tt> for pre-selection error-control counts,
only created if <em>avgepsilon</em> is not zero.</li>
<li><tt class="docutils literal"><span class="pre">_errpost.txt</span></tt> for post-selection error-control counts,
only created if <em>avgrho</em> is not zero.</li>
</ul>
</div></blockquote>
</li>
</ul>
</div></blockquote>
</li>
<li><p class="first"><em>prefs_file</em> : name of an existing file giving the actual site-specific amino-acid
preferences, either with or without stop codons. Data are simulated using these preferences.
If stop codons are not specified, the preferences for such codons are set to zero.</p>
</li>
<li><p class="first"><em>avgmu</em> : average mutation rate at each site to <strong>any</strong> non-wildtype codon. The actual
mutation rate to each possible codon is drawn from a Dirichlet.</p>
</li>
<li><p class="first"><em>avgepsilon</em> : average error rate in pre-selection library. This is the average
rate at which each nucleotide is mis-read as some other nucleotide. Set to zero
if there are no errors.</p>
</li>
<li><p class="first"><em>avgrho</em> : average error rate in post-selection library. This is the average
rate at which each nucleotide is mis-read as some other nucleotide. Set to zero
if there are no errors.</p>
</li>
<li><p class="first"><em>mu_concentration</em> : mutation rates for any given codon drawn
from Dirichlet with this concentration parameter.</p>
</li>
<li><p class="first"><em>epsilon_concentration</em> : concentration parameter for pre-selection error rate.</p>
</li>
<li><p class="first"><em>rho_concentration</em> : concentratiion parameter for post-selection error rate.</p>
</li>
<li><p class="first"><em>seed</em> is seed for random number generators</p>
</li>
</ul>
<p>RESULT OF THIS FUNCTION:</p>
<p>The result of this function is creation of the files specified in <em>depths_outfiles</em>.</p>
</dd></dl>

</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Python API</a><ul>
<li><a class="reference internal" href="#dms-tools-package"><tt class="docutils literal"><span class="pre">dms_tools</span></tt> package</a><ul>
<li><a class="reference internal" href="#modules-in-the-package">Modules in the package</a></li>
<li><a class="reference internal" href="#package-level-variables">Package-level variables</a></li>
</ul>
</li>
<li><a class="reference internal" href="#mcmc-module"><tt class="docutils literal"><span class="pre">mcmc</span></tt> module</a><ul>
<li><a class="reference internal" href="#functions-in-this-module">Functions in this module</a></li>
<li><a class="reference internal" href="#function-documentation">Function documentation</a></li>
</ul>
</li>
<li><a class="reference internal" href="#file-io-module"><tt class="docutils literal"><span class="pre">file_io</span></tt> module</a><ul>
<li><a class="reference internal" href="#id1">Functions in this module</a></li>
<li><a class="reference internal" href="#id2">Function documentation</a></li>
</ul>
</li>
<li><a class="reference internal" href="#utils-module"><tt class="docutils literal"><span class="pre">utils</span></tt> module</a><ul>
<li><a class="reference internal" href="#id3">Functions in this module</a></li>
<li><a class="reference internal" href="#id4">Function documentation</a></li>
</ul>
</li>
<li><a class="reference internal" href="#parsearguments-module"><tt class="docutils literal"><span class="pre">parsearguments</span></tt> module</a><ul>
<li><a class="reference internal" href="#defined-in-this-module">Defined in this module</a></li>
<li><a class="reference internal" href="#detailed-documentation">Detailed documentation</a></li>
</ul>
</li>
<li><a class="reference internal" href="#plot-module"><tt class="docutils literal"><span class="pre">plot</span></tt> module</a><ul>
<li><a class="reference internal" href="#id5">Functions in this module</a></li>
<li><a class="reference internal" href="#id6">Function documentation</a></li>
</ul>
</li>
<li><a class="reference internal" href="#weblogo-module"><tt class="docutils literal"><span class="pre">weblogo</span></tt> module</a><ul>
<li><a class="reference internal" href="#id7">Functions in this module</a></li>
<li><a class="reference internal" href="#id8">Function documentation</a></li>
</ul>
</li>
<li><a class="reference internal" href="#simulate-module"><tt class="docutils literal"><span class="pre">simulate</span></tt> module</a><ul>
<li><a class="reference internal" href="#id9">Functions in this module</a></li>
<li><a class="reference internal" href="#id10">Function documentation</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="inferdiffprefs_algorithm.html"
                        title="previous chapter">Algorithm to infer differential preferences</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="citations.html"
                        title="next chapter">Citations and Acknowledgments</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/pythonapi.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="citations.html" title="Citations and Acknowledgments"
             >next</a> |</li>
        <li class="right" >
          <a href="inferdiffprefs_algorithm.html" title="Algorithm to infer differential preferences"
             >previous</a> |</li>
        <li><a href="index.html">dms_tools 1.1.6 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2014, Jesse Bloom.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3a0.
    </div>
  </body>
</html>